// @metamask/smart-accounts-kit/dist/chunk-2ZOWIAP3.cjs
"use strict";Object.defineProperty(exports, "__esModule", {value: true}); function _nullishCoalesce(lhs, rhsFn) { if (lhs != null) { return lhs; } else { return rhsFn(); } } function _optionalChain(ops) { let lastAccessLHS = undefined; let value = ops[0]; let i = 1; while (i < ops.length) { const op = ops[i]; const fn = ops[i + 1]; i += 2; if ((op === 'optionalAccess' || op === 'optionalCall') && value == null) { return undefined; } if (op === 'access' || op === 'optionalAccess') { lastAccessLHS = value; value = fn(value); } else if (op === 'call' || op === 'optionalCall') { value = fn((...args) => value.call(lastAccessLHS, ...args)); lastAccessLHS = undefined; } } return value; }





var _chunkPQPAEHJZcjs = require('./chunk-PQPAEHJZ.cjs');



var _chunk2OS36DVVcjs = require('./chunk-2OS36DVV.cjs');














var _chunkW6HPAK4Mcjs = require('./chunk-W6HPAK4M.cjs');

// src/actions/index.ts
var actions_exports = {};
_chunkW6HPAK4Mcjs.__export.call(void 0, actions_exports, {
  caveatEnforcerActions: () => caveatEnforcerActions,
  erc7710BundlerActions: () => erc7710BundlerActions,
  erc7710WalletActions: () => erc7710WalletActions,
  erc7715ProviderActions: () => erc7715ProviderActions,
  getErc20PeriodTransferEnforcerAvailableAmount: () => getErc20PeriodTransferEnforcerAvailableAmount,
  getErc20StreamingEnforcerAvailableAmount: () => getErc20StreamingEnforcerAvailableAmount,
  getGrantedExecutionPermissions: () => erc7715GetGrantedExecutionPermissionsAction,
  getMultiTokenPeriodEnforcerAvailableAmount: () => getMultiTokenPeriodEnforcerAvailableAmount,
  getNativeTokenPeriodTransferEnforcerAvailableAmount: () => getNativeTokenPeriodTransferEnforcerAvailableAmount,
  getNativeTokenStreamingEnforcerAvailableAmount: () => getNativeTokenStreamingEnforcerAvailableAmount,
  getSupportedExecutionPermissions: () => erc7715GetSupportedExecutionPermissionsAction,
  isValid7702Implementation: () => isValid7702Implementation,
  redelegatePermissionContextAction: () => redelegatePermissionContextAction,
  redelegatePermissionContextOpenAction: () => redelegatePermissionContextOpenAction,
  requestExecutionPermissions: () => erc7715RequestExecutionPermissionsAction,
  signDelegation: () => signDelegation,
  signDelegationActions: () => signDelegationActions,
  signUserOperation: () => signUserOperation,
  signUserOperationActions: () => signUserOperationActions
});

// src/actions/erc7710RedeemDelegationAction.ts
var _delegationabis = require('@metamask/delegation-abis');
var _viem = require('viem');
async function sendTransactionWithDelegationAction(client, args) {
  _chunkW6HPAK4Mcjs.trackSmartAccountsKitFunctionCall.call(void 0, "sendTransactionWithDelegationAction", {
    chainId: _nullishCoalesce(_optionalChain([client, 'access', _ => _.chain, 'optionalAccess', _2 => _2.id]), () => ( null))
  });
  if (!args.to) {
    throw new Error(
      "`to` is required. `sendTransactionWithDelegation` cannot be used to deploy contracts."
    );
  }
  const chainId = _optionalChain([client, 'access', _3 => _3.chain, 'optionalAccess', _4 => _4.id]);
  if (!chainId) {
    throw new Error("Chain ID is not set");
  }
  const { DelegationManager: expectedDelegationManager } = _chunk2OS36DVVcjs.getSmartAccountsEnvironment.call(void 0, chainId);
  if (!_viem.isAddressEqual.call(void 0, args.delegationManager, expectedDelegationManager)) {
    throw new Error(
      `Invalid DelegationManager: expected ${expectedDelegationManager} for chain ${chainId}, but got ${args.delegationManager}`
    );
  }
  const executions = [
    _chunkW6HPAK4Mcjs.createExecution.call(void 0, {
      target: args.to,
      value: args.value,
      callData: args.data
    })
  ];
  const calldata = _viem.encodeFunctionData.call(void 0, {
    abi: _delegationabis.DelegationManager,
    functionName: "redeemDelegations",
    args: [
      [_chunkW6HPAK4Mcjs.encodeDelegations.call(void 0, args.permissionContext)],
      ["0x0000000000000000000000000000000000000000000000000000000000000000" /* SingleDefault */],
      _chunkW6HPAK4Mcjs.encodeExecutionCalldatas.call(void 0, [executions])
    ]
  });
  const {
    value: _value,
    permissionContext: _permissionContext,
    delegationManager: _delegationManager,
    ...rest
  } = args;
  const hash = await client.sendTransaction({
    ...rest,
    to: args.delegationManager,
    data: calldata
  });
  return hash;
}
async function sendUserOperationWithDelegationAction(client, parameters) {
  _chunkW6HPAK4Mcjs.trackSmartAccountsKitFunctionCall.call(void 0, "sendUserOperationWithDelegationAction", {
    chainId: _nullishCoalesce(_optionalChain([client, 'access', _5 => _5.chain, 'optionalAccess', _6 => _6.id]), () => ( null)),
    callCount: parameters.calls.length,
    hasDependencies: parameters.dependencies !== void 0
  });
  if (parameters.dependencies) {
    const { publicClient } = parameters;
    const includedAccountKeys = {};
    const chainId = _optionalChain([publicClient, 'access', _7 => _7.chain, 'optionalAccess', _8 => _8.id]);
    if (!chainId) {
      throw new Error("Chain ID is not set");
    }
    const { SimpleFactory } = _chunk2OS36DVVcjs.getSmartAccountsEnvironment.call(void 0, chainId);
    const uniqueDependencies = parameters.dependencies.filter((dependency) => {
      if (!_viem.isAddressEqual.call(void 0, dependency.factory, SimpleFactory)) {
        throw new Error(
          `Invalid dependency: ${dependency.factory} is not allowed.`
        );
      }
      const accountKey = _viem.concat.call(void 0, [dependency.factory, dependency.factoryData]);
      const isDuplicate = includedAccountKeys[accountKey];
      includedAccountKeys[accountKey] = true;
      return !isDuplicate;
    });
    const factoryCalls = (await Promise.all(
      uniqueDependencies.map(async ({ factory, factoryData }) => {
        const isDeployed = await publicClient.call({
          to: factory,
          data: factoryData
        }).then(() => false).catch(() => true);
        if (isDeployed) {
          return void 0;
        }
        return {
          to: factory,
          value: 0n,
          data: factoryData
        };
      })
    )).filter((call) => call !== void 0);
    parameters.calls = [
      ...factoryCalls,
      ...parameters.calls
    ];
  }
  parameters.calls = parameters.calls.map((call) => {
    if (!("permissionContext" in call)) {
      return call;
    }
    const { permissionContext } = call;
    if (!permissionContext) {
      return call;
    }
    return {
      ...call,
      permissionContext: _chunkW6HPAK4Mcjs.encodeDelegations.call(void 0, permissionContext)
    };
  });
  return client.sendUserOperation(
    parameters
  );
}

// src/actions/erc7715Mapping.ts

function permissionRequestToRpc(parameters) {
  const { chainId, from, expiry, redeemer, payee } = parameters;
  const converter = getPermissionRequestToRpcConverter(
    parameters.permission.type
  );
  const rules = [];
  if (_chunkW6HPAK4Mcjs.isDefined.call(void 0, expiry)) {
    rules.push({
      type: "expiry",
      data: {
        timestamp: expiry
      }
    });
  }
  if (_chunkW6HPAK4Mcjs.isDefined.call(void 0, redeemer)) {
    if (redeemer.length === 0) {
      throw new Error(
        "Invalid redeemers: must specify at least one redeemer address"
      );
    }
    const addresses = [];
    for (const addr of redeemer) {
      if (!_viem.isAddress.call(void 0, addr)) {
        throw new Error("Invalid redeemers: must be a valid address");
      }
      addresses.push(_viem.getAddress.call(void 0, addr));
    }
    rules.push({
      type: "redeemer",
      data: { addresses }
    });
  }
  if (_chunkW6HPAK4Mcjs.isDefined.call(void 0, payee)) {
    if (payee.length === 0) {
      throw new Error(
        "Invalid payees: must specify at least one payee address"
      );
    }
    const payeeAddresses = [];
    for (const addr of payee) {
      if (!_viem.isAddress.call(void 0, addr)) {
        throw new Error("Invalid payees: must be a valid address");
      }
      payeeAddresses.push(_viem.getAddress.call(void 0, addr));
    }
    rules.push({
      type: "payee",
      data: { addresses: payeeAddresses }
    });
  }
  const optionalFields = {
    ...from ? { from } : {}
  };
  return {
    ...optionalFields,
    chainId: _viem.toHex.call(void 0, chainId),
    permission: converter(parameters.permission),
    to: parameters.to,
    rules
  };
}
function getPermissionRequestToRpcConverter(permissionType) {
  switch (permissionType) {
    case "native-token-stream":
      return (permission) => nativeTokenStreamPermissionToRpc(
        permission
      );
    case "erc20-token-stream":
      return (permission) => erc20TokenStreamPermissionToRpc(
        permission
      );
    case "native-token-periodic":
      return (permission) => nativeTokenPeriodicPermissionToRpc(
        permission
      );
    case "native-token-allowance":
      return (permission) => nativeTokenAllowancePermissionToRpc(
        permission
      );
    case "erc20-token-periodic":
      return (permission) => erc20TokenPeriodicPermissionToRpc(
        permission
      );
    case "erc20-token-allowance":
      return (permission) => erc20TokenAllowancePermissionToRpc(
        permission
      );
    case "erc20-token-revocation":
      return (permission) => erc20TokenRevocationPermissionToRpc(
        permission
      );
    case "token-approval-revocation":
      return (permission) => tokenApprovalRevocationPermissionToRpc(
        permission
      );
    default:
      throw new Error(`Unsupported permission type: ${permissionType}`);
  }
}
function nativeTokenStreamPermissionToRpc(permission) {
  const {
    data: {
      initialAmount,
      justification,
      maxAmount,
      startTime,
      amountPerSecond
    },
    isAdjustmentAllowed
  } = permission;
  const optionalFields = {
    ..._chunkW6HPAK4Mcjs.isDefined.call(void 0, initialAmount) && {
      initialAmount: _chunkW6HPAK4Mcjs.toHexOrThrow.call(void 0, initialAmount, "initialAmount")
    },
    ..._chunkW6HPAK4Mcjs.isDefined.call(void 0, maxAmount) && {
      maxAmount: _chunkW6HPAK4Mcjs.toHexOrThrow.call(void 0, maxAmount, "maxAmount")
    },
    ..._chunkW6HPAK4Mcjs.isDefined.call(void 0, startTime) && {
      startTime: Number(startTime)
    },
    ...justification ? { justification } : {}
  };
  return {
    type: "native-token-stream",
    data: {
      amountPerSecond: _chunkW6HPAK4Mcjs.toHexOrThrow.call(void 0, amountPerSecond, "amountPerSecond"),
      ...optionalFields
    },
    isAdjustmentAllowed
  };
}
function erc20TokenStreamPermissionToRpc(permission) {
  const {
    data: {
      tokenAddress,
      amountPerSecond,
      initialAmount,
      startTime,
      maxAmount,
      justification
    },
    isAdjustmentAllowed
  } = permission;
  const optionalFields = {
    ..._chunkW6HPAK4Mcjs.isDefined.call(void 0, initialAmount) && {
      initialAmount: _chunkW6HPAK4Mcjs.toHexOrThrow.call(void 0, initialAmount, "initialAmount")
    },
    ..._chunkW6HPAK4Mcjs.isDefined.call(void 0, maxAmount) && {
      maxAmount: _chunkW6HPAK4Mcjs.toHexOrThrow.call(void 0, maxAmount, "maxAmount")
    },
    ..._chunkW6HPAK4Mcjs.isDefined.call(void 0, startTime) && {
      startTime: Number(startTime)
    },
    ...justification ? { justification } : {}
  };
  return {
    type: "erc20-token-stream",
    data: {
      tokenAddress: _chunkW6HPAK4Mcjs.toHexOrThrow.call(void 0, tokenAddress, "tokenAddress"),
      amountPerSecond: _chunkW6HPAK4Mcjs.toHexOrThrow.call(void 0, amountPerSecond, "amountPerSecond"),
      ...optionalFields
    },
    isAdjustmentAllowed
  };
}
function nativeTokenPeriodicPermissionToRpc(permission) {
  const {
    data: { periodAmount, periodDuration, startTime, justification },
    isAdjustmentAllowed
  } = permission;
  const optionalFields = {
    ..._chunkW6HPAK4Mcjs.isDefined.call(void 0, startTime) && {
      startTime: Number(startTime)
    },
    ...justification ? { justification } : {}
  };
  return {
    type: "native-token-periodic",
    data: {
      periodAmount: _chunkW6HPAK4Mcjs.toHexOrThrow.call(void 0, periodAmount, "periodAmount"),
      periodDuration: Number(periodDuration),
      ...optionalFields
    },
    isAdjustmentAllowed
  };
}
function erc20TokenPeriodicPermissionToRpc(permission) {
  const {
    data: {
      tokenAddress,
      periodAmount,
      periodDuration,
      startTime,
      justification
    },
    isAdjustmentAllowed
  } = permission;
  const optionalFields = {
    ..._chunkW6HPAK4Mcjs.isDefined.call(void 0, startTime) && {
      startTime: Number(startTime)
    },
    ...justification ? { justification } : {}
  };
  return {
    type: "erc20-token-periodic",
    data: {
      tokenAddress: _chunkW6HPAK4Mcjs.toHexOrThrow.call(void 0, tokenAddress, "tokenAddress"),
      periodAmount: _chunkW6HPAK4Mcjs.toHexOrThrow.call(void 0, periodAmount, "periodAmount"),
      periodDuration: Number(periodDuration),
      ...optionalFields
    },
    isAdjustmentAllowed
  };
}
function nativeTokenAllowancePermissionToRpc(permission) {
  const {
    data: { allowanceAmount, startTime, justification },
    isAdjustmentAllowed
  } = permission;
  const optionalFields = {
    ..._chunkW6HPAK4Mcjs.isDefined.call(void 0, startTime) && {
      startTime: Number(startTime)
    },
    ...justification ? { justification } : {}
  };
  return {
    type: "native-token-allowance",
    data: {
      allowanceAmount: _chunkW6HPAK4Mcjs.toHexOrThrow.call(void 0, allowanceAmount, "allowanceAmount"),
      ...optionalFields
    },
    isAdjustmentAllowed
  };
}
function erc20TokenAllowancePermissionToRpc(permission) {
  const {
    data: { tokenAddress, allowanceAmount, startTime, justification },
    isAdjustmentAllowed
  } = permission;
  const optionalFields = {
    ..._chunkW6HPAK4Mcjs.isDefined.call(void 0, startTime) && {
      startTime: Number(startTime)
    },
    ...justification ? { justification } : {}
  };
  return {
    type: "erc20-token-allowance",
    data: {
      tokenAddress: _chunkW6HPAK4Mcjs.toHexOrThrow.call(void 0, tokenAddress, "tokenAddress"),
      allowanceAmount: _chunkW6HPAK4Mcjs.toHexOrThrow.call(void 0, allowanceAmount, "allowanceAmount"),
      ...optionalFields
    },
    isAdjustmentAllowed
  };
}
function erc20TokenRevocationPermissionToRpc(permission) {
  const {
    data: { justification },
    isAdjustmentAllowed
  } = permission;
  const data = {
    ...justification ? { justification } : {}
  };
  return {
    type: "erc20-token-revocation",
    data,
    isAdjustmentAllowed
  };
}
function tokenApprovalRevocationPermissionToRpc(permission) {
  const {
    data: {
      erc20Approve,
      erc721Approve,
      erc721SetApprovalForAll,
      permit2Approve,
      permit2Lockdown,
      permit2InvalidateNonces,
      justification
    },
    isAdjustmentAllowed
  } = permission;
  const data = {
    erc20Approve,
    erc721Approve,
    erc721SetApprovalForAll,
    permit2Approve,
    permit2Lockdown,
    permit2InvalidateNonces,
    ...justification ? { justification } : {}
  };
  return {
    type: "token-approval-revocation",
    data,
    isAdjustmentAllowed
  };
}
function permissionResponsesFromRpc(result) {
  return result.map((permission) => ({
    ...permission,
    chainId: _viem.hexToNumber.call(void 0, permission.chainId),
    permission: permissionTypeFromRpc(permission.permission),
    rules: normalizeRulesFromRpc(permission.rules)
  }));
}
function normalizeRulesFromRpc(rules) {
  if (rules === void 0 || rules === null) {
    return rules;
  }
  return rules.map((rule) => {
    if (rule.type !== "redeemer" && rule.type !== "payee") {
      return rule;
    }
    const rawAddresses = _optionalChain([rule, 'access', _9 => _9.data, 'optionalAccess', _10 => _10.addresses]);
    if (!Array.isArray(rawAddresses)) {
      return rule;
    }
    return {
      type: rule.type,
      data: {
        addresses: rawAddresses.map((addr) => _viem.getAddress.call(void 0, addr))
      }
    };
  });
}
function permissionTypeFromRpc(permission) {
  const convertedData = { ...permission.data };
  if ("amountPerSecond" in convertedData && convertedData.amountPerSecond) {
    convertedData.amountPerSecond = BigInt(
      convertedData.amountPerSecond
    );
  }
  if ("periodAmount" in convertedData && convertedData.periodAmount) {
    convertedData.periodAmount = BigInt(
      convertedData.periodAmount
    );
  }
  if ("initialAmount" in convertedData && convertedData.initialAmount) {
    convertedData.initialAmount = BigInt(
      convertedData.initialAmount
    );
  }
  if ("maxAmount" in convertedData && convertedData.maxAmount) {
    convertedData.maxAmount = BigInt(convertedData.maxAmount);
  }
  if ("allowanceAmount" in convertedData && convertedData.allowanceAmount) {
    convertedData.allowanceAmount = BigInt(
      convertedData.allowanceAmount
    );
  }
  return {
    ...permission,
    data: convertedData
  };
}
function rpcSupportedPermissionsToDeveloper(result) {
  const converted = {};
  for (const [permissionType, permissionInfo] of Object.entries(result)) {
    converted[permissionType] = {
      chainIds: permissionInfo.chainIds.map((chainId) => _viem.hexToNumber.call(void 0, chainId)),
      ruleTypes: permissionInfo.ruleTypes
    };
  }
  return converted;
}

// src/actions/erc7715GetGrantedExecutionPermissionsAction.ts
async function erc7715GetGrantedExecutionPermissionsAction(client) {
  _chunkW6HPAK4Mcjs.trackSmartAccountsKitFunctionCall.call(void 0, 
    "erc7715GetGrantedExecutionPermissionsAction",
    {
      chainId: _nullishCoalesce(_optionalChain([client, 'access', _11 => _11.chain, 'optionalAccess', _12 => _12.id]), () => ( null))
    }
  );
  const result = await client.request(
    {
      method: "wallet_getGrantedExecutionPermissions",
      params: []
    },
    { retryCount: 0 }
  );
  if (!result) {
    throw new Error("Failed to get granted execution permissions");
  }
  return permissionResponsesFromRpc(result);
}

// src/actions/erc7715GetSupportedExecutionPermissionsAction.ts
async function erc7715GetSupportedExecutionPermissionsAction(client) {
  _chunkW6HPAK4Mcjs.trackSmartAccountsKitFunctionCall.call(void 0, 
    "erc7715GetSupportedExecutionPermissionsAction",
    {
      chainId: _nullishCoalesce(_optionalChain([client, 'access', _13 => _13.chain, 'optionalAccess', _14 => _14.id]), () => ( null))
    }
  );
  const result = await client.request(
    {
      method: "wallet_getSupportedExecutionPermissions",
      params: []
    },
    { retryCount: 0 }
  );
  if (!result) {
    throw new Error("Failed to get supported execution permissions");
  }
  return rpcSupportedPermissionsToDeveloper(result);
}

// src/actions/erc7715RequestExecutionPermissionsAction.ts
async function erc7715RequestExecutionPermissionsAction(client, parameters) {
  _chunkW6HPAK4Mcjs.trackSmartAccountsKitFunctionCall.call(void 0, 
    "erc7715RequestExecutionPermissionsAction",
    {
      chainId: _nullishCoalesce(_optionalChain([client, 'access', _15 => _15.chain, 'optionalAccess', _16 => _16.id]), () => ( null)),
      requestCount: parameters.length
    }
  );
  const formattedPermissionRequest = parameters.map(permissionRequestToRpc);
  const result = await client.request(
    {
      method: "wallet_requestExecutionPermissions",
      params: formattedPermissionRequest
    },
    { retryCount: 0 }
  );
  if (!result) {
    throw new Error("Failed to grant permissions");
  }
  return permissionResponsesFromRpc(result);
}

// src/actions/redelegatePermissionContext.ts

var _accounts = require('viem/accounts');

// src/actions/signDelegation.ts


async function signDelegation(client, parameters) {
  const {
    account: accountParam = client.account,
    delegation,
    delegationManager,
    chainId,
    name = "DelegationManager",
    version = "1",
    allowInsecureUnrestrictedDelegation = false
  } = parameters;
  if (!accountParam) {
    throw new (0, _viem.BaseError)("Account not found. Please provide an account.");
  }
  const account = _accounts.parseAccount.call(void 0, accountParam);
  const typedData = _chunkW6HPAK4Mcjs.prepareSignDelegationTypedData.call(void 0, {
    delegation,
    delegationManager,
    chainId,
    name,
    version,
    allowInsecureUnrestrictedDelegation
  });
  return client.signTypedData({
    account,
    ...typedData
  });
}
function signDelegationActions() {
  return (client) => ({
    signDelegation: async (parameters) => signDelegation(client, {
      chainId: _nullishCoalesce(parameters.chainId, () => ( (() => {
        if (!_optionalChain([client, 'access', _17 => _17.chain, 'optionalAccess', _18 => _18.id])) {
          throw new (0, _viem.BaseError)(
            "Chain ID is required. Either provide it in parameters or configure the client with a chain."
          );
        }
        return client.chain.id;
      })())),
      ...parameters
    })
  });
}

// src/actions/redelegatePermissionContext.ts
async function signAndPrependRedelegation(client, options) {
  const { account, environment, delegations, unsignedDelegation, chainId } = options;
  const signature = await signDelegation(client, {
    account,
    delegation: unsignedDelegation,
    delegationManager: environment.DelegationManager,
    chainId,
    // Redelegations always inherit from a parent delegation (enforced by
    // `resolveRedelegationArgs`), so the parent's caveats provide the
    // restriction even when the redelegation itself adds no extra caveats.
    // This mirrors `resolveCaveats`, which also allows empty caveats in this
    // case.
    allowInsecureUnrestrictedDelegation: true
  });
  const signedDelegation = {
    ...unsignedDelegation,
    signature
  };
  const newPermissionContext = _chunkW6HPAK4Mcjs.encodeDelegations.call(void 0, [
    signedDelegation,
    ...delegations
  ]);
  return {
    delegation: signedDelegation,
    permissionContext: newPermissionContext
  };
}
function resolveRedelegationArgs(client, parameters) {
  const { account: accountParam = client.account, permissionContext } = parameters;
  if (!accountParam) {
    throw new (0, _viem.BaseError)("Account not found. Please provide an account.");
  }
  const account = _accounts.parseAccount.call(void 0, accountParam);
  const delegations = _chunkW6HPAK4Mcjs.decodeDelegations.call(void 0, permissionContext);
  const parentDelegation = delegations[0];
  if (!parentDelegation) {
    throw new (0, _viem.BaseError)(
      "Permission context must contain at least one delegation"
    );
  }
  const isParentOpenDelegation = parentDelegation.delegate.toLowerCase() === _chunkW6HPAK4Mcjs.ANY_BENEFICIARY.toLowerCase();
  const from = isParentOpenDelegation ? account.address : parentDelegation.delegate;
  return { account, delegations, parentDelegation, from };
}
async function redelegatePermissionContextAction(client, parameters) {
  const { environment, scope, caveats, salt, to } = parameters;
  const chainId = resolveChainId(client, parameters.chainId);
  const { account, delegations, parentDelegation, from } = resolveRedelegationArgs(client, parameters);
  _chunkW6HPAK4Mcjs.trackSmartAccountsKitFunctionCall.call(void 0, "redelegatePermissionContext", {
    chainId,
    hasScope: scope !== void 0,
    hasCaveats: caveats !== void 0
  });
  const unsignedDelegation = _chunkW6HPAK4Mcjs.createDelegation.call(void 0, {
    environment,
    from,
    to,
    scope,
    caveats,
    parentDelegation,
    salt
  });
  return signAndPrependRedelegation(client, {
    account,
    environment,
    delegations,
    unsignedDelegation,
    chainId
  });
}
async function redelegatePermissionContextOpenAction(client, parameters) {
  const { environment, scope, caveats, salt } = parameters;
  const chainId = resolveChainId(client, parameters.chainId);
  const { account, delegations, parentDelegation, from } = resolveRedelegationArgs(client, parameters);
  _chunkW6HPAK4Mcjs.trackSmartAccountsKitFunctionCall.call(void 0, "redelegatePermissionContextOpen", {
    chainId,
    hasScope: scope !== void 0,
    hasCaveats: caveats !== void 0
  });
  const unsignedDelegation = _chunkW6HPAK4Mcjs.createOpenDelegation.call(void 0, {
    environment,
    from,
    scope,
    caveats,
    parentDelegation,
    salt
  });
  return signAndPrependRedelegation(client, {
    account,
    environment,
    delegations,
    unsignedDelegation,
    chainId
  });
}
function resolveChainId(client, chainId) {
  if (chainId !== void 0) {
    return chainId;
  }
  if (!_optionalChain([client, 'access', _19 => _19.chain, 'optionalAccess', _20 => _20.id])) {
    throw new (0, _viem.BaseError)(
      "Chain ID is required. Either provide it in parameters or configure the client with a chain."
    );
  }
  return client.chain.id;
}

// src/actions/getCaveatAvailableAmount.ts
function findMatchingCaveat({
  delegation,
  enforcerAddress,
  enforcerName
}) {
  const matchingCaveats = delegation.caveats.filter(
    (caveat) => caveat.enforcer.toLowerCase() === enforcerAddress.toLowerCase()
  );
  if (matchingCaveats.length === 0) {
    throw new Error(`No caveat found with enforcer matching ${enforcerName}`);
  }
  if (matchingCaveats.length > 1) {
    throw new Error(
      `Multiple caveats found with enforcer matching ${enforcerName}`
    );
  }
  const [{ terms, args }] = matchingCaveats;
  return {
    terms,
    args
  };
}
function getDelegationManager(environment) {
  if (!environment.DelegationManager) {
    throw new Error("Delegation manager address not found");
  }
  return environment.DelegationManager;
}
function getEnforcerAddress({
  enforcerName,
  environment
}) {
  const enforcerAddress = environment.caveatEnforcers[enforcerName];
  if (!enforcerAddress) {
    throw new Error(`${enforcerName} not found in environment`);
  }
  return enforcerAddress;
}
async function getErc20PeriodTransferEnforcerAvailableAmount(client, environment, params) {
  _chunkW6HPAK4Mcjs.trackSmartAccountsKitFunctionCall.call(void 0, 
    "getErc20PeriodTransferEnforcerAvailableAmount",
    {
      chainId: _nullishCoalesce(_optionalChain([client, 'access', _21 => _21.chain, 'optionalAccess', _22 => _22.id]), () => ( null))
    }
  );
  const enforcerName = "ERC20PeriodTransferEnforcer";
  const delegationManager = getDelegationManager(environment);
  const enforcerAddress = getEnforcerAddress({
    enforcerName,
    environment
  });
  const delegationHash = _chunkW6HPAK4Mcjs.hashDelegation.call(void 0, params.delegation);
  const { terms } = findMatchingCaveat({
    delegation: params.delegation,
    enforcerAddress,
    enforcerName
  });
  return _chunkPQPAEHJZcjs.read_exports.getAvailableAmount({
    client,
    contractAddress: enforcerAddress,
    delegationHash,
    delegationManager,
    terms
  });
}
async function getErc20StreamingEnforcerAvailableAmount(client, environment, params) {
  _chunkW6HPAK4Mcjs.trackSmartAccountsKitFunctionCall.call(void 0, 
    "getErc20StreamingEnforcerAvailableAmount",
    {
      chainId: _nullishCoalesce(_optionalChain([client, 'access', _23 => _23.chain, 'optionalAccess', _24 => _24.id]), () => ( null))
    }
  );
  const enforcerName = "ERC20StreamingEnforcer";
  const delegationManager = getDelegationManager(environment);
  const enforcerAddress = getEnforcerAddress({
    enforcerName,
    environment
  });
  const delegationHash = _chunkW6HPAK4Mcjs.hashDelegation.call(void 0, params.delegation);
  const { terms } = findMatchingCaveat({
    delegation: params.delegation,
    enforcerAddress,
    enforcerName
  });
  return _chunkPQPAEHJZcjs.read_exports2.getAvailableAmount({
    client,
    contractAddress: enforcerAddress,
    delegationManager,
    delegationHash,
    terms
  });
}
async function getMultiTokenPeriodEnforcerAvailableAmount(client, environment, params) {
  _chunkW6HPAK4Mcjs.trackSmartAccountsKitFunctionCall.call(void 0, 
    "getMultiTokenPeriodEnforcerAvailableAmount",
    {
      chainId: _nullishCoalesce(_optionalChain([client, 'access', _25 => _25.chain, 'optionalAccess', _26 => _26.id]), () => ( null))
    }
  );
  const enforcerName = "MultiTokenPeriodEnforcer";
  const delegationManager = getDelegationManager(environment);
  const enforcerAddress = getEnforcerAddress({
    enforcerName,
    environment
  });
  const delegationHash = _chunkW6HPAK4Mcjs.hashDelegation.call(void 0, params.delegation);
  const { terms, args } = findMatchingCaveat({
    delegation: params.delegation,
    enforcerAddress,
    enforcerName
  });
  return _chunkPQPAEHJZcjs.read_exports3.getAvailableAmount({
    client,
    contractAddress: enforcerAddress,
    delegationHash,
    delegationManager,
    terms,
    args
  });
}
async function getNativeTokenPeriodTransferEnforcerAvailableAmount(client, environment, params) {
  _chunkW6HPAK4Mcjs.trackSmartAccountsKitFunctionCall.call(void 0, 
    "getNativeTokenPeriodTransferEnforcerAvailableAmount",
    {
      chainId: _nullishCoalesce(_optionalChain([client, 'access', _27 => _27.chain, 'optionalAccess', _28 => _28.id]), () => ( null))
    }
  );
  const enforcerName = "NativeTokenPeriodTransferEnforcer";
  const delegationManager = getDelegationManager(environment);
  const enforcerAddress = getEnforcerAddress({
    enforcerName,
    environment
  });
  const delegationHash = _chunkW6HPAK4Mcjs.hashDelegation.call(void 0, params.delegation);
  const { terms } = findMatchingCaveat({
    delegation: params.delegation,
    enforcerAddress,
    enforcerName
  });
  return _chunkPQPAEHJZcjs.read_exports4.getAvailableAmount({
    client,
    contractAddress: enforcerAddress,
    delegationHash,
    delegationManager,
    terms
  });
}
async function getNativeTokenStreamingEnforcerAvailableAmount(client, environment, params) {
  _chunkW6HPAK4Mcjs.trackSmartAccountsKitFunctionCall.call(void 0, 
    "getNativeTokenStreamingEnforcerAvailableAmount",
    {
      chainId: _nullishCoalesce(_optionalChain([client, 'access', _29 => _29.chain, 'optionalAccess', _30 => _30.id]), () => ( null))
    }
  );
  const enforcerName = "NativeTokenStreamingEnforcer";
  const delegationManager = getDelegationManager(environment);
  const enforcerAddress = getEnforcerAddress({
    enforcerName,
    environment
  });
  const delegationHash = _chunkW6HPAK4Mcjs.hashDelegation.call(void 0, params.delegation);
  const { terms } = findMatchingCaveat({
    delegation: params.delegation,
    enforcerAddress,
    enforcerName
  });
  return _chunkPQPAEHJZcjs.read_exports5.getAvailableAmount({
    client,
    contractAddress: enforcerAddress,
    delegationManager,
    delegationHash,
    terms
  });
}
var caveatEnforcerActions = ({ environment }) => (client) => ({
  /**
   * Get available amount for ERC20 period transfer enforcer.
   *
   * @param params - The parameters for the ERC20 period transfer enforcer.
   * @returns Promise resolving to the period transfer result.
   */
  getErc20PeriodTransferEnforcerAvailableAmount: async (params) => {
    return getErc20PeriodTransferEnforcerAvailableAmount(
      client,
      environment,
      params
    );
  },
  /**
   * Get available amount for ERC20 streaming enforcer.
   *
   * @param params - The parameters for the ERC20 streaming enforcer.
   * @returns Promise resolving to the streaming result.
   */
  getErc20StreamingEnforcerAvailableAmount: async (params) => {
    return getErc20StreamingEnforcerAvailableAmount(
      client,
      environment,
      params
    );
  },
  /**
   * Get available amount for multi-token period enforcer.
   *
   * @param params - The parameters for the multi-token period enforcer.
   * @returns Promise resolving to the period transfer result.
   */
  getMultiTokenPeriodEnforcerAvailableAmount: async (params) => {
    return getMultiTokenPeriodEnforcerAvailableAmount(
      client,
      environment,
      params
    );
  },
  /**
   * Get available amount for native token period transfer enforcer.
   *
   * @param params - The parameters for the native token period transfer enforcer.
   * @returns Promise resolving to the period transfer result.
   */
  getNativeTokenPeriodTransferEnforcerAvailableAmount: async (params) => {
    return getNativeTokenPeriodTransferEnforcerAvailableAmount(
      client,
      environment,
      params
    );
  },
  /**
   * Get available amount for native token streaming enforcer.
   *
   * @param params - The parameters for the native token streaming enforcer.
   * @returns Promise resolving to the streaming result.
   */
  getNativeTokenStreamingEnforcerAvailableAmount: async (params) => {
    return getNativeTokenStreamingEnforcerAvailableAmount(
      client,
      environment,
      params
    );
  }
});

// src/actions/isValid7702Implementation.ts

var _actions = require('viem/actions');
var DELEGATION_PREFIX = "0xef0100";
function extractDelegatedAddress(code) {
  if (_optionalChain([code, 'optionalAccess', _31 => _31.length]) !== 48) {
    return null;
  }
  if (!code.toLowerCase().startsWith(DELEGATION_PREFIX.toLowerCase())) {
    return null;
  }
  const addressHex = code.slice(8);
  return `0x${addressHex}`;
}
async function isValid7702Implementation({
  client,
  accountAddress,
  environment
}) {
  try {
    const code = await _actions.getCode.call(void 0, client, {
      address: accountAddress
    });
    const delegatedAddress = extractDelegatedAddress(code);
    if (!delegatedAddress) {
      return false;
    }
    const expectedImplementation = environment.implementations.EIP7702StatelessDeleGatorImpl;
    if (!expectedImplementation) {
      return false;
    }
    return _viem.isAddressEqual.call(void 0, delegatedAddress, expectedImplementation);
  } catch (e) {
    return false;
  }
}

// src/actions/signUserOperation.ts


async function signUserOperation(client, parameters) {
  const {
    account: accountParam = client.account,
    userOperation,
    entryPoint,
    chainId,
    name,
    address,
    version = "1"
  } = parameters;
  if (!accountParam) {
    throw new (0, _viem.BaseError)("Account not found. Please provide an account.");
  }
  const account = _accounts.parseAccount.call(void 0, accountParam);
  const typedData = _chunk2OS36DVVcjs.prepareSignUserOperationTypedData.call(void 0, {
    userOperation,
    entryPoint,
    chainId,
    name,
    address,
    version
  });
  return client.signTypedData({
    account,
    ...typedData
  });
}
function signUserOperationActions() {
  return (client) => ({
    signUserOperation: async (parameters) => signUserOperation(client, {
      chainId: _nullishCoalesce(parameters.chainId, () => ( (() => {
        if (!_optionalChain([client, 'access', _32 => _32.chain, 'optionalAccess', _33 => _33.id])) {
          throw new (0, _viem.BaseError)(
            "Chain ID is required. Either provide it in parameters or configure the client with a chain."
          );
        }
        return client.chain.id;
      })())),
      ...parameters
    })
  });
}

// src/actions/index.ts
var erc7715ProviderActions = () => (client) => ({
  requestExecutionPermissions: async (parameters) => {
    return erc7715RequestExecutionPermissionsAction(
      client,
      parameters
    );
  },
  getSupportedExecutionPermissions: async () => {
    return erc7715GetSupportedExecutionPermissionsAction(
      client
    );
  },
  getGrantedExecutionPermissions: async () => {
    return erc7715GetGrantedExecutionPermissionsAction(
      client
    );
  }
});
var erc7710WalletActions = () => (client) => ({
  sendTransactionWithDelegation: async (args) => sendTransactionWithDelegationAction(client, args),
  redelegatePermissionContext: async (parameters) => redelegatePermissionContextAction(client, parameters),
  redelegatePermissionContextOpen: async (parameters) => redelegatePermissionContextOpenAction(client, parameters)
});
var erc7710BundlerActions = () => (client) => ({
  sendUserOperationWithDelegation: async (args) => sendUserOperationWithDelegationAction(client, args)
});






















exports.isValid7702Implementation = isValid7702Implementation; exports.erc7715GetGrantedExecutionPermissionsAction = erc7715GetGrantedExecutionPermissionsAction; exports.erc7715GetSupportedExecutionPermissionsAction = erc7715GetSupportedExecutionPermissionsAction; exports.erc7715RequestExecutionPermissionsAction = erc7715RequestExecutionPermissionsAction; exports.signDelegation = signDelegation; exports.signDelegationActions = signDelegationActions; exports.redelegatePermissionContextAction = redelegatePermissionContextAction; exports.redelegatePermissionContextOpenAction = redelegatePermissionContextOpenAction; exports.getErc20PeriodTransferEnforcerAvailableAmount = getErc20PeriodTransferEnforcerAvailableAmount; exports.getErc20StreamingEnforcerAvailableAmount = getErc20StreamingEnforcerAvailableAmount; exports.getMultiTokenPeriodEnforcerAvailableAmount = getMultiTokenPeriodEnforcerAvailableAmount; exports.getNativeTokenPeriodTransferEnforcerAvailableAmount = getNativeTokenPeriodTransferEnforcerAvailableAmount; exports.getNativeTokenStreamingEnforcerAvailableAmount = getNativeTokenStreamingEnforcerAvailableAmount; exports.caveatEnforcerActions = caveatEnforcerActions; exports.signUserOperation = signUserOperation; exports.signUserOperationActions = signUserOperationActions; exports.erc7715ProviderActions = erc7715ProviderActions; exports.erc7710WalletActions = erc7710WalletActions; exports.erc7710BundlerActions = erc7710BundlerActions; exports.actions_exports = actions_exports;
