// @metamask/smart-accounts-kit/dist/chunk-TNGQTVSY.mjs
var __defProp = Object.defineProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};

// src/constants.ts
var Implementation = /* @__PURE__ */ ((Implementation2) => {
  Implementation2["MultiSig"] = "MultiSig";
  Implementation2["Hybrid"] = "Hybrid";
  Implementation2["Stateless7702"] = "Stateless7702";
  return Implementation2;
})(Implementation || {});
var TransferWindow = /* @__PURE__ */ ((TransferWindow2) => {
  TransferWindow2[TransferWindow2["Hourly"] = 3600] = "Hourly";
  TransferWindow2[TransferWindow2["Daily"] = 86400] = "Daily";
  TransferWindow2[TransferWindow2["Weekly"] = 604800] = "Weekly";
  TransferWindow2[TransferWindow2["BiWeekly"] = 1209600] = "BiWeekly";
  TransferWindow2[TransferWindow2["Monthly"] = 2592e3] = "Monthly";
  TransferWindow2[TransferWindow2["Quarterly"] = 7776e3] = "Quarterly";
  TransferWindow2[TransferWindow2["Yearly"] = 31536e3] = "Yearly";
  return TransferWindow2;
})(TransferWindow || {});
var ScopeType = /* @__PURE__ */ ((ScopeType2) => {
  ScopeType2["Erc20TransferAmount"] = "erc20TransferAmount";
  ScopeType2["Erc20Streaming"] = "erc20Streaming";
  ScopeType2["Erc20PeriodTransfer"] = "erc20PeriodTransfer";
  ScopeType2["NativeTokenTransferAmount"] = "nativeTokenTransferAmount";
  ScopeType2["NativeTokenStreaming"] = "nativeTokenStreaming";
  ScopeType2["NativeTokenPeriodTransfer"] = "nativeTokenPeriodTransfer";
  ScopeType2["Erc721Transfer"] = "erc721Transfer";
  ScopeType2["OwnershipTransfer"] = "ownershipTransfer";
  ScopeType2["FunctionCall"] = "functionCall";
  return ScopeType2;
})(ScopeType || {});
var CaveatType = /* @__PURE__ */ ((CaveatType2) => {
  CaveatType2["AllowedMethods"] = "allowedMethods";
  CaveatType2["AllowedTargets"] = "allowedTargets";
  CaveatType2["ApprovalRevocation"] = "approvalRevocation";
  CaveatType2["Deployed"] = "deployed";
  CaveatType2["AllowedCalldata"] = "allowedCalldata";
  CaveatType2["Erc20BalanceChange"] = "erc20BalanceChange";
  CaveatType2["Erc721BalanceChange"] = "erc721BalanceChange";
  CaveatType2["Erc1155BalanceChange"] = "erc1155BalanceChange";
  CaveatType2["ValueLte"] = "valueLte";
  CaveatType2["LimitedCalls"] = "limitedCalls";
  CaveatType2["Id"] = "id";
  CaveatType2["Nonce"] = "nonce";
  CaveatType2["Timestamp"] = "timestamp";
  CaveatType2["BlockNumber"] = "blockNumber";
  CaveatType2["Erc20TransferAmount"] = "erc20TransferAmount";
  CaveatType2["Erc20Streaming"] = "erc20Streaming";
  CaveatType2["NativeTokenStreaming"] = "nativeTokenStreaming";
  CaveatType2["Erc721Transfer"] = "erc721Transfer";
  CaveatType2["NativeTokenTransferAmount"] = "nativeTokenTransferAmount";
  CaveatType2["NativeBalanceChange"] = "nativeBalanceChange";
  CaveatType2["Redeemer"] = "redeemer";
  CaveatType2["NativeTokenPayment"] = "nativeTokenPayment";
  CaveatType2["ArgsEqualityCheck"] = "argsEqualityCheck";
  CaveatType2["SpecificActionERC20TransferBatch"] = "specificActionERC20TransferBatch";
  CaveatType2["Erc20PeriodTransfer"] = "erc20PeriodTransfer";
  CaveatType2["NativeTokenPeriodTransfer"] = "nativeTokenPeriodTransfer";
  CaveatType2["ExactCalldataBatch"] = "exactCalldataBatch";
  CaveatType2["ExactCalldata"] = "exactCalldata";
  CaveatType2["ExactExecution"] = "exactExecution";
  CaveatType2["ExactExecutionBatch"] = "exactExecutionBatch";
  CaveatType2["MultiTokenPeriod"] = "multiTokenPeriod";
  CaveatType2["OwnershipTransfer"] = "ownershipTransfer";
  return CaveatType2;
})(CaveatType || {});

// src/caveatBuilder/caveatBuilder.ts
var INSECURE_UNRESTRICTED_DELEGATION_ERROR_MESSAGE = "No caveats found. If you definitely want to create an empty caveat collection, set `allowInsecureUnrestrictedDelegation` to `true`.";
var CaveatBuilder = class _CaveatBuilder {
  #results = [];
  #hasBeenBuilt = false;
  #environment;
  #config;
  #enforcerBuilders;
  constructor(environment, config = {}, enforcerBuilders = {}, builtCaveats = []) {
    this.#environment = environment;
    this.#config = config;
    this.#enforcerBuilders = enforcerBuilders;
    this.#results = builtCaveats;
  }
  /**
   * Extends the CaveatBuilder with a new enforcer function.
   *
   * @template TEnforcerName - The name of the enforcer.
   * @template TFunction - The type of the enforcer function.
   * @param name - The name of the enforcer.
   * @param fn - The enforcer function.
   * @returns The extended CaveatBuilder instance.
   */
  extend(name, fn) {
    return new _CaveatBuilder(
      this.#environment,
      this.#config,
      { ...this.#enforcerBuilders, [name]: fn },
      this.#results
    );
  }
  addCaveat(nameOrCaveat, config) {
    if (typeof nameOrCaveat === "object") {
      const caveat = {
        args: "0x00",
        ...nameOrCaveat
      };
      this.#results = [...this.#results, caveat];
      return this;
    }
    const name = nameOrCaveat;
    const func = this.#enforcerBuilders[name];
    if (typeof func === "function") {
      const result = func(this.#environment, config);
      this.#results = [...this.#results, result];
      return this;
    }
    throw new Error(`Function "${String(name)}" does not exist.`);
  }
  /**
   * Returns the caveats that have been built using this CaveatBuilder.
   *
   * @returns The array of built caveats.
   * @throws Error if the builder has already been built or if no caveats are found and empty caveats are not allowed.
   */
  build() {
    if (this.#hasBeenBuilt) {
      throw new Error("This CaveatBuilder has already been built.");
    }
    if (this.#results.length === 0 && !this.#config.allowInsecureUnrestrictedDelegation) {
      throw new Error(INSECURE_UNRESTRICTED_DELEGATION_ERROR_MESSAGE);
    }
    this.#hasBeenBuilt = true;
    return this.#results;
  }
};

// src/caveatBuilder/allowedCalldataBuilder.ts
import { createAllowedCalldataTerms } from "@metamask/delegation-core";
var allowedCalldata = "allowedCalldata";
var allowedCalldataBuilder = (environment, config) => {
  const { startIndex, value } = config;
  const terms = createAllowedCalldataTerms({ startIndex, value });
  const {
    caveatEnforcers: { AllowedCalldataEnforcer }
  } = environment;
  if (!AllowedCalldataEnforcer) {
    throw new Error("AllowedCalldataEnforcer not found in environment");
  }
  return {
    enforcer: AllowedCalldataEnforcer,
    terms,
    args: "0x00"
  };
};

// src/caveatBuilder/allowedMethodsBuilder.ts
import { createAllowedMethodsTerms } from "@metamask/delegation-core";
import { isHex, toFunctionSelector } from "viem";
var allowedMethods = "allowedMethods";
var FUNCTION_SELECTOR_STRING_LENGTH = 10;
var allowedMethodsBuilder = (environment, config) => {
  const { selectors } = config;
  if (selectors.length === 0) {
    throw new Error("Invalid selectors: must provide at least one selector");
  }
  const parsedSelectors = selectors.map(parseSelector);
  const terms = createAllowedMethodsTerms({ selectors: parsedSelectors });
  const {
    caveatEnforcers: { AllowedMethodsEnforcer }
  } = environment;
  if (!AllowedMethodsEnforcer) {
    throw new Error("AllowedMethodsEnforcer not found in environment");
  }
  return {
    enforcer: AllowedMethodsEnforcer,
    terms,
    args: "0x00"
  };
};
function parseSelector(selector) {
  if (isHex(selector)) {
    if (selector.length === FUNCTION_SELECTOR_STRING_LENGTH) {
      return selector;
    }
    throw new Error(
      "Invalid selector: must be a 4 byte hex string, abi function signature, or AbiFunction"
    );
  }
  try {
    return toFunctionSelector(selector);
  } catch (rootError) {
    throw new Error(
      "Invalid selector: must be a 4 byte hex string, abi function signature, or AbiFunction",
      { cause: rootError }
    );
  }
}

// src/caveatBuilder/allowedTargetsBuilder.ts
import { createAllowedTargetsTerms } from "@metamask/delegation-core";
import { isAddress } from "viem";
var allowedTargets = "allowedTargets";
var allowedTargetsBuilder = (environment, config) => {
  const { targets } = config;
  if (targets.length === 0) {
    throw new Error(
      "Invalid targets: must provide at least one target address"
    );
  }
  const invalidAddresses = targets.filter(
    (target) => !isAddress(target, { strict: false })
  );
  if (invalidAddresses.length > 0) {
    throw new Error("Invalid targets: must be valid addresses");
  }
  const terms = createAllowedTargetsTerms({ targets });
  const {
    caveatEnforcers: { AllowedTargetsEnforcer }
  } = environment;
  if (!AllowedTargetsEnforcer) {
    throw new Error("AllowedTargetsEnforcer not found in environment");
  }
  return {
    enforcer: AllowedTargetsEnforcer,
    terms,
    args: "0x00"
  };
};

// src/caveatBuilder/approvalRevocationBuilder.ts
import {
  createApprovalRevocationTerms
} from "@metamask/delegation-core";
var approvalRevocation = "approvalRevocation";
var approvalRevocationBuilder = (environment, config) => {
  const terms = createApprovalRevocationTerms(config);
  const {
    caveatEnforcers: { ApprovalRevocationEnforcer }
  } = environment;
  if (!ApprovalRevocationEnforcer) {
    throw new Error("ApprovalRevocationEnforcer not found in environment");
  }
  return {
    enforcer: ApprovalRevocationEnforcer,
    terms,
    args: "0x00"
  };
};

// src/caveatBuilder/argsEqualityCheckBuilder.ts
import { createArgsEqualityCheckTerms } from "@metamask/delegation-core";
import { isHex as isHex2 } from "viem";
var argsEqualityCheck = "argsEqualityCheck";
var argsEqualityCheckBuilder = (environment, config) => {
  const { args } = config;
  if (!isHex2(args)) {
    throw new Error("Invalid config: args must be a valid hex string");
  }
  const {
    caveatEnforcers: { ArgsEqualityCheckEnforcer }
  } = environment;
  if (!ArgsEqualityCheckEnforcer) {
    throw new Error("ArgsEqualityCheckEnforcer not found in environment");
  }
  return {
    enforcer: ArgsEqualityCheckEnforcer,
    terms: createArgsEqualityCheckTerms({ args }),
    args: "0x00"
  };
};

// src/caveatBuilder/blockNumberBuilder.ts
import { createBlockNumberTerms } from "@metamask/delegation-core";
var blockNumber = "blockNumber";
var blockNumberBuilder = (environment, config) => {
  const { afterThreshold, beforeThreshold } = config;
  if (afterThreshold === 0n && beforeThreshold === 0n) {
    throw new Error(
      "Invalid thresholds: At least one of afterThreshold or beforeThreshold must be specified"
    );
  }
  if (beforeThreshold !== 0n && afterThreshold >= beforeThreshold) {
    throw new Error(
      "Invalid thresholds: afterThreshold must be less than beforeThreshold if both are specified"
    );
  }
  const terms = createBlockNumberTerms({ afterThreshold, beforeThreshold });
  const {
    caveatEnforcers: { BlockNumberEnforcer }
  } = environment;
  if (!BlockNumberEnforcer) {
    throw new Error("BlockNumberEnforcer not found in environment");
  }
  return {
    enforcer: BlockNumberEnforcer,
    terms,
    args: "0x00"
  };
};

// src/caveatBuilder/deployedBuilder.ts
import { createDeployedTerms } from "@metamask/delegation-core";
import { isAddress as isAddress2, isHex as isHex3 } from "viem";
var deployed = "deployed";
var deployedBuilder = (environment, config) => {
  const { contractAddress, salt, bytecode } = config;
  if (!isAddress2(contractAddress, { strict: false })) {
    throw new Error(
      `Invalid contractAddress: must be a valid Ethereum address`
    );
  }
  if (!isHex3(salt)) {
    throw new Error("Invalid salt: must be a valid hexadecimal string");
  }
  if (!isHex3(bytecode)) {
    throw new Error("Invalid bytecode: must be a valid hexadecimal string");
  }
  const terms = createDeployedTerms({ contractAddress, salt, bytecode });
  const {
    caveatEnforcers: { DeployedEnforcer }
  } = environment;
  if (!DeployedEnforcer) {
    throw new Error("DeployedEnforcer not found in environment");
  }
  return {
    enforcer: DeployedEnforcer,
    terms,
    args: "0x00"
  };
};

// src/caveatBuilder/erc1155BalanceChangeBuilder.ts
import {
  createERC1155BalanceChangeTerms,
  BalanceChangeType
} from "@metamask/delegation-core";
import { isAddress as isAddress3 } from "viem";
var erc1155BalanceChange = "erc1155BalanceChange";
var erc1155BalanceChangeBuilder = (environment, config) => {
  const { tokenAddress, recipient, tokenId, balance, changeType } = config;
  if (!isAddress3(tokenAddress, { strict: false })) {
    throw new Error("Invalid tokenAddress: must be a valid address");
  }
  if (!isAddress3(recipient, { strict: false })) {
    throw new Error("Invalid recipient: must be a valid address");
  }
  if (balance <= 0n) {
    throw new Error("Invalid balance: must be a positive number");
  }
  if (tokenId < 0n) {
    throw new Error("Invalid tokenId: must be a non-negative number");
  }
  if (changeType !== BalanceChangeType.Increase && changeType !== BalanceChangeType.Decrease) {
    throw new Error("Invalid changeType: must be either Increase or Decrease");
  }
  const terms = createERC1155BalanceChangeTerms({
    tokenAddress,
    recipient,
    tokenId,
    balance,
    changeType
  });
  const {
    caveatEnforcers: { ERC1155BalanceChangeEnforcer }
  } = environment;
  if (!ERC1155BalanceChangeEnforcer) {
    throw new Error("ERC1155BalanceChangeEnforcer not found in environment");
  }
  return {
    enforcer: ERC1155BalanceChangeEnforcer,
    terms,
    args: "0x00"
  };
};

// src/caveatBuilder/erc20BalanceChangeBuilder.ts
import {
  createERC20BalanceChangeTerms,
  BalanceChangeType as BalanceChangeType2
} from "@metamask/delegation-core";
import { isAddress as isAddress4 } from "viem";
var erc20BalanceChange = "erc20BalanceChange";
var erc20BalanceChangeBuilder = (environment, config) => {
  const { tokenAddress, recipient, balance, changeType } = config;
  if (!isAddress4(tokenAddress, { strict: false })) {
    throw new Error("Invalid tokenAddress: must be a valid address");
  }
  if (balance <= 0n) {
    throw new Error("Invalid balance: must be a positive number");
  }
  if (changeType !== BalanceChangeType2.Increase && changeType !== BalanceChangeType2.Decrease) {
    throw new Error("Invalid changeType: must be either Increase or Decrease");
  }
  const terms = createERC20BalanceChangeTerms({
    tokenAddress,
    recipient,
    balance,
    changeType
  });
  const {
    caveatEnforcers: { ERC20BalanceChangeEnforcer }
  } = environment;
  if (!ERC20BalanceChangeEnforcer) {
    throw new Error("ERC20BalanceChangeEnforcer not found in environment");
  }
  return {
    enforcer: ERC20BalanceChangeEnforcer,
    terms,
    args: "0x00"
  };
};

// src/caveatBuilder/erc20PeriodTransferBuilder.ts
import { createERC20TokenPeriodTransferTerms } from "@metamask/delegation-core";
var erc20PeriodTransfer = "erc20PeriodTransfer";
var erc20PeriodTransferBuilder = (environment, config) => {
  const { tokenAddress, periodAmount, periodDuration, startDate } = config;
  const terms = createERC20TokenPeriodTransferTerms({
    tokenAddress,
    periodAmount,
    periodDuration,
    startDate
  });
  const {
    caveatEnforcers: { ERC20PeriodTransferEnforcer }
  } = environment;
  if (!ERC20PeriodTransferEnforcer) {
    throw new Error("ERC20PeriodTransferEnforcer not found in environment");
  }
  return {
    enforcer: ERC20PeriodTransferEnforcer,
    terms,
    args: "0x00"
  };
};

// src/caveatBuilder/erc20StreamingBuilder.ts
import { createERC20StreamingTerms } from "@metamask/delegation-core";
var erc20Streaming = "erc20Streaming";
var erc20StreamingBuilder = (environment, config) => {
  const { tokenAddress, initialAmount, maxAmount, amountPerSecond, startTime } = config;
  const terms = createERC20StreamingTerms({
    tokenAddress,
    initialAmount,
    maxAmount,
    amountPerSecond,
    startTime
  });
  const {
    caveatEnforcers: { ERC20StreamingEnforcer }
  } = environment;
  if (!ERC20StreamingEnforcer) {
    throw new Error("ERC20StreamingEnforcer not found in environment");
  }
  return {
    enforcer: ERC20StreamingEnforcer,
    terms,
    args: "0x00"
  };
};

// src/caveatBuilder/erc20TransferAmountBuilder.ts
import { createERC20TransferAmountTerms } from "@metamask/delegation-core";
import { isAddress as isAddress5 } from "viem";
var erc20TransferAmount = "erc20TransferAmount";
var erc20TransferAmountBuilder = (environment, config) => {
  const { tokenAddress, maxAmount } = config;
  if (!isAddress5(tokenAddress, { strict: false })) {
    throw new Error("Invalid tokenAddress: must be a valid address");
  }
  if (maxAmount <= 0n) {
    throw new Error("Invalid maxAmount: must be a positive number");
  }
  const terms = createERC20TransferAmountTerms({ tokenAddress, maxAmount });
  const {
    caveatEnforcers: { ERC20TransferAmountEnforcer }
  } = environment;
  if (!ERC20TransferAmountEnforcer) {
    throw new Error("ERC20TransferAmountEnforcer not found in environment");
  }
  return {
    enforcer: ERC20TransferAmountEnforcer,
    terms,
    args: "0x00"
  };
};

// src/caveatBuilder/erc721BalanceChangeBuilder.ts
import {
  createERC721BalanceChangeTerms,
  BalanceChangeType as BalanceChangeType3
} from "@metamask/delegation-core";
import { isAddress as isAddress6 } from "viem";
var erc721BalanceChange = "erc721BalanceChange";
var erc721BalanceChangeBuilder = (environment, config) => {
  const { tokenAddress, recipient, amount, changeType } = config;
  if (!isAddress6(tokenAddress, { strict: false })) {
    throw new Error("Invalid tokenAddress: must be a valid address");
  }
  if (!isAddress6(recipient, { strict: false })) {
    throw new Error("Invalid recipient: must be a valid address");
  }
  if (amount <= 0n) {
    throw new Error("Invalid balance: must be a positive number");
  }
  if (changeType !== BalanceChangeType3.Increase && changeType !== BalanceChangeType3.Decrease) {
    throw new Error("Invalid changeType: must be either Increase or Decrease");
  }
  const terms = createERC721BalanceChangeTerms({
    tokenAddress,
    recipient,
    amount,
    changeType
  });
  const {
    caveatEnforcers: { ERC721BalanceChangeEnforcer }
  } = environment;
  if (!ERC721BalanceChangeEnforcer) {
    throw new Error("ERC721BalanceChangeEnforcer not found in environment");
  }
  return {
    enforcer: ERC721BalanceChangeEnforcer,
    terms,
    args: "0x00"
  };
};

// src/caveatBuilder/erc721TransferBuilder.ts
import { createERC721TransferTerms } from "@metamask/delegation-core";
import { isAddress as isAddress7 } from "viem";
var erc721Transfer = "erc721Transfer";
var erc721TransferBuilder = (environment, config) => {
  const { tokenAddress, tokenId } = config;
  if (!isAddress7(tokenAddress, { strict: false })) {
    throw new Error("Invalid tokenAddress: must be a valid address");
  }
  if (tokenId < 0n) {
    throw new Error("Invalid tokenId: must be a non-negative number");
  }
  const terms = createERC721TransferTerms({ tokenAddress, tokenId });
  const {
    caveatEnforcers: { ERC721TransferEnforcer }
  } = environment;
  if (!ERC721TransferEnforcer) {
    throw new Error("ERC721TransferEnforcer not found in environment");
  }
  return {
    enforcer: ERC721TransferEnforcer,
    terms,
    args: "0x00"
  };
};

// src/caveatBuilder/exactCalldataBatchBuilder.ts
import { createExactCalldataBatchTerms } from "@metamask/delegation-core";
import { isAddress as isAddress8 } from "viem";
var exactCalldataBatch = "exactCalldataBatch";
var exactCalldataBatchBuilder = (environment, config) => {
  const { executions } = config;
  if (executions.length === 0) {
    throw new Error("Invalid executions: array cannot be empty");
  }
  for (const execution of executions) {
    if (!isAddress8(execution.target, { strict: false })) {
      throw new Error("Invalid target: must be a valid address");
    }
    if (execution.value < 0n) {
      throw new Error("Invalid value: must be a non-negative number");
    }
    if (!execution.callData.startsWith("0x")) {
      throw new Error(
        "Invalid calldata: must be a hex string starting with 0x"
      );
    }
  }
  const terms = createExactCalldataBatchTerms({ executions });
  const {
    caveatEnforcers: { ExactCalldataBatchEnforcer }
  } = environment;
  if (!ExactCalldataBatchEnforcer) {
    throw new Error("ExactCalldataBatchEnforcer not found in environment");
  }
  return {
    enforcer: ExactCalldataBatchEnforcer,
    terms,
    args: "0x00"
  };
};

// src/caveatBuilder/exactCalldataBuilder.ts
import { createExactCalldataTerms } from "@metamask/delegation-core";
var exactCalldata = "exactCalldata";
var exactCalldataBuilder = (environment, config) => {
  const { calldata } = config;
  const terms = createExactCalldataTerms({ calldata });
  const {
    caveatEnforcers: { ExactCalldataEnforcer }
  } = environment;
  if (!ExactCalldataEnforcer) {
    throw new Error("ExactCalldataEnforcer not found in environment");
  }
  return {
    enforcer: ExactCalldataEnforcer,
    terms,
    args: "0x00"
  };
};

// src/caveatBuilder/exactExecutionBatchBuilder.ts
import { createExactExecutionBatchTerms } from "@metamask/delegation-core";
import { isAddress as isAddress9 } from "viem";
var exactExecutionBatch = "exactExecutionBatch";
var exactExecutionBatchBuilder = (environment, config) => {
  const { executions } = config;
  if (executions.length === 0) {
    throw new Error("Invalid executions: array cannot be empty");
  }
  for (const execution of executions) {
    if (!isAddress9(execution.target, { strict: false })) {
      throw new Error("Invalid target: must be a valid address");
    }
    if (execution.value < 0n) {
      throw new Error("Invalid value: must be a non-negative number");
    }
    if (!execution.callData.startsWith("0x")) {
      throw new Error(
        "Invalid calldata: must be a hex string starting with 0x"
      );
    }
  }
  const terms = createExactExecutionBatchTerms({ executions });
  const {
    caveatEnforcers: { ExactExecutionBatchEnforcer }
  } = environment;
  if (!ExactExecutionBatchEnforcer) {
    throw new Error("ExactExecutionBatchEnforcer not found in environment");
  }
  return {
    enforcer: ExactExecutionBatchEnforcer,
    terms,
    args: "0x00"
  };
};

// src/caveatBuilder/exactExecutionBuilder.ts
import { createExactExecutionTerms } from "@metamask/delegation-core";
import { isAddress as isAddress10 } from "viem";
var exactExecution = "exactExecution";
var exactExecutionBuilder = (environment, config) => {
  const { execution } = config;
  if (!isAddress10(execution.target, { strict: false })) {
    throw new Error("Invalid target: must be a valid address");
  }
  if (execution.value < 0n) {
    throw new Error("Invalid value: must be a non-negative number");
  }
  if (!execution.callData.startsWith("0x")) {
    throw new Error("Invalid calldata: must be a hex string starting with 0x");
  }
  const terms = createExactExecutionTerms({ execution });
  const {
    caveatEnforcers: { ExactExecutionEnforcer }
  } = environment;
  if (!ExactExecutionEnforcer) {
    throw new Error("ExactExecutionEnforcer not found in environment");
  }
  return {
    enforcer: ExactExecutionEnforcer,
    terms,
    args: "0x00"
  };
};

// src/caveatBuilder/idBuilder.ts
import { createIdTerms } from "@metamask/delegation-core";
import { maxUint256 } from "viem";
var id = "id";
var idBuilder = (environment, config) => {
  const { id: idValue } = config;
  let idBigInt;
  if (typeof idValue === "number") {
    if (!Number.isInteger(idValue)) {
      throw new Error("Invalid id: must be an integer");
    }
    idBigInt = BigInt(idValue);
  } else if (typeof idValue === "bigint") {
    idBigInt = idValue;
  } else {
    throw new Error("Invalid id: must be a bigint or number");
  }
  if (idBigInt < 0n) {
    throw new Error("Invalid id: must be a non-negative number");
  }
  if (idBigInt > maxUint256) {
    throw new Error("Invalid id: must be less than 2^256");
  }
  const terms = createIdTerms({ id: idBigInt });
  const {
    caveatEnforcers: { IdEnforcer }
  } = environment;
  if (!IdEnforcer) {
    throw new Error("IdEnforcer not found in environment");
  }
  return {
    enforcer: IdEnforcer,
    terms,
    args: "0x00"
  };
};

// src/caveatBuilder/limitedCallsBuilder.ts
import { createLimitedCallsTerms } from "@metamask/delegation-core";
var limitedCalls = "limitedCalls";
var limitedCallsBuilder = (environment, config) => {
  const { limit } = config;
  if (!Number.isInteger(limit)) {
    throw new Error("Invalid limit: must be an integer");
  }
  if (limit <= 0) {
    throw new Error("Invalid limit: must be a positive integer");
  }
  const terms = createLimitedCallsTerms({ limit });
  const {
    caveatEnforcers: { LimitedCallsEnforcer }
  } = environment;
  if (!LimitedCallsEnforcer) {
    throw new Error("LimitedCallsEnforcer not found in environment");
  }
  return {
    enforcer: LimitedCallsEnforcer,
    terms,
    args: "0x00"
  };
};

// src/caveatBuilder/multiTokenPeriodBuilder.ts
import { createMultiTokenPeriodTerms } from "@metamask/delegation-core";
import { isAddress as isAddress11 } from "viem";
var multiTokenPeriod = "multiTokenPeriod";
var multiTokenPeriodBuilder = (environment, config) => {
  if (!config?.tokenConfigs || config.tokenConfigs.length === 0) {
    throw new Error(
      "MultiTokenPeriodBuilder: tokenConfigs array cannot be empty"
    );
  }
  config.tokenConfigs.forEach((tokenConfig) => {
    if (!isAddress11(tokenConfig.token)) {
      throw new Error(`Invalid token address: ${String(tokenConfig.token)}`);
    }
    if (tokenConfig.periodAmount <= 0) {
      throw new Error("Invalid period amount: must be greater than 0");
    }
    if (tokenConfig.periodDuration <= 0) {
      throw new Error("Invalid period duration: must be greater than 0");
    }
  });
  const terms = createMultiTokenPeriodTerms({
    tokenConfigs: config.tokenConfigs
  });
  const {
    caveatEnforcers: { MultiTokenPeriodEnforcer }
  } = environment;
  if (!MultiTokenPeriodEnforcer) {
    throw new Error("MultiTokenPeriodEnforcer not found in environment");
  }
  return {
    enforcer: MultiTokenPeriodEnforcer,
    terms,
    args: "0x00"
  };
};

// src/caveatBuilder/nativeBalanceChangeBuilder.ts
import {
  createNativeBalanceChangeTerms,
  BalanceChangeType as BalanceChangeType4
} from "@metamask/delegation-core";
import { isAddress as isAddress12 } from "viem";
var nativeBalanceChange = "nativeBalanceChange";
var nativeBalanceChangeBuilder = (environment, config) => {
  const { recipient, balance, changeType } = config;
  if (!isAddress12(recipient)) {
    throw new Error("Invalid recipient: must be a valid Address");
  }
  if (balance <= 0n) {
    throw new Error("Invalid balance: must be a positive number");
  }
  if (changeType !== BalanceChangeType4.Increase && changeType !== BalanceChangeType4.Decrease) {
    throw new Error("Invalid changeType: must be either Increase or Decrease");
  }
  const terms = createNativeBalanceChangeTerms({
    recipient,
    balance,
    changeType
  });
  const {
    caveatEnforcers: { NativeBalanceChangeEnforcer }
  } = environment;
  if (!NativeBalanceChangeEnforcer) {
    throw new Error("NativeBalanceChangeEnforcer not found in environment");
  }
  return {
    enforcer: NativeBalanceChangeEnforcer,
    terms,
    args: "0x00"
  };
};

// src/caveatBuilder/nativeTokenPaymentBuilder.ts
import { createNativeTokenPaymentTerms } from "@metamask/delegation-core";
import { isAddress as isAddress13 } from "viem";
var nativeTokenPayment = "nativeTokenPayment";
var nativeTokenPaymentBuilder = (environment, config) => {
  const { recipient, amount } = config;
  if (amount <= 0n) {
    throw new Error("Invalid amount: must be positive");
  }
  if (!isAddress13(recipient)) {
    throw new Error("Invalid recipient: must be a valid address");
  }
  const terms = createNativeTokenPaymentTerms({ recipient, amount });
  const {
    caveatEnforcers: { NativeTokenPaymentEnforcer }
  } = environment;
  if (!NativeTokenPaymentEnforcer) {
    throw new Error("NativeTokenPaymentEnforcer not found in environment");
  }
  return {
    enforcer: NativeTokenPaymentEnforcer,
    terms,
    args: "0x00"
  };
};

// src/caveatBuilder/nativeTokenPeriodTransferBuilder.ts
import { createNativeTokenPeriodTransferTerms } from "@metamask/delegation-core";
var nativeTokenPeriodTransfer = "nativeTokenPeriodTransfer";
var nativeTokenPeriodTransferBuilder = (environment, config) => {
  const { periodAmount, periodDuration, startDate } = config;
  const terms = createNativeTokenPeriodTransferTerms({
    periodAmount,
    periodDuration,
    startDate
  });
  const {
    caveatEnforcers: { NativeTokenPeriodTransferEnforcer }
  } = environment;
  if (!NativeTokenPeriodTransferEnforcer) {
    throw new Error(
      "NativeTokenPeriodTransferEnforcer not found in environment"
    );
  }
  return {
    enforcer: NativeTokenPeriodTransferEnforcer,
    terms,
    args: "0x00"
  };
};

// src/caveatBuilder/nativeTokenStreamingBuilder.ts
import { createNativeTokenStreamingTerms } from "@metamask/delegation-core";
var nativeTokenStreaming = "nativeTokenStreaming";
var nativeTokenStreamingBuilder = (environment, config) => {
  const { initialAmount, maxAmount, amountPerSecond, startTime } = config;
  const terms = createNativeTokenStreamingTerms({
    initialAmount,
    maxAmount,
    amountPerSecond,
    startTime
  });
  const {
    caveatEnforcers: { NativeTokenStreamingEnforcer }
  } = environment;
  if (!NativeTokenStreamingEnforcer) {
    throw new Error("NativeTokenStreamingEnforcer not found in environment");
  }
  return {
    enforcer: NativeTokenStreamingEnforcer,
    terms,
    args: "0x00"
  };
};

// src/caveatBuilder/nativeTokenTransferAmountBuilder.ts
import { createNativeTokenTransferAmountTerms } from "@metamask/delegation-core";
var nativeTokenTransferAmount = "nativeTokenTransferAmount";
var nativeTokenTransferAmountBuilder = (environment, config) => {
  const { maxAmount } = config;
  if (maxAmount < 0n) {
    throw new Error("Invalid maxAmount: must be zero or positive");
  }
  const terms = createNativeTokenTransferAmountTerms({ maxAmount });
  const {
    caveatEnforcers: { NativeTokenTransferAmountEnforcer }
  } = environment;
  if (!NativeTokenTransferAmountEnforcer) {
    throw new Error(
      "NativeTokenTransferAmountEnforcer not found in environment"
    );
  }
  return {
    enforcer: NativeTokenTransferAmountEnforcer,
    terms,
    args: "0x00"
  };
};

// src/caveatBuilder/nonceBuilder.ts
import { createNonceTerms } from "@metamask/delegation-core";
var nonce = "nonce";
var nonceBuilder = (environment, config) => {
  const { nonce: nonceValue } = config;
  const terms = createNonceTerms({ nonce: nonceValue });
  const {
    caveatEnforcers: { NonceEnforcer }
  } = environment;
  if (!NonceEnforcer) {
    throw new Error("NonceEnforcer not found in environment");
  }
  return {
    enforcer: NonceEnforcer,
    terms,
    args: "0x00"
  };
};

// src/caveatBuilder/ownershipTransferBuilder.ts
import { createOwnershipTransferTerms } from "@metamask/delegation-core";
import { isAddress as isAddress14 } from "viem";
var ownershipTransfer = "ownershipTransfer";
var ownershipTransferBuilder = (environment, config) => {
  const { contractAddress } = config;
  if (!isAddress14(contractAddress, { strict: false })) {
    throw new Error("Invalid contractAddress: must be a valid address");
  }
  const terms = createOwnershipTransferTerms({ contractAddress });
  const {
    caveatEnforcers: { OwnershipTransferEnforcer }
  } = environment;
  if (!OwnershipTransferEnforcer) {
    throw new Error("OwnershipTransferEnforcer not found in environment");
  }
  return {
    enforcer: OwnershipTransferEnforcer,
    terms,
    args: "0x00"
  };
};

// src/caveatBuilder/redeemerBuilder.ts
import { createRedeemerTerms } from "@metamask/delegation-core";
import { isAddress as isAddress15 } from "viem";
var redeemer = "redeemer";
var redeemerBuilder = (environment, config) => {
  const { redeemers } = config;
  if (redeemers.length === 0) {
    throw new Error(
      "Invalid redeemers: must specify at least one redeemer address"
    );
  }
  for (const redeemerAddress of redeemers) {
    if (!isAddress15(redeemerAddress)) {
      throw new Error("Invalid redeemers: must be a valid address");
    }
  }
  const terms = createRedeemerTerms({ redeemers });
  const {
    caveatEnforcers: { RedeemerEnforcer }
  } = environment;
  if (!RedeemerEnforcer) {
    throw new Error("RedeemerEnforcer not found in environment");
  }
  return {
    enforcer: RedeemerEnforcer,
    terms,
    args: "0x00"
  };
};

// src/caveatBuilder/specificActionERC20TransferBatchBuilder.ts
import { createSpecificActionERC20TransferBatchTerms } from "@metamask/delegation-core";
import { isAddress as isAddress16 } from "viem";
var specificActionERC20TransferBatch = "specificActionERC20TransferBatch";
var specificActionERC20TransferBatchBuilder = (environment, config) => {
  const { tokenAddress, recipient, amount, target, calldata } = config;
  if (!isAddress16(tokenAddress, { strict: false })) {
    throw new Error("Invalid tokenAddress: must be a valid address");
  }
  if (!isAddress16(recipient, { strict: false })) {
    throw new Error("Invalid recipient: must be a valid address");
  }
  if (!isAddress16(target, { strict: false })) {
    throw new Error("Invalid target: must be a valid address");
  }
  if (amount <= 0n) {
    throw new Error("Invalid amount: must be a positive number");
  }
  const terms = createSpecificActionERC20TransferBatchTerms({
    tokenAddress,
    recipient,
    amount,
    target,
    calldata
  });
  const {
    caveatEnforcers: { SpecificActionERC20TransferBatchEnforcer }
  } = environment;
  if (!SpecificActionERC20TransferBatchEnforcer) {
    throw new Error(
      "SpecificActionERC20TransferBatchEnforcer not found in environment"
    );
  }
  return {
    enforcer: SpecificActionERC20TransferBatchEnforcer,
    terms,
    args: "0x00"
  };
};

// src/caveatBuilder/timestampBuilder.ts
import { createTimestampTerms } from "@metamask/delegation-core";
var timestamp = "timestamp";
var timestampBuilder = (environment, config) => {
  const { afterThreshold, beforeThreshold } = config;
  const terms = createTimestampTerms({
    afterThreshold,
    beforeThreshold
  });
  const {
    caveatEnforcers: { TimestampEnforcer }
  } = environment;
  if (!TimestampEnforcer) {
    throw new Error("TimestampEnforcer not found in environment");
  }
  return {
    enforcer: TimestampEnforcer,
    terms,
    args: "0x00"
  };
};

// src/caveatBuilder/valueLteBuilder.ts
import { createValueLteTerms } from "@metamask/delegation-core";
var valueLte = "valueLte";
var valueLteBuilder = (environment, config) => {
  const { maxValue } = config;
  const terms = createValueLteTerms({ maxValue });
  const {
    caveatEnforcers: { ValueLteEnforcer }
  } = environment;
  if (!ValueLteEnforcer) {
    throw new Error("ValueLteEnforcer not found in environment");
  }
  return {
    enforcer: ValueLteEnforcer,
    terms,
    args: "0x00"
  };
};

// src/caveatBuilder/coreCaveatBuilder.ts
var createCaveatBuilder = (environment, config) => {
  const caveatBuilder = new CaveatBuilder(environment, config).extend(allowedMethods, allowedMethodsBuilder).extend(allowedTargets, allowedTargetsBuilder).extend(approvalRevocation, approvalRevocationBuilder).extend(deployed, deployedBuilder).extend(allowedCalldata, allowedCalldataBuilder).extend(erc20BalanceChange, erc20BalanceChangeBuilder).extend(erc721BalanceChange, erc721BalanceChangeBuilder).extend(erc1155BalanceChange, erc1155BalanceChangeBuilder).extend(valueLte, valueLteBuilder).extend(limitedCalls, limitedCallsBuilder).extend(id, idBuilder).extend(nonce, nonceBuilder).extend(timestamp, timestampBuilder).extend(blockNumber, blockNumberBuilder).extend(erc20TransferAmount, erc20TransferAmountBuilder).extend(erc20Streaming, erc20StreamingBuilder).extend(nativeTokenStreaming, nativeTokenStreamingBuilder).extend(erc721Transfer, erc721TransferBuilder).extend(nativeTokenTransferAmount, nativeTokenTransferAmountBuilder).extend(nativeBalanceChange, nativeBalanceChangeBuilder).extend(redeemer, redeemerBuilder).extend(nativeTokenPayment, nativeTokenPaymentBuilder).extend(argsEqualityCheck, argsEqualityCheckBuilder).extend(
    specificActionERC20TransferBatch,
    specificActionERC20TransferBatchBuilder
  ).extend(erc20PeriodTransfer, erc20PeriodTransferBuilder).extend(nativeTokenPeriodTransfer, nativeTokenPeriodTransferBuilder).extend(exactCalldataBatch, exactCalldataBatchBuilder).extend(exactCalldata, exactCalldataBuilder).extend(exactExecution, exactExecutionBuilder).extend(exactExecutionBatch, exactExecutionBatchBuilder).extend(multiTokenPeriod, multiTokenPeriodBuilder).extend(ownershipTransfer, ownershipTransferBuilder);
  return caveatBuilder;
};

// src/caveatBuilder/scope/erc20PeriodicScope.ts
function createErc20PeriodicCaveatBuilder(environment, config) {
  return createCaveatBuilder(environment).addCaveat("valueLte", {
    maxValue: 0n
  }).addCaveat("erc20PeriodTransfer", {
    tokenAddress: config.tokenAddress,
    periodAmount: config.periodAmount,
    periodDuration: config.periodDuration,
    startDate: config.startDate
  });
}

// src/caveatBuilder/scope/erc20StreamingScope.ts
function createErc20StreamingCaveatBuilder(environment, config) {
  return createCaveatBuilder(environment).addCaveat("valueLte", {
    maxValue: 0n
  }).addCaveat("erc20Streaming", {
    tokenAddress: config.tokenAddress,
    initialAmount: config.initialAmount,
    maxAmount: config.maxAmount,
    amountPerSecond: config.amountPerSecond,
    startTime: config.startTime
  });
}

// src/caveatBuilder/scope/erc20TransferScope.ts
function createErc20TransferCaveatBuilder(environment, config) {
  return createCaveatBuilder(environment).addCaveat("valueLte", {
    maxValue: 0n
  }).addCaveat("erc20TransferAmount", {
    tokenAddress: config.tokenAddress,
    maxAmount: config.maxAmount
  });
}

// src/utils.ts
import { isHex as isHex4, toHex } from "viem";
var hasProperties = (object, properties) => {
  return properties.every(
    (prop) => prop in object && object[prop] !== void 0
  );
};
function isDefined(value) {
  return value !== void 0 && value !== null;
}
function assertIsDefined(value, parameterName) {
  if (!isDefined(value)) {
    throw new Error(
      `Invalid parameters: ${parameterName ?? "value"} is required`
    );
  }
}
function toHexOrThrow(value, parameterName) {
  assertIsDefined(value, parameterName);
  if (typeof value === "string") {
    if (!isHex4(value)) {
      throw new Error(
        `Invalid parameters: ${parameterName ?? "value"} is not a valid hex value`
      );
    }
    return value;
  }
  return toHex(value);
}

// src/caveatBuilder/scope/erc721Scope.ts
var isErc721TransferConfig = (config) => {
  return hasProperties(
    config,
    ["tokenAddress", "tokenId"]
  );
};
function createErc721CaveatBuilder(environment, config) {
  if (!isErc721TransferConfig(config)) {
    throw new Error("Invalid ERC721 configuration");
  }
  const caveatBuilder = createCaveatBuilder(environment).addCaveat(
    "erc721Transfer",
    config
  );
  return caveatBuilder;
}

// src/caveatBuilder/scope/functionCallScope.ts
var isFunctionCallConfig = (config) => {
  return hasProperties(config, ["targets", "selectors"]);
};
function createFunctionCallCaveatBuilder(environment, config) {
  const { targets, selectors, allowedCalldata: allowedCalldata2, exactCalldata: exactCalldata2 } = config;
  if (!isFunctionCallConfig(config)) {
    throw new Error("Invalid Function Call configuration");
  }
  if (allowedCalldata2 && allowedCalldata2.length > 0 && exactCalldata2) {
    throw new Error(
      "Cannot specify both allowedCalldata and exactCalldata. Please use only one calldata restriction type."
    );
  }
  const valueLteConfig = config.valueLte ?? { maxValue: 0n };
  const caveatBuilder = createCaveatBuilder(environment).addCaveat("allowedTargets", { targets }).addCaveat("allowedMethods", { selectors }).addCaveat("valueLte", valueLteConfig);
  if (allowedCalldata2 && allowedCalldata2.length > 0) {
    allowedCalldata2.forEach((calldataConfig) => {
      caveatBuilder.addCaveat("allowedCalldata", calldataConfig);
    });
  } else if (exactCalldata2) {
    caveatBuilder.addCaveat("exactCalldata", exactCalldata2);
  }
  return caveatBuilder;
}

// src/caveatBuilder/scope/nativeTokenPeriodicScope.ts
function createNativeTokenPeriodicCaveatBuilder(environment, config) {
  const {
    periodAmount,
    periodDuration,
    startDate,
    allowedCalldata: allowedCalldata2,
    exactCalldata: exactCalldata2
  } = config;
  if (allowedCalldata2 && allowedCalldata2.length > 0 && exactCalldata2) {
    throw new Error(
      "Cannot specify both allowedCalldata and exactCalldata. Please use only one calldata restriction type."
    );
  }
  const caveatBuilder = createCaveatBuilder(environment);
  if (allowedCalldata2 && allowedCalldata2.length > 0) {
    allowedCalldata2.forEach((calldataConfig) => {
      caveatBuilder.addCaveat("allowedCalldata", calldataConfig);
    });
  } else if (exactCalldata2) {
    caveatBuilder.addCaveat("exactCalldata", exactCalldata2);
  } else {
    caveatBuilder.addCaveat("exactCalldata", {
      calldata: "0x"
    });
  }
  caveatBuilder.addCaveat("nativeTokenPeriodTransfer", {
    periodAmount,
    periodDuration,
    startDate
  });
  return caveatBuilder;
}

// src/caveatBuilder/scope/nativeTokenStreamingScope.ts
function createNativeTokenStreamingCaveatBuilder(environment, config) {
  const {
    initialAmount,
    maxAmount,
    amountPerSecond,
    startTime,
    allowedCalldata: allowedCalldata2,
    exactCalldata: exactCalldata2
  } = config;
  if (allowedCalldata2 && allowedCalldata2.length > 0 && exactCalldata2) {
    throw new Error(
      "Cannot specify both allowedCalldata and exactCalldata. Please use only one calldata restriction type."
    );
  }
  const caveatBuilder = createCaveatBuilder(environment);
  if (allowedCalldata2 && allowedCalldata2.length > 0) {
    allowedCalldata2.forEach((calldataConfig) => {
      caveatBuilder.addCaveat("allowedCalldata", calldataConfig);
    });
  } else if (exactCalldata2) {
    caveatBuilder.addCaveat("exactCalldata", exactCalldata2);
  } else {
    caveatBuilder.addCaveat("exactCalldata", {
      calldata: "0x"
    });
  }
  caveatBuilder.addCaveat("nativeTokenStreaming", {
    initialAmount,
    maxAmount,
    amountPerSecond,
    startTime
  });
  return caveatBuilder;
}

// src/caveatBuilder/scope/nativeTokenTransferScope.ts
function createNativeTokenTransferCaveatBuilder(environment, config) {
  const { maxAmount, allowedCalldata: allowedCalldata2, exactCalldata: exactCalldata2 } = config;
  if (allowedCalldata2 && allowedCalldata2.length > 0 && exactCalldata2) {
    throw new Error(
      "Cannot specify both allowedCalldata and exactCalldata. Please use only one calldata restriction type."
    );
  }
  const caveatBuilder = createCaveatBuilder(environment);
  if (allowedCalldata2 && allowedCalldata2.length > 0) {
    allowedCalldata2.forEach((calldataConfig) => {
      caveatBuilder.addCaveat("allowedCalldata", calldataConfig);
    });
  } else if (exactCalldata2) {
    caveatBuilder.addCaveat("exactCalldata", exactCalldata2);
  } else {
    caveatBuilder.addCaveat("exactCalldata", {
      calldata: "0x"
    });
  }
  caveatBuilder.addCaveat("nativeTokenTransferAmount", {
    maxAmount
  });
  return caveatBuilder;
}

// src/caveatBuilder/scope/ownershipScope.ts
var isOwnershipTransferConfig = (config) => {
  return hasProperties(
    config,
    ["contractAddress"]
  );
};
function createOwnershipCaveatBuilder(environment, config) {
  if (!isOwnershipTransferConfig(config)) {
    throw new Error("Invalid ownership transfer configuration");
  }
  const caveatBuilder = createCaveatBuilder(environment).addCaveat(
    "ownershipTransfer",
    config
  );
  return caveatBuilder;
}

// src/caveatBuilder/scope/index.ts
var normalizeScopeConfig = (config) => {
  return {
    ...config,
    type: config.type
  };
};
var createCaveatBuilderFromScope = (environment, scopeConfig) => {
  const normalizedScopeConfig = normalizeScopeConfig(scopeConfig);
  switch (normalizedScopeConfig.type) {
    case "erc20TransferAmount" /* Erc20TransferAmount */:
      return createErc20TransferCaveatBuilder(
        environment,
        normalizedScopeConfig
      );
    case "erc20Streaming" /* Erc20Streaming */:
      return createErc20StreamingCaveatBuilder(
        environment,
        normalizedScopeConfig
      );
    case "erc20PeriodTransfer" /* Erc20PeriodTransfer */:
      return createErc20PeriodicCaveatBuilder(
        environment,
        normalizedScopeConfig
      );
    case "nativeTokenTransferAmount" /* NativeTokenTransferAmount */:
      return createNativeTokenTransferCaveatBuilder(
        environment,
        normalizedScopeConfig
      );
    case "nativeTokenStreaming" /* NativeTokenStreaming */:
      return createNativeTokenStreamingCaveatBuilder(
        environment,
        normalizedScopeConfig
      );
    case "nativeTokenPeriodTransfer" /* NativeTokenPeriodTransfer */:
      return createNativeTokenPeriodicCaveatBuilder(
        environment,
        normalizedScopeConfig
      );
    case "erc721Transfer" /* Erc721Transfer */:
      return createErc721CaveatBuilder(environment, normalizedScopeConfig);
    case "ownershipTransfer" /* OwnershipTransfer */:
      return createOwnershipCaveatBuilder(environment, normalizedScopeConfig);
    case "functionCall" /* FunctionCall */:
      return createFunctionCallCaveatBuilder(
        environment,
        normalizedScopeConfig
      );
    default:
      const exhaustivenessCheck = normalizedScopeConfig;
      throw new Error(
        `Invalid scope type: ${exhaustivenessCheck.type}`
      );
  }
};

// src/caveatBuilder/resolveCaveats.ts
var resolveCaveats = ({
  environment,
  scope,
  caveats,
  isScopeOptional
}) => {
  if (!scope && !isScopeOptional) {
    throw new Error("Scope is required");
  }
  const scopeCaveatBuilder = scope ? createCaveatBuilderFromScope(environment, scope) : createCaveatBuilder(environment, {
    allowInsecureUnrestrictedDelegation: true
  });
  if (caveats) {
    if ("build" in caveats && typeof caveats.build === "function") {
      caveats.build().forEach((caveat) => {
        scopeCaveatBuilder.addCaveat(caveat);
      });
    } else if (Array.isArray(caveats)) {
      caveats.forEach((caveat) => {
        try {
          if ("type" in caveat) {
            const { type, ...config } = caveat;
            scopeCaveatBuilder.addCaveat(type, config);
          } else {
            scopeCaveatBuilder.addCaveat(caveat);
          }
        } catch (error) {
          throw new Error(`Invalid caveat: ${error.message}`);
        }
      });
    }
  }
  return scopeCaveatBuilder.build();
};

// src/caveats.ts
import {
  decodeAllowedCalldataTerms,
  decodeERC20StreamingTerms,
  decodeERC20TransferAmountTerms,
  decodeERC20BalanceChangeTerms,
  decodeApprovalRevocationTerms,
  decodeAllowedMethodsTerms,
  decodeAllowedTargetsTerms,
  decodeArgsEqualityCheckTerms,
  decodeBlockNumberTerms,
  decodeDeployedTerms,
  decodeERC721BalanceChangeTerms,
  decodeERC721TransferTerms,
  decodeERC1155BalanceChangeTerms,
  decodeTimestampTerms,
  decodeNonceTerms,
  decodeValueLteTerms,
  decodeLimitedCallsTerms,
  decodeIdTerms,
  decodeNativeTokenTransferAmountTerms,
  decodeNativeBalanceChangeTerms,
  decodeNativeTokenStreamingTerms,
  decodeNativeTokenPaymentTerms,
  decodeRedeemerTerms,
  decodeSpecificActionERC20TransferBatchTerms,
  decodeNativeTokenPeriodTransferTerms,
  decodeERC20TokenPeriodTransferTerms,
  decodeExactExecutionTerms,
  decodeExactCalldataTerms,
  decodeExactCalldataBatchTerms,
  decodeExactExecutionBatchTerms,
  decodeMultiTokenPeriodTerms,
  decodeOwnershipTransferTerms
} from "@metamask/delegation-core";
import {
  encodeAbiParameters,
  parseAbiParameters,
  keccak256,
  toHex as toHex2
} from "viem";
var CAVEAT_ABI_TYPE_COMPONENTS = [
  { type: "address", name: "enforcer" },
  { type: "bytes", name: "terms" },
  { type: "bytes", name: "args" }
];
var CAVEAT_TYPEHASH = keccak256(
  toHex2("Caveat(address enforcer,bytes terms)")
);
var createCaveat = (enforcer, terms, args = "0x00") => {
  return {
    enforcer,
    terms,
    args
  };
};
var decodeCaveat = ({
  caveat: { enforcer, terms },
  environment: { caveatEnforcers }
}) => {
  switch (enforcer.toLowerCase()) {
    case caveatEnforcers.AllowedCalldataEnforcer?.toLowerCase():
      return {
        type: "allowedCalldata" /* AllowedCalldata */,
        ...decodeAllowedCalldataTerms(terms)
      };
    case caveatEnforcers.AllowedMethodsEnforcer?.toLowerCase():
      return {
        type: "allowedMethods" /* AllowedMethods */,
        ...decodeAllowedMethodsTerms(terms)
      };
    case caveatEnforcers.ApprovalRevocationEnforcer?.toLowerCase():
      return {
        type: "approvalRevocation" /* ApprovalRevocation */,
        ...decodeApprovalRevocationTerms(terms)
      };
    case caveatEnforcers.AllowedTargetsEnforcer?.toLowerCase():
      return {
        type: "allowedTargets" /* AllowedTargets */,
        ...decodeAllowedTargetsTerms(terms)
      };
    case caveatEnforcers.ArgsEqualityCheckEnforcer?.toLowerCase():
      return {
        type: "argsEqualityCheck" /* ArgsEqualityCheck */,
        ...decodeArgsEqualityCheckTerms(terms)
      };
    case caveatEnforcers.BlockNumberEnforcer?.toLowerCase():
      return { type: "blockNumber" /* BlockNumber */, ...decodeBlockNumberTerms(terms) };
    case caveatEnforcers.DeployedEnforcer?.toLowerCase():
      return { type: "deployed" /* Deployed */, ...decodeDeployedTerms(terms) };
    case caveatEnforcers.ERC20BalanceChangeEnforcer?.toLowerCase():
      return {
        type: "erc20BalanceChange" /* Erc20BalanceChange */,
        ...decodeERC20BalanceChangeTerms(terms)
      };
    case caveatEnforcers.ERC20TransferAmountEnforcer?.toLowerCase():
      return {
        type: "erc20TransferAmount" /* Erc20TransferAmount */,
        ...decodeERC20TransferAmountTerms(terms)
      };
    case caveatEnforcers.ERC20StreamingEnforcer?.toLowerCase():
      return {
        type: "erc20Streaming" /* Erc20Streaming */,
        ...decodeERC20StreamingTerms(terms)
      };
    case caveatEnforcers.ERC721BalanceChangeEnforcer?.toLowerCase():
      return {
        type: "erc721BalanceChange" /* Erc721BalanceChange */,
        ...decodeERC721BalanceChangeTerms(terms)
      };
    case caveatEnforcers.ERC721TransferEnforcer?.toLowerCase():
      return {
        type: "erc721Transfer" /* Erc721Transfer */,
        ...decodeERC721TransferTerms(terms)
      };
    case caveatEnforcers.ERC1155BalanceChangeEnforcer?.toLowerCase():
      return {
        type: "erc1155BalanceChange" /* Erc1155BalanceChange */,
        ...decodeERC1155BalanceChangeTerms(terms)
      };
    case caveatEnforcers.IdEnforcer?.toLowerCase():
      return { type: "id" /* Id */, ...decodeIdTerms(terms) };
    case caveatEnforcers.LimitedCallsEnforcer?.toLowerCase():
      return {
        type: "limitedCalls" /* LimitedCalls */,
        ...decodeLimitedCallsTerms(terms)
      };
    case caveatEnforcers.NonceEnforcer?.toLowerCase():
      return { type: "nonce" /* Nonce */, ...decodeNonceTerms(terms) };
    case caveatEnforcers.TimestampEnforcer?.toLowerCase():
      return { type: "timestamp" /* Timestamp */, ...decodeTimestampTerms(terms) };
    case caveatEnforcers.ValueLteEnforcer?.toLowerCase():
      return { type: "valueLte" /* ValueLte */, ...decodeValueLteTerms(terms) };
    case caveatEnforcers.NativeTokenTransferAmountEnforcer?.toLowerCase():
      return {
        type: "nativeTokenTransferAmount" /* NativeTokenTransferAmount */,
        ...decodeNativeTokenTransferAmountTerms(terms)
      };
    case caveatEnforcers.NativeBalanceChangeEnforcer?.toLowerCase():
      return {
        type: "nativeBalanceChange" /* NativeBalanceChange */,
        ...decodeNativeBalanceChangeTerms(terms)
      };
    case caveatEnforcers.NativeTokenStreamingEnforcer?.toLowerCase():
      return {
        type: "nativeTokenStreaming" /* NativeTokenStreaming */,
        ...decodeNativeTokenStreamingTerms(terms)
      };
    case caveatEnforcers.NativeTokenPaymentEnforcer?.toLowerCase():
      return {
        type: "nativeTokenPayment" /* NativeTokenPayment */,
        ...decodeNativeTokenPaymentTerms(terms)
      };
    case caveatEnforcers.RedeemerEnforcer?.toLowerCase():
      return { type: "redeemer" /* Redeemer */, ...decodeRedeemerTerms(terms) };
    case caveatEnforcers.SpecificActionERC20TransferBatchEnforcer?.toLowerCase():
      return {
        type: "specificActionERC20TransferBatch" /* SpecificActionERC20TransferBatch */,
        ...decodeSpecificActionERC20TransferBatchTerms(terms)
      };
    case caveatEnforcers.ERC20PeriodTransferEnforcer?.toLowerCase():
      return {
        type: "erc20PeriodTransfer" /* Erc20PeriodTransfer */,
        ...decodeERC20TokenPeriodTransferTerms(terms)
      };
    case caveatEnforcers.NativeTokenPeriodTransferEnforcer?.toLowerCase():
      return {
        type: "nativeTokenPeriodTransfer" /* NativeTokenPeriodTransfer */,
        ...decodeNativeTokenPeriodTransferTerms(terms)
      };
    case caveatEnforcers.ExactCalldataBatchEnforcer?.toLowerCase():
      return {
        type: "exactCalldataBatch" /* ExactCalldataBatch */,
        ...decodeExactCalldataBatchTerms(terms)
      };
    case caveatEnforcers.ExactCalldataEnforcer?.toLowerCase():
      return {
        type: "exactCalldata" /* ExactCalldata */,
        ...decodeExactCalldataTerms(terms)
      };
    case caveatEnforcers.ExactExecutionEnforcer?.toLowerCase():
      return {
        type: "exactExecution" /* ExactExecution */,
        ...decodeExactExecutionTerms(terms)
      };
    case caveatEnforcers.ExactExecutionBatchEnforcer?.toLowerCase():
      return {
        type: "exactExecutionBatch" /* ExactExecutionBatch */,
        ...decodeExactExecutionBatchTerms(terms)
      };
    case caveatEnforcers.MultiTokenPeriodEnforcer?.toLowerCase():
      return {
        type: "multiTokenPeriod" /* MultiTokenPeriod */,
        ...decodeMultiTokenPeriodTerms(terms)
      };
    case caveatEnforcers.OwnershipTransferEnforcer?.toLowerCase():
      return {
        type: "ownershipTransfer" /* OwnershipTransfer */,
        ...decodeOwnershipTransferTerms(terms)
      };
    default:
      throw new Error(`Unknown enforcer address: ${enforcer}`);
  }
};

// src/delegation.ts
import {
  encodeDelegations as encodeDelegationsCore,
  decodeDelegations as decodeDelegationsCore,
  encodeDelegation as encodeDelegationCore,
  decodeDelegation as decodeDelegationCore,
  hashDelegation as hashDelegationCore,
  ANY_BENEFICIARY,
  DELEGATION_TYPEHASH,
  CAVEAT_TYPEHASH as CAVEAT_TYPEHASH2,
  ROOT_AUTHORITY
} from "@metamask/delegation-core";
import { toHex as toHex3, getAddress, isHex as isHex5 } from "viem";
import { signTypedData } from "viem/accounts";

// src/analytics/index.ts
import createClient from "openapi-fetch";

// src/analytics/environment.ts
var session;
function createAnonId() {
  if (typeof globalThis.crypto?.randomUUID === "function") {
    return globalThis.crypto.randomUUID();
  }
  return `00000000-0000-4000-8000-${Math.random().toString(16).slice(2, 14).padEnd(12, "0")}`;
}
function inferPlatform() {
  if (typeof globalThis === "undefined" || !("window" in globalThis)) {
    return "nodejs";
  }
  const nav = globalThis.navigator;
  const ua = typeof nav?.userAgent === "string" ? nav.userAgent : "";
  if (/Mobile|Android|iPhone|iPod|webOS|BlackBerry|IEMobile|Opera Mini/iu.test(ua)) {
    return "web-mobile";
  }
  return "web-desktop";
}
function inferDomain() {
  if (typeof globalThis !== "undefined" && "window" in globalThis) {
    const win = globalThis;
    const hostname = win.location?.hostname;
    if (typeof hostname === "string" && hostname.length > 0) {
      return { domain: hostname };
    }
  }
  return {};
}
function getInitializationContext(params) {
  if (!session) {
    session = {
      sdk_version: params.sdk_version,
      anon_id: params.anon_id ?? createAnonId(),
      platform: inferPlatform(),
      ...inferDomain()
    };
    return { ...session };
  }
  session = {
    ...session,
    sdk_version: params.sdk_version
  };
  return { ...session };
}
function getSessionBaseProperties() {
  if (!session) {
    throw new Error(
      "Smart Accounts Kit analytics: call getInitializationContext({ sdk_version }) at SDK startup before recording events."
    );
  }
  return { ...session };
}
function mergeSessionProperties(partial) {
  if (!session) {
    throw new Error(
      "Smart Accounts Kit analytics: call getInitializationContext before mergeSessionProperties."
    );
  }
  session = { ...session, ...partial };
}

// src/analytics/sender.ts
var Sender = class {
  #sendFn;
  #batchSize;
  #baseTimeoutMs;
  #maxFailureCount;
  #maxTimeoutMs;
  #isDisabled = false;
  #batch = [];
  #failureCount = 0;
  #timeoutId = null;
  #isSending = false;
  constructor(options) {
    this.#batchSize = options.batchSize;
    this.#baseTimeoutMs = options.baseTimeoutMs;
    this.#maxFailureCount = options.maxFailureCount;
    this.#sendFn = options.sendFn;
    this.#maxTimeoutMs = options.maxTimeoutMs;
  }
  enqueue(item) {
    if (this.#isDisabled) {
      return;
    }
    this.#batch.push(item);
    this.#schedule();
  }
  #schedule() {
    if (this.#isDisabled) {
      return;
    }
    if (this.#batch.length > 0 && !this.#timeoutId) {
      const { isRetry, timeoutMs } = this.#getTimeout();
      const timeout = setTimeout(() => {
        this.#timeoutId = null;
        this.#flush();
      }, timeoutMs);
      if (isRetry && timeout.unref) {
        timeout.unref();
      }
      this.#timeoutId = timeout;
    }
  }
  async #flush() {
    if (this.#isDisabled || this.#isSending || this.#batch.length === 0) {
      return;
    }
    this.#isSending = true;
    const current = this.#batch.slice(0, this.#batchSize);
    this.#batch = this.#batch.slice(this.#batchSize);
    try {
      await this.#sendFn(current);
      this.#failureCount = 0;
    } catch {
      this.#failureCount += 1;
      if (this.#failureCount >= this.#maxFailureCount) {
        this.#isDisabled = true;
        this.#batch = [];
        if (this.#timeoutId !== null) {
          clearTimeout(this.#timeoutId);
          this.#timeoutId = null;
        }
      } else {
        this.#batch = [...current, ...this.#batch];
      }
    } finally {
      this.#isSending = false;
      this.#schedule();
    }
  }
  #getTimeout() {
    const isRetry = this.#failureCount > 0;
    const timeoutMs = Math.min(
      this.#baseTimeoutMs * 2 ** this.#failureCount,
      this.#maxTimeoutMs
    );
    return { isRetry, timeoutMs };
  }
};

// package.json
var version = "1.6.0";

// src/analytics/index.ts
function isCompleteBase(value) {
  return typeof value.sdk_version === "string" && value.sdk_version.length > 0 && typeof value.anon_id === "string" && value.anon_id.length > 0 && typeof value.platform === "string" && value.platform.length > 0;
}
function normalise(batch) {
  const walk = (value) => {
    if (typeof value === "bigint") {
      return `${value}n`;
    }
    if (value === null || typeof value !== "object") {
      return value;
    }
    if (Array.isArray(value)) {
      return value.map((item) => walk(item));
    }
    const result = {};
    for (const [key, child] of Object.entries(
      value
    )) {
      result[key] = walk(child);
    }
    return result;
  };
  return walk(batch);
}
var METAMASK_ANALYTICS_ENDPOINT = "https://mm-sdk-analytics.api.cx.metamask.io/";
var Analytics = class {
  enabled = false;
  sender;
  constructor(baseUrl) {
    const client = createClient({ baseUrl });
    const sendFn = async (batch) => {
      const normalisedBatch = normalise(batch);
      const res = await client.POST("/v2/events", { body: normalisedBatch });
      if (res.response.status !== 200) {
        throw new Error(String(res.error));
      }
    };
    this.sender = new Sender({
      batchSize: 100,
      baseTimeoutMs: 100,
      maxFailureCount: 6,
      maxTimeoutMs: 1e4,
      sendFn
    });
  }
  enable() {
    this.enabled = true;
  }
  /**
   * Merges a field into the session base (shared with {@link getSessionBaseProperties}).
   *
   * @param key - Base property name.
   * @param value - Value for that property.
   */
  setGlobalProperty(key, value) {
    mergeSessionProperties({
      [key]: value
    });
  }
  /**
   * Sends `smart_accounts_kit_initialized` using the session base from {@link getInitializationContext}
   * plus optional per-event overrides. Updates the stored session with the merged snapshot.
   *
   * @param properties - Optional overrides; omit to use the current session base only.
   */
  trackInitialized(properties = {}) {
    if (!this.enabled) {
      return;
    }
    const merged = {
      ...getSessionBaseProperties(),
      ...properties
    };
    if (!isCompleteBase(merged)) {
      throw new Error(
        "Analytics: trackInitialized produced incomplete base configuration (ensure getInitializationContext ran and sdk_version, anon_id, platform are set)"
      );
    }
    mergeSessionProperties(merged);
    const event = {
      namespace: "metamask/smart-accounts-kit",
      event_name: "smart_accounts_kit_initialized",
      properties: merged
    };
    this.sender.enqueue(event);
  }
  /**
   * Sends `smart_accounts_kit_function_called` with session base plus the function name and optional
   * non-sensitive parameters. Does not mutate the session store (unlike {@link trackInitialized}).
   *
   * @param functionName - Public SDK entry name (use a stable string, e.g. `createDelegation`).
   * @param parameters - Safe primitives only; omit secrets, keys, and raw addresses if sensitive.
   * @param baseOverrides - Optional overrides for base fields (same as {@link trackInitialized}).
   */
  trackSdkFunctionCall(functionName, parameters, baseOverrides = {}) {
    if (!this.enabled) {
      return;
    }
    const mergedBase = {
      ...getSessionBaseProperties(),
      ...baseOverrides
    };
    if (!isCompleteBase(mergedBase)) {
      throw new Error(
        "Analytics: trackSdkFunctionCall requires session (call getInitializationContext before tracking)"
      );
    }
    const props = {
      ...mergedBase,
      function_name: functionName,
      ...parameters !== void 0 && Object.keys(parameters).length > 0 ? { parameters } : {}
    };
    const event = {
      namespace: "metamask/smart-accounts-kit",
      event_name: "smart_accounts_kit_function_called",
      properties: props
    };
    this.sender.enqueue(event);
  }
};
var analytics = new Analytics(METAMASK_ANALYTICS_ENDPOINT);
function isAnalyticsDisabled() {
  const dntValues = [];
  if (typeof process !== "undefined") {
    dntValues.push(process.env?.CI);
    dntValues.push(process.env?.DO_NOT_TRACK);
  }
  if (typeof navigator !== "undefined") {
    dntValues.push(navigator.doNotTrack);
  }
  if (typeof window !== "undefined") {
    dntValues.push(window.doNotTrack);
  }
  return dntValues.some(
    (dntValue) => dntValue === "1" || dntValue?.toLowerCase() === "yes" || dntValue?.toLowerCase() === "true"
  );
}
var hasBootstrapped = false;
function ensureSmartAccountsKitAnalyticsBootstrapped() {
  if (hasBootstrapped) {
    return;
  }
  hasBootstrapped = true;
  if (isAnalyticsDisabled()) {
    return;
  }
  try {
    getInitializationContext({ sdk_version: version });
    analytics.enable();
    analytics.trackInitialized();
  } catch {
  }
}
function trackSmartAccountsKitFunctionCall(functionName, parameters) {
  ensureSmartAccountsKitAnalyticsBootstrapped();
  try {
    analytics.trackSdkFunctionCall(functionName, parameters);
  } catch {
  }
}

// src/delegation.ts
var DELEGATION_ABI_TYPE_COMPONENTS = [
  { type: "address", name: "delegate" },
  { type: "address", name: "delegator" },
  { type: "bytes32", name: "authority" },
  { type: "tuple[]", name: "caveats", components: CAVEAT_ABI_TYPE_COMPONENTS },
  { type: "uint256", name: "salt" },
  { type: "bytes", name: "signature" }
];
var toDelegationStruct = (delegation) => {
  const caveats = delegation.caveats.map((caveat) => ({
    enforcer: getAddress(caveat.enforcer),
    terms: caveat.terms,
    args: caveat.args
  }));
  const salt = delegation.salt === "0x" ? 0n : BigInt(delegation.salt);
  return {
    delegate: getAddress(delegation.delegate),
    delegator: getAddress(delegation.delegator),
    authority: delegation.authority ?? ROOT_AUTHORITY,
    caveats,
    salt,
    signature: delegation.signature
  };
};
var toDelegation = (delegationStruct) => {
  return {
    ...delegationStruct,
    salt: toHex3(delegationStruct.salt)
  };
};
var encodeDelegations = (delegations) => {
  if (Array.isArray(delegations)) {
    const delegationStructs = delegations.map(toDelegationStruct);
    return encodeDelegationsCore(delegationStructs);
  }
  if (isHex5(delegations)) {
    return delegations;
  }
  throw new Error(
    "Invalid delegations - must be an array of delegations or a hex string"
  );
};
var encodeDelegation = (delegation) => {
  const delegationStruct = toDelegationStruct(delegation);
  return encodeDelegationCore(delegationStruct);
};
var decodeDelegations = (delegations) => {
  if (Array.isArray(delegations)) {
    return delegations;
  }
  if (isHex5(delegations)) {
    return decodeDelegationsCore(delegations).map(toDelegation);
  }
  throw new Error(
    "Invalid delegations - must be an array of delegations or a hex string"
  );
};
var decodeDelegation = (encoded) => {
  return toDelegation(decodeDelegationCore(encoded));
};
var SIGNABLE_DELEGATION_TYPED_DATA = {
  Caveat: [
    { name: "enforcer", type: "address" },
    { name: "terms", type: "bytes" }
  ],
  Delegation: [
    { name: "delegate", type: "address" },
    { name: "delegator", type: "address" },
    { name: "authority", type: "bytes32" },
    { name: "caveats", type: "Caveat[]" },
    { name: "salt", type: "uint256" }
  ]
};
var DELEGATION_ARRAY_ABI_TYPE = {
  type: "tuple[]",
  components: DELEGATION_ABI_TYPE_COMPONENTS
};
var DELEGATION_ABI_TYPE = {
  type: "tuple",
  components: DELEGATION_ABI_TYPE_COMPONENTS
};
var hashDelegation = (input) => {
  const delegationStruct = toDelegationStruct(input);
  return hashDelegationCore(delegationStruct);
};
var extractLeafDelegation = (permissionContext) => {
  const delegations = decodeDelegations(permissionContext);
  if (delegations.length === 0) {
    throw new Error("Permission context must contain at least one delegation");
  }
  const leafDelegation = delegations[0];
  if (!leafDelegation) {
    throw new Error("Failed to extract leaf delegation");
  }
  return leafDelegation;
};
var getCaveatNames = ({
  caveats,
  environment: { caveatEnforcers }
}) => {
  if (Array.isArray(caveats)) {
    const knownEnforcers = Object.entries(caveatEnforcers).map(
      ([name, address]) => ({ name, address: address.toLowerCase() })
    );
    return caveats.map((caveat) => {
      const enforcerAddressLowercase = caveat.enforcer?.toLowerCase();
      const matchingCaveat = knownEnforcers.find(
        ({ address }) => address === enforcerAddressLowercase
      );
      return matchingCaveat?.name ?? "Unknown";
    });
  }
  return [];
};
var resolveDelegationArgs = (options) => {
  const optionsHasParentPermissionContext = "parentPermissionContext" in options && options.parentPermissionContext;
  const resolvedParentDelegation = optionsHasParentPermissionContext ? extractLeafDelegation(options.parentPermissionContext) : options.parentDelegation;
  const authority = (() => {
    if (!resolvedParentDelegation) {
      if (!options.scope) {
        throw new Error(
          "Invalid arguments - must specify either parentDelegation, parentPermissionContext, or scope"
        );
      }
      return ROOT_AUTHORITY;
    }
    if (typeof resolvedParentDelegation === "string") {
      if (resolvedParentDelegation.toLowerCase() === ROOT_AUTHORITY.toLowerCase()) {
        throw new Error(
          `Invalid parent delegation - cannot be ${ROOT_AUTHORITY}`
        );
      }
      return resolvedParentDelegation;
    }
    return hashDelegation(resolvedParentDelegation);
  })();
  const hasInheritance = resolvedParentDelegation !== void 0;
  const caveats = resolveCaveats({
    environment: options.environment,
    scope: options.scope,
    caveats: options.caveats,
    isScopeOptional: hasInheritance
  });
  return {
    authority,
    caveats,
    hasInheritance,
    salt: options.salt ?? "0x00"
  };
};
var createDelegation = (options) => {
  const { authority, caveats, hasInheritance, salt } = resolveDelegationArgs(options);
  trackSmartAccountsKitFunctionCall("createDelegation", {
    hasInheritance,
    scope: options.scope?.type ?? null,
    caveatNames: getCaveatNames({
      caveats,
      environment: options.environment
    })
  });
  return {
    delegate: options.to,
    delegator: options.from,
    authority,
    caveats,
    salt,
    signature: "0x"
  };
};
var createOpenDelegation = (options) => {
  const { authority, caveats, hasInheritance, salt } = resolveDelegationArgs(options);
  trackSmartAccountsKitFunctionCall("createOpenDelegation", {
    hasInheritance,
    scope: options.scope?.type ?? null,
    caveatNames: getCaveatNames({
      caveats,
      environment: options.environment
    })
  });
  return {
    delegate: ANY_BENEFICIARY,
    delegator: options.from,
    authority,
    caveats,
    salt,
    signature: "0x"
  };
};
var prepareSignDelegationTypedData = ({
  delegation,
  delegationManager,
  chainId,
  name = "DelegationManager",
  version: version2 = "1",
  allowInsecureUnrestrictedDelegation = false
}) => {
  const delegationStruct = toDelegationStruct({
    ...delegation,
    signature: "0x"
  });
  if (delegationStruct.caveats.length === 0 && !allowInsecureUnrestrictedDelegation) {
    throw new Error(
      "No caveats found. If you definitely want to sign a delegation without caveats, set `allowInsecureUnrestrictedDelegation` to `true`."
    );
  }
  return {
    domain: {
      chainId,
      name,
      version: version2,
      verifyingContract: delegationManager
    },
    types: SIGNABLE_DELEGATION_TYPED_DATA,
    primaryType: "Delegation",
    message: delegationStruct
  };
};
var signDelegation = async ({
  privateKey,
  delegation,
  delegationManager,
  chainId,
  name = "DelegationManager",
  version: version2 = "1",
  allowInsecureUnrestrictedDelegation = false
}) => {
  const typedData = prepareSignDelegationTypedData({
    delegation,
    delegationManager,
    chainId,
    name,
    version: version2,
    allowInsecureUnrestrictedDelegation
  });
  return signTypedData({
    privateKey,
    ...typedData
  });
};

// src/executions.ts
import {
  encodeAbiParameters as encodeAbiParameters2,
  encodePacked
} from "viem";
var createExecution = ({
  target,
  value = 0n,
  callData = "0x"
}) => {
  return {
    target,
    value,
    callData
  };
};
var ExecutionMode = /* @__PURE__ */ ((ExecutionMode2) => {
  ExecutionMode2["SingleDefault"] = "0x0000000000000000000000000000000000000000000000000000000000000000";
  ExecutionMode2["SingleTry"] = "0x0001000000000000000000000000000000000000000000000000000000000000";
  ExecutionMode2["BatchDefault"] = "0x0100000000000000000000000000000000000000000000000000000000000000";
  ExecutionMode2["BatchTry"] = "0x0101000000000000000000000000000000000000000000000000000000000000";
  return ExecutionMode2;
})(ExecutionMode || {});
var EXECUTION_ABI_TYPE_COMPONENTS = [
  { type: "address", name: "target" },
  { type: "uint256", name: "value" },
  { type: "bytes", name: "callData" }
];
var encodeSingleExecution = (execution) => {
  return encodePacked(
    ["address", "uint256", "bytes"],
    [execution.target, execution.value, execution.callData]
  );
};
var encodeBatchExecution = (executions) => {
  return encodeAbiParameters2(
    [
      {
        components: EXECUTION_ABI_TYPE_COMPONENTS,
        name: "executions",
        type: "tuple[]"
      }
    ],
    [executions]
  );
};
var encodeExecutionCalldata = (executions) => {
  if (executions.length === 0) {
    throw new Error(
      "Error while getting the execution calldatas, executions is empty"
    );
  }
  if (executions.length === 1) {
    const execution = executions[0];
    if (!execution) {
      throw new Error("Execution not found");
    }
    return encodeSingleExecution(execution);
  }
  return encodeBatchExecution(executions);
};
var encodeExecutionCalldatas = (executionsBatch) => {
  if (executionsBatch.length === 0) {
    throw new Error(
      "Error while getting the execution calldatas, executionsBatch is empty"
    );
  }
  return executionsBatch.map(encodeExecutionCalldata);
};

export {
  __export,
  trackSmartAccountsKitFunctionCall,
  Implementation,
  TransferWindow,
  ScopeType,
  CaveatType,
  CaveatBuilder,
  createCaveatBuilder,
  isDefined,
  toHexOrThrow,
  resolveCaveats,
  createCaveat,
  decodeCaveat,
  ANY_BENEFICIARY,
  DELEGATION_TYPEHASH,
  ROOT_AUTHORITY,
  DELEGATION_ABI_TYPE_COMPONENTS,
  toDelegationStruct,
  toDelegation,
  encodeDelegations,
  encodeDelegation,
  decodeDelegations,
  decodeDelegation,
  SIGNABLE_DELEGATION_TYPED_DATA,
  DELEGATION_ARRAY_ABI_TYPE,
  DELEGATION_ABI_TYPE,
  hashDelegation,
  createDelegation,
  createOpenDelegation,
  prepareSignDelegationTypedData,
  signDelegation,
  createExecution,
  ExecutionMode,
  encodeSingleExecution,
  encodeBatchExecution,
  encodeExecutionCalldata,
  encodeExecutionCalldatas
};
