// @metamask/smart-accounts-kit/dist/chunk-W6HPAK4M.cjs
"use strict";Object.defineProperty(exports, "__esModule", {value: true}); function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; } function _nullishCoalesce(lhs, rhsFn) { if (lhs != null) { return lhs; } else { return rhsFn(); } } function _optionalChain(ops) { let lastAccessLHS = undefined; let value = ops[0]; let i = 1; while (i < ops.length) { const op = ops[i]; const fn = ops[i + 1]; i += 2; if ((op === 'optionalAccess' || op === 'optionalCall') && value == null) { return undefined; } if (op === 'access' || op === 'optionalAccess') { lastAccessLHS = value; value = fn(value); } else if (op === 'call' || op === 'optionalCall') { value = fn((...args) => value.call(lastAccessLHS, ...args)); lastAccessLHS = undefined; } } return value; } var _class;var __defProp = Object.defineProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};

// src/constants.ts
var Implementation = /* @__PURE__ */ ((Implementation2) => {
  Implementation2["MultiSig"] = "MultiSig";
  Implementation2["Hybrid"] = "Hybrid";
  Implementation2["Stateless7702"] = "Stateless7702";
  return Implementation2;
})(Implementation || {});
var TransferWindow = /* @__PURE__ */ ((TransferWindow2) => {
  TransferWindow2[TransferWindow2["Hourly"] = 3600] = "Hourly";
  TransferWindow2[TransferWindow2["Daily"] = 86400] = "Daily";
  TransferWindow2[TransferWindow2["Weekly"] = 604800] = "Weekly";
  TransferWindow2[TransferWindow2["BiWeekly"] = 1209600] = "BiWeekly";
  TransferWindow2[TransferWindow2["Monthly"] = 2592e3] = "Monthly";
  TransferWindow2[TransferWindow2["Quarterly"] = 7776e3] = "Quarterly";
  TransferWindow2[TransferWindow2["Yearly"] = 31536e3] = "Yearly";
  return TransferWindow2;
})(TransferWindow || {});
var ScopeType = /* @__PURE__ */ ((ScopeType2) => {
  ScopeType2["Erc20TransferAmount"] = "erc20TransferAmount";
  ScopeType2["Erc20Streaming"] = "erc20Streaming";
  ScopeType2["Erc20PeriodTransfer"] = "erc20PeriodTransfer";
  ScopeType2["NativeTokenTransferAmount"] = "nativeTokenTransferAmount";
  ScopeType2["NativeTokenStreaming"] = "nativeTokenStreaming";
  ScopeType2["NativeTokenPeriodTransfer"] = "nativeTokenPeriodTransfer";
  ScopeType2["Erc721Transfer"] = "erc721Transfer";
  ScopeType2["OwnershipTransfer"] = "ownershipTransfer";
  ScopeType2["FunctionCall"] = "functionCall";
  return ScopeType2;
})(ScopeType || {});
var CaveatType = /* @__PURE__ */ ((CaveatType2) => {
  CaveatType2["AllowedMethods"] = "allowedMethods";
  CaveatType2["AllowedTargets"] = "allowedTargets";
  CaveatType2["ApprovalRevocation"] = "approvalRevocation";
  CaveatType2["Deployed"] = "deployed";
  CaveatType2["AllowedCalldata"] = "allowedCalldata";
  CaveatType2["Erc20BalanceChange"] = "erc20BalanceChange";
  CaveatType2["Erc721BalanceChange"] = "erc721BalanceChange";
  CaveatType2["Erc1155BalanceChange"] = "erc1155BalanceChange";
  CaveatType2["ValueLte"] = "valueLte";
  CaveatType2["LimitedCalls"] = "limitedCalls";
  CaveatType2["Id"] = "id";
  CaveatType2["Nonce"] = "nonce";
  CaveatType2["Timestamp"] = "timestamp";
  CaveatType2["BlockNumber"] = "blockNumber";
  CaveatType2["Erc20TransferAmount"] = "erc20TransferAmount";
  CaveatType2["Erc20Streaming"] = "erc20Streaming";
  CaveatType2["NativeTokenStreaming"] = "nativeTokenStreaming";
  CaveatType2["Erc721Transfer"] = "erc721Transfer";
  CaveatType2["NativeTokenTransferAmount"] = "nativeTokenTransferAmount";
  CaveatType2["NativeBalanceChange"] = "nativeBalanceChange";
  CaveatType2["Redeemer"] = "redeemer";
  CaveatType2["NativeTokenPayment"] = "nativeTokenPayment";
  CaveatType2["ArgsEqualityCheck"] = "argsEqualityCheck";
  CaveatType2["SpecificActionERC20TransferBatch"] = "specificActionERC20TransferBatch";
  CaveatType2["Erc20PeriodTransfer"] = "erc20PeriodTransfer";
  CaveatType2["NativeTokenPeriodTransfer"] = "nativeTokenPeriodTransfer";
  CaveatType2["ExactCalldataBatch"] = "exactCalldataBatch";
  CaveatType2["ExactCalldata"] = "exactCalldata";
  CaveatType2["ExactExecution"] = "exactExecution";
  CaveatType2["ExactExecutionBatch"] = "exactExecutionBatch";
  CaveatType2["MultiTokenPeriod"] = "multiTokenPeriod";
  CaveatType2["OwnershipTransfer"] = "ownershipTransfer";
  return CaveatType2;
})(CaveatType || {});

// src/caveatBuilder/caveatBuilder.ts
var INSECURE_UNRESTRICTED_DELEGATION_ERROR_MESSAGE = "No caveats found. If you definitely want to create an empty caveat collection, set `allowInsecureUnrestrictedDelegation` to `true`.";
var CaveatBuilder = class _CaveatBuilder {
  #results = [];
  #hasBeenBuilt = false;
  #environment;
  #config;
  #enforcerBuilders;
  constructor(environment, config = {}, enforcerBuilders = {}, builtCaveats = []) {
    this.#environment = environment;
    this.#config = config;
    this.#enforcerBuilders = enforcerBuilders;
    this.#results = builtCaveats;
  }
  /**
   * Extends the CaveatBuilder with a new enforcer function.
   *
   * @template TEnforcerName - The name of the enforcer.
   * @template TFunction - The type of the enforcer function.
   * @param name - The name of the enforcer.
   * @param fn - The enforcer function.
   * @returns The extended CaveatBuilder instance.
   */
  extend(name, fn) {
    return new _CaveatBuilder(
      this.#environment,
      this.#config,
      { ...this.#enforcerBuilders, [name]: fn },
      this.#results
    );
  }
  addCaveat(nameOrCaveat, config) {
    if (typeof nameOrCaveat === "object") {
      const caveat = {
        args: "0x00",
        ...nameOrCaveat
      };
      this.#results = [...this.#results, caveat];
      return this;
    }
    const name = nameOrCaveat;
    const func = this.#enforcerBuilders[name];
    if (typeof func === "function") {
      const result = func(this.#environment, config);
      this.#results = [...this.#results, result];
      return this;
    }
    throw new Error(`Function "${String(name)}" does not exist.`);
  }
  /**
   * Returns the caveats that have been built using this CaveatBuilder.
   *
   * @returns The array of built caveats.
   * @throws Error if the builder has already been built or if no caveats are found and empty caveats are not allowed.
   */
  build() {
    if (this.#hasBeenBuilt) {
      throw new Error("This CaveatBuilder has already been built.");
    }
    if (this.#results.length === 0 && !this.#config.allowInsecureUnrestrictedDelegation) {
      throw new Error(INSECURE_UNRESTRICTED_DELEGATION_ERROR_MESSAGE);
    }
    this.#hasBeenBuilt = true;
    return this.#results;
  }
};

// src/caveatBuilder/allowedCalldataBuilder.ts
var _delegationcore = require('@metamask/delegation-core');
var allowedCalldata = "allowedCalldata";
var allowedCalldataBuilder = (environment, config) => {
  const { startIndex, value } = config;
  const terms = _delegationcore.createAllowedCalldataTerms.call(void 0, { startIndex, value });
  const {
    caveatEnforcers: { AllowedCalldataEnforcer }
  } = environment;
  if (!AllowedCalldataEnforcer) {
    throw new Error("AllowedCalldataEnforcer not found in environment");
  }
  return {
    enforcer: AllowedCalldataEnforcer,
    terms,
    args: "0x00"
  };
};

// src/caveatBuilder/allowedMethodsBuilder.ts

var _viem = require('viem');
var allowedMethods = "allowedMethods";
var FUNCTION_SELECTOR_STRING_LENGTH = 10;
var allowedMethodsBuilder = (environment, config) => {
  const { selectors } = config;
  if (selectors.length === 0) {
    throw new Error("Invalid selectors: must provide at least one selector");
  }
  const parsedSelectors = selectors.map(parseSelector);
  const terms = _delegationcore.createAllowedMethodsTerms.call(void 0, { selectors: parsedSelectors });
  const {
    caveatEnforcers: { AllowedMethodsEnforcer }
  } = environment;
  if (!AllowedMethodsEnforcer) {
    throw new Error("AllowedMethodsEnforcer not found in environment");
  }
  return {
    enforcer: AllowedMethodsEnforcer,
    terms,
    args: "0x00"
  };
};
function parseSelector(selector) {
  if (_viem.isHex.call(void 0, selector)) {
    if (selector.length === FUNCTION_SELECTOR_STRING_LENGTH) {
      return selector;
    }
    throw new Error(
      "Invalid selector: must be a 4 byte hex string, abi function signature, or AbiFunction"
    );
  }
  try {
    return _viem.toFunctionSelector.call(void 0, selector);
  } catch (rootError) {
    throw new Error(
      "Invalid selector: must be a 4 byte hex string, abi function signature, or AbiFunction",
      { cause: rootError }
    );
  }
}

// src/caveatBuilder/allowedTargetsBuilder.ts


var allowedTargets = "allowedTargets";
var allowedTargetsBuilder = (environment, config) => {
  const { targets } = config;
  if (targets.length === 0) {
    throw new Error(
      "Invalid targets: must provide at least one target address"
    );
  }
  const invalidAddresses = targets.filter(
    (target) => !_viem.isAddress.call(void 0, target, { strict: false })
  );
  if (invalidAddresses.length > 0) {
    throw new Error("Invalid targets: must be valid addresses");
  }
  const terms = _delegationcore.createAllowedTargetsTerms.call(void 0, { targets });
  const {
    caveatEnforcers: { AllowedTargetsEnforcer }
  } = environment;
  if (!AllowedTargetsEnforcer) {
    throw new Error("AllowedTargetsEnforcer not found in environment");
  }
  return {
    enforcer: AllowedTargetsEnforcer,
    terms,
    args: "0x00"
  };
};

// src/caveatBuilder/approvalRevocationBuilder.ts



var approvalRevocation = "approvalRevocation";
var approvalRevocationBuilder = (environment, config) => {
  const terms = _delegationcore.createApprovalRevocationTerms.call(void 0, config);
  const {
    caveatEnforcers: { ApprovalRevocationEnforcer }
  } = environment;
  if (!ApprovalRevocationEnforcer) {
    throw new Error("ApprovalRevocationEnforcer not found in environment");
  }
  return {
    enforcer: ApprovalRevocationEnforcer,
    terms,
    args: "0x00"
  };
};

// src/caveatBuilder/argsEqualityCheckBuilder.ts


var argsEqualityCheck = "argsEqualityCheck";
var argsEqualityCheckBuilder = (environment, config) => {
  const { args } = config;
  if (!_viem.isHex.call(void 0, args)) {
    throw new Error("Invalid config: args must be a valid hex string");
  }
  const {
    caveatEnforcers: { ArgsEqualityCheckEnforcer }
  } = environment;
  if (!ArgsEqualityCheckEnforcer) {
    throw new Error("ArgsEqualityCheckEnforcer not found in environment");
  }
  return {
    enforcer: ArgsEqualityCheckEnforcer,
    terms: _delegationcore.createArgsEqualityCheckTerms.call(void 0, { args }),
    args: "0x00"
  };
};

// src/caveatBuilder/blockNumberBuilder.ts

var blockNumber = "blockNumber";
var blockNumberBuilder = (environment, config) => {
  const { afterThreshold, beforeThreshold } = config;
  if (afterThreshold === 0n && beforeThreshold === 0n) {
    throw new Error(
      "Invalid thresholds: At least one of afterThreshold or beforeThreshold must be specified"
    );
  }
  if (beforeThreshold !== 0n && afterThreshold >= beforeThreshold) {
    throw new Error(
      "Invalid thresholds: afterThreshold must be less than beforeThreshold if both are specified"
    );
  }
  const terms = _delegationcore.createBlockNumberTerms.call(void 0, { afterThreshold, beforeThreshold });
  const {
    caveatEnforcers: { BlockNumberEnforcer }
  } = environment;
  if (!BlockNumberEnforcer) {
    throw new Error("BlockNumberEnforcer not found in environment");
  }
  return {
    enforcer: BlockNumberEnforcer,
    terms,
    args: "0x00"
  };
};

// src/caveatBuilder/deployedBuilder.ts


var deployed = "deployed";
var deployedBuilder = (environment, config) => {
  const { contractAddress, salt, bytecode } = config;
  if (!_viem.isAddress.call(void 0, contractAddress, { strict: false })) {
    throw new Error(
      `Invalid contractAddress: must be a valid Ethereum address`
    );
  }
  if (!_viem.isHex.call(void 0, salt)) {
    throw new Error("Invalid salt: must be a valid hexadecimal string");
  }
  if (!_viem.isHex.call(void 0, bytecode)) {
    throw new Error("Invalid bytecode: must be a valid hexadecimal string");
  }
  const terms = _delegationcore.createDeployedTerms.call(void 0, { contractAddress, salt, bytecode });
  const {
    caveatEnforcers: { DeployedEnforcer }
  } = environment;
  if (!DeployedEnforcer) {
    throw new Error("DeployedEnforcer not found in environment");
  }
  return {
    enforcer: DeployedEnforcer,
    terms,
    args: "0x00"
  };
};

// src/caveatBuilder/erc1155BalanceChangeBuilder.ts





var erc1155BalanceChange = "erc1155BalanceChange";
var erc1155BalanceChangeBuilder = (environment, config) => {
  const { tokenAddress, recipient, tokenId, balance, changeType } = config;
  if (!_viem.isAddress.call(void 0, tokenAddress, { strict: false })) {
    throw new Error("Invalid tokenAddress: must be a valid address");
  }
  if (!_viem.isAddress.call(void 0, recipient, { strict: false })) {
    throw new Error("Invalid recipient: must be a valid address");
  }
  if (balance <= 0n) {
    throw new Error("Invalid balance: must be a positive number");
  }
  if (tokenId < 0n) {
    throw new Error("Invalid tokenId: must be a non-negative number");
  }
  if (changeType !== _delegationcore.BalanceChangeType.Increase && changeType !== _delegationcore.BalanceChangeType.Decrease) {
    throw new Error("Invalid changeType: must be either Increase or Decrease");
  }
  const terms = _delegationcore.createERC1155BalanceChangeTerms.call(void 0, {
    tokenAddress,
    recipient,
    tokenId,
    balance,
    changeType
  });
  const {
    caveatEnforcers: { ERC1155BalanceChangeEnforcer }
  } = environment;
  if (!ERC1155BalanceChangeEnforcer) {
    throw new Error("ERC1155BalanceChangeEnforcer not found in environment");
  }
  return {
    enforcer: ERC1155BalanceChangeEnforcer,
    terms,
    args: "0x00"
  };
};

// src/caveatBuilder/erc20BalanceChangeBuilder.ts





var erc20BalanceChange = "erc20BalanceChange";
var erc20BalanceChangeBuilder = (environment, config) => {
  const { tokenAddress, recipient, balance, changeType } = config;
  if (!_viem.isAddress.call(void 0, tokenAddress, { strict: false })) {
    throw new Error("Invalid tokenAddress: must be a valid address");
  }
  if (balance <= 0n) {
    throw new Error("Invalid balance: must be a positive number");
  }
  if (changeType !== _delegationcore.BalanceChangeType.Increase && changeType !== _delegationcore.BalanceChangeType.Decrease) {
    throw new Error("Invalid changeType: must be either Increase or Decrease");
  }
  const terms = _delegationcore.createERC20BalanceChangeTerms.call(void 0, {
    tokenAddress,
    recipient,
    balance,
    changeType
  });
  const {
    caveatEnforcers: { ERC20BalanceChangeEnforcer }
  } = environment;
  if (!ERC20BalanceChangeEnforcer) {
    throw new Error("ERC20BalanceChangeEnforcer not found in environment");
  }
  return {
    enforcer: ERC20BalanceChangeEnforcer,
    terms,
    args: "0x00"
  };
};

// src/caveatBuilder/erc20PeriodTransferBuilder.ts

var erc20PeriodTransfer = "erc20PeriodTransfer";
var erc20PeriodTransferBuilder = (environment, config) => {
  const { tokenAddress, periodAmount, periodDuration, startDate } = config;
  const terms = _delegationcore.createERC20TokenPeriodTransferTerms.call(void 0, {
    tokenAddress,
    periodAmount,
    periodDuration,
    startDate
  });
  const {
    caveatEnforcers: { ERC20PeriodTransferEnforcer }
  } = environment;
  if (!ERC20PeriodTransferEnforcer) {
    throw new Error("ERC20PeriodTransferEnforcer not found in environment");
  }
  return {
    enforcer: ERC20PeriodTransferEnforcer,
    terms,
    args: "0x00"
  };
};

// src/caveatBuilder/erc20StreamingBuilder.ts

var erc20Streaming = "erc20Streaming";
var erc20StreamingBuilder = (environment, config) => {
  const { tokenAddress, initialAmount, maxAmount, amountPerSecond, startTime } = config;
  const terms = _delegationcore.createERC20StreamingTerms.call(void 0, {
    tokenAddress,
    initialAmount,
    maxAmount,
    amountPerSecond,
    startTime
  });
  const {
    caveatEnforcers: { ERC20StreamingEnforcer }
  } = environment;
  if (!ERC20StreamingEnforcer) {
    throw new Error("ERC20StreamingEnforcer not found in environment");
  }
  return {
    enforcer: ERC20StreamingEnforcer,
    terms,
    args: "0x00"
  };
};

// src/caveatBuilder/erc20TransferAmountBuilder.ts


var erc20TransferAmount = "erc20TransferAmount";
var erc20TransferAmountBuilder = (environment, config) => {
  const { tokenAddress, maxAmount } = config;
  if (!_viem.isAddress.call(void 0, tokenAddress, { strict: false })) {
    throw new Error("Invalid tokenAddress: must be a valid address");
  }
  if (maxAmount <= 0n) {
    throw new Error("Invalid maxAmount: must be a positive number");
  }
  const terms = _delegationcore.createERC20TransferAmountTerms.call(void 0, { tokenAddress, maxAmount });
  const {
    caveatEnforcers: { ERC20TransferAmountEnforcer }
  } = environment;
  if (!ERC20TransferAmountEnforcer) {
    throw new Error("ERC20TransferAmountEnforcer not found in environment");
  }
  return {
    enforcer: ERC20TransferAmountEnforcer,
    terms,
    args: "0x00"
  };
};

// src/caveatBuilder/erc721BalanceChangeBuilder.ts





var erc721BalanceChange = "erc721BalanceChange";
var erc721BalanceChangeBuilder = (environment, config) => {
  const { tokenAddress, recipient, amount, changeType } = config;
  if (!_viem.isAddress.call(void 0, tokenAddress, { strict: false })) {
    throw new Error("Invalid tokenAddress: must be a valid address");
  }
  if (!_viem.isAddress.call(void 0, recipient, { strict: false })) {
    throw new Error("Invalid recipient: must be a valid address");
  }
  if (amount <= 0n) {
    throw new Error("Invalid balance: must be a positive number");
  }
  if (changeType !== _delegationcore.BalanceChangeType.Increase && changeType !== _delegationcore.BalanceChangeType.Decrease) {
    throw new Error("Invalid changeType: must be either Increase or Decrease");
  }
  const terms = _delegationcore.createERC721BalanceChangeTerms.call(void 0, {
    tokenAddress,
    recipient,
    amount,
    changeType
  });
  const {
    caveatEnforcers: { ERC721BalanceChangeEnforcer }
  } = environment;
  if (!ERC721BalanceChangeEnforcer) {
    throw new Error("ERC721BalanceChangeEnforcer not found in environment");
  }
  return {
    enforcer: ERC721BalanceChangeEnforcer,
    terms,
    args: "0x00"
  };
};

// src/caveatBuilder/erc721TransferBuilder.ts


var erc721Transfer = "erc721Transfer";
var erc721TransferBuilder = (environment, config) => {
  const { tokenAddress, tokenId } = config;
  if (!_viem.isAddress.call(void 0, tokenAddress, { strict: false })) {
    throw new Error("Invalid tokenAddress: must be a valid address");
  }
  if (tokenId < 0n) {
    throw new Error("Invalid tokenId: must be a non-negative number");
  }
  const terms = _delegationcore.createERC721TransferTerms.call(void 0, { tokenAddress, tokenId });
  const {
    caveatEnforcers: { ERC721TransferEnforcer }
  } = environment;
  if (!ERC721TransferEnforcer) {
    throw new Error("ERC721TransferEnforcer not found in environment");
  }
  return {
    enforcer: ERC721TransferEnforcer,
    terms,
    args: "0x00"
  };
};

// src/caveatBuilder/exactCalldataBatchBuilder.ts


var exactCalldataBatch = "exactCalldataBatch";
var exactCalldataBatchBuilder = (environment, config) => {
  const { executions } = config;
  if (executions.length === 0) {
    throw new Error("Invalid executions: array cannot be empty");
  }
  for (const execution of executions) {
    if (!_viem.isAddress.call(void 0, execution.target, { strict: false })) {
      throw new Error("Invalid target: must be a valid address");
    }
    if (execution.value < 0n) {
      throw new Error("Invalid value: must be a non-negative number");
    }
    if (!execution.callData.startsWith("0x")) {
      throw new Error(
        "Invalid calldata: must be a hex string starting with 0x"
      );
    }
  }
  const terms = _delegationcore.createExactCalldataBatchTerms.call(void 0, { executions });
  const {
    caveatEnforcers: { ExactCalldataBatchEnforcer }
  } = environment;
  if (!ExactCalldataBatchEnforcer) {
    throw new Error("ExactCalldataBatchEnforcer not found in environment");
  }
  return {
    enforcer: ExactCalldataBatchEnforcer,
    terms,
    args: "0x00"
  };
};

// src/caveatBuilder/exactCalldataBuilder.ts

var exactCalldata = "exactCalldata";
var exactCalldataBuilder = (environment, config) => {
  const { calldata } = config;
  const terms = _delegationcore.createExactCalldataTerms.call(void 0, { calldata });
  const {
    caveatEnforcers: { ExactCalldataEnforcer }
  } = environment;
  if (!ExactCalldataEnforcer) {
    throw new Error("ExactCalldataEnforcer not found in environment");
  }
  return {
    enforcer: ExactCalldataEnforcer,
    terms,
    args: "0x00"
  };
};

// src/caveatBuilder/exactExecutionBatchBuilder.ts


var exactExecutionBatch = "exactExecutionBatch";
var exactExecutionBatchBuilder = (environment, config) => {
  const { executions } = config;
  if (executions.length === 0) {
    throw new Error("Invalid executions: array cannot be empty");
  }
  for (const execution of executions) {
    if (!_viem.isAddress.call(void 0, execution.target, { strict: false })) {
      throw new Error("Invalid target: must be a valid address");
    }
    if (execution.value < 0n) {
      throw new Error("Invalid value: must be a non-negative number");
    }
    if (!execution.callData.startsWith("0x")) {
      throw new Error(
        "Invalid calldata: must be a hex string starting with 0x"
      );
    }
  }
  const terms = _delegationcore.createExactExecutionBatchTerms.call(void 0, { executions });
  const {
    caveatEnforcers: { ExactExecutionBatchEnforcer }
  } = environment;
  if (!ExactExecutionBatchEnforcer) {
    throw new Error("ExactExecutionBatchEnforcer not found in environment");
  }
  return {
    enforcer: ExactExecutionBatchEnforcer,
    terms,
    args: "0x00"
  };
};

// src/caveatBuilder/exactExecutionBuilder.ts


var exactExecution = "exactExecution";
var exactExecutionBuilder = (environment, config) => {
  const { execution } = config;
  if (!_viem.isAddress.call(void 0, execution.target, { strict: false })) {
    throw new Error("Invalid target: must be a valid address");
  }
  if (execution.value < 0n) {
    throw new Error("Invalid value: must be a non-negative number");
  }
  if (!execution.callData.startsWith("0x")) {
    throw new Error("Invalid calldata: must be a hex string starting with 0x");
  }
  const terms = _delegationcore.createExactExecutionTerms.call(void 0, { execution });
  const {
    caveatEnforcers: { ExactExecutionEnforcer }
  } = environment;
  if (!ExactExecutionEnforcer) {
    throw new Error("ExactExecutionEnforcer not found in environment");
  }
  return {
    enforcer: ExactExecutionEnforcer,
    terms,
    args: "0x00"
  };
};

// src/caveatBuilder/idBuilder.ts


var id = "id";
var idBuilder = (environment, config) => {
  const { id: idValue } = config;
  let idBigInt;
  if (typeof idValue === "number") {
    if (!Number.isInteger(idValue)) {
      throw new Error("Invalid id: must be an integer");
    }
    idBigInt = BigInt(idValue);
  } else if (typeof idValue === "bigint") {
    idBigInt = idValue;
  } else {
    throw new Error("Invalid id: must be a bigint or number");
  }
  if (idBigInt < 0n) {
    throw new Error("Invalid id: must be a non-negative number");
  }
  if (idBigInt > _viem.maxUint256) {
    throw new Error("Invalid id: must be less than 2^256");
  }
  const terms = _delegationcore.createIdTerms.call(void 0, { id: idBigInt });
  const {
    caveatEnforcers: { IdEnforcer }
  } = environment;
  if (!IdEnforcer) {
    throw new Error("IdEnforcer not found in environment");
  }
  return {
    enforcer: IdEnforcer,
    terms,
    args: "0x00"
  };
};

// src/caveatBuilder/limitedCallsBuilder.ts

var limitedCalls = "limitedCalls";
var limitedCallsBuilder = (environment, config) => {
  const { limit } = config;
  if (!Number.isInteger(limit)) {
    throw new Error("Invalid limit: must be an integer");
  }
  if (limit <= 0) {
    throw new Error("Invalid limit: must be a positive integer");
  }
  const terms = _delegationcore.createLimitedCallsTerms.call(void 0, { limit });
  const {
    caveatEnforcers: { LimitedCallsEnforcer }
  } = environment;
  if (!LimitedCallsEnforcer) {
    throw new Error("LimitedCallsEnforcer not found in environment");
  }
  return {
    enforcer: LimitedCallsEnforcer,
    terms,
    args: "0x00"
  };
};

// src/caveatBuilder/multiTokenPeriodBuilder.ts


var multiTokenPeriod = "multiTokenPeriod";
var multiTokenPeriodBuilder = (environment, config) => {
  if (!_optionalChain([config, 'optionalAccess', _ => _.tokenConfigs]) || config.tokenConfigs.length === 0) {
    throw new Error(
      "MultiTokenPeriodBuilder: tokenConfigs array cannot be empty"
    );
  }
  config.tokenConfigs.forEach((tokenConfig) => {
    if (!_viem.isAddress.call(void 0, tokenConfig.token)) {
      throw new Error(`Invalid token address: ${String(tokenConfig.token)}`);
    }
    if (tokenConfig.periodAmount <= 0) {
      throw new Error("Invalid period amount: must be greater than 0");
    }
    if (tokenConfig.periodDuration <= 0) {
      throw new Error("Invalid period duration: must be greater than 0");
    }
  });
  const terms = _delegationcore.createMultiTokenPeriodTerms.call(void 0, {
    tokenConfigs: config.tokenConfigs
  });
  const {
    caveatEnforcers: { MultiTokenPeriodEnforcer }
  } = environment;
  if (!MultiTokenPeriodEnforcer) {
    throw new Error("MultiTokenPeriodEnforcer not found in environment");
  }
  return {
    enforcer: MultiTokenPeriodEnforcer,
    terms,
    args: "0x00"
  };
};

// src/caveatBuilder/nativeBalanceChangeBuilder.ts





var nativeBalanceChange = "nativeBalanceChange";
var nativeBalanceChangeBuilder = (environment, config) => {
  const { recipient, balance, changeType } = config;
  if (!_viem.isAddress.call(void 0, recipient)) {
    throw new Error("Invalid recipient: must be a valid Address");
  }
  if (balance <= 0n) {
    throw new Error("Invalid balance: must be a positive number");
  }
  if (changeType !== _delegationcore.BalanceChangeType.Increase && changeType !== _delegationcore.BalanceChangeType.Decrease) {
    throw new Error("Invalid changeType: must be either Increase or Decrease");
  }
  const terms = _delegationcore.createNativeBalanceChangeTerms.call(void 0, {
    recipient,
    balance,
    changeType
  });
  const {
    caveatEnforcers: { NativeBalanceChangeEnforcer }
  } = environment;
  if (!NativeBalanceChangeEnforcer) {
    throw new Error("NativeBalanceChangeEnforcer not found in environment");
  }
  return {
    enforcer: NativeBalanceChangeEnforcer,
    terms,
    args: "0x00"
  };
};

// src/caveatBuilder/nativeTokenPaymentBuilder.ts


var nativeTokenPayment = "nativeTokenPayment";
var nativeTokenPaymentBuilder = (environment, config) => {
  const { recipient, amount } = config;
  if (amount <= 0n) {
    throw new Error("Invalid amount: must be positive");
  }
  if (!_viem.isAddress.call(void 0, recipient)) {
    throw new Error("Invalid recipient: must be a valid address");
  }
  const terms = _delegationcore.createNativeTokenPaymentTerms.call(void 0, { recipient, amount });
  const {
    caveatEnforcers: { NativeTokenPaymentEnforcer }
  } = environment;
  if (!NativeTokenPaymentEnforcer) {
    throw new Error("NativeTokenPaymentEnforcer not found in environment");
  }
  return {
    enforcer: NativeTokenPaymentEnforcer,
    terms,
    args: "0x00"
  };
};

// src/caveatBuilder/nativeTokenPeriodTransferBuilder.ts

var nativeTokenPeriodTransfer = "nativeTokenPeriodTransfer";
var nativeTokenPeriodTransferBuilder = (environment, config) => {
  const { periodAmount, periodDuration, startDate } = config;
  const terms = _delegationcore.createNativeTokenPeriodTransferTerms.call(void 0, {
    periodAmount,
    periodDuration,
    startDate
  });
  const {
    caveatEnforcers: { NativeTokenPeriodTransferEnforcer }
  } = environment;
  if (!NativeTokenPeriodTransferEnforcer) {
    throw new Error(
      "NativeTokenPeriodTransferEnforcer not found in environment"
    );
  }
  return {
    enforcer: NativeTokenPeriodTransferEnforcer,
    terms,
    args: "0x00"
  };
};

// src/caveatBuilder/nativeTokenStreamingBuilder.ts

var nativeTokenStreaming = "nativeTokenStreaming";
var nativeTokenStreamingBuilder = (environment, config) => {
  const { initialAmount, maxAmount, amountPerSecond, startTime } = config;
  const terms = _delegationcore.createNativeTokenStreamingTerms.call(void 0, {
    initialAmount,
    maxAmount,
    amountPerSecond,
    startTime
  });
  const {
    caveatEnforcers: { NativeTokenStreamingEnforcer }
  } = environment;
  if (!NativeTokenStreamingEnforcer) {
    throw new Error("NativeTokenStreamingEnforcer not found in environment");
  }
  return {
    enforcer: NativeTokenStreamingEnforcer,
    terms,
    args: "0x00"
  };
};

// src/caveatBuilder/nativeTokenTransferAmountBuilder.ts

var nativeTokenTransferAmount = "nativeTokenTransferAmount";
var nativeTokenTransferAmountBuilder = (environment, config) => {
  const { maxAmount } = config;
  if (maxAmount < 0n) {
    throw new Error("Invalid maxAmount: must be zero or positive");
  }
  const terms = _delegationcore.createNativeTokenTransferAmountTerms.call(void 0, { maxAmount });
  const {
    caveatEnforcers: { NativeTokenTransferAmountEnforcer }
  } = environment;
  if (!NativeTokenTransferAmountEnforcer) {
    throw new Error(
      "NativeTokenTransferAmountEnforcer not found in environment"
    );
  }
  return {
    enforcer: NativeTokenTransferAmountEnforcer,
    terms,
    args: "0x00"
  };
};

// src/caveatBuilder/nonceBuilder.ts

var nonce = "nonce";
var nonceBuilder = (environment, config) => {
  const { nonce: nonceValue } = config;
  const terms = _delegationcore.createNonceTerms.call(void 0, { nonce: nonceValue });
  const {
    caveatEnforcers: { NonceEnforcer }
  } = environment;
  if (!NonceEnforcer) {
    throw new Error("NonceEnforcer not found in environment");
  }
  return {
    enforcer: NonceEnforcer,
    terms,
    args: "0x00"
  };
};

// src/caveatBuilder/ownershipTransferBuilder.ts


var ownershipTransfer = "ownershipTransfer";
var ownershipTransferBuilder = (environment, config) => {
  const { contractAddress } = config;
  if (!_viem.isAddress.call(void 0, contractAddress, { strict: false })) {
    throw new Error("Invalid contractAddress: must be a valid address");
  }
  const terms = _delegationcore.createOwnershipTransferTerms.call(void 0, { contractAddress });
  const {
    caveatEnforcers: { OwnershipTransferEnforcer }
  } = environment;
  if (!OwnershipTransferEnforcer) {
    throw new Error("OwnershipTransferEnforcer not found in environment");
  }
  return {
    enforcer: OwnershipTransferEnforcer,
    terms,
    args: "0x00"
  };
};

// src/caveatBuilder/redeemerBuilder.ts


var redeemer = "redeemer";
var redeemerBuilder = (environment, config) => {
  const { redeemers } = config;
  if (redeemers.length === 0) {
    throw new Error(
      "Invalid redeemers: must specify at least one redeemer address"
    );
  }
  for (const redeemerAddress of redeemers) {
    if (!_viem.isAddress.call(void 0, redeemerAddress)) {
      throw new Error("Invalid redeemers: must be a valid address");
    }
  }
  const terms = _delegationcore.createRedeemerTerms.call(void 0, { redeemers });
  const {
    caveatEnforcers: { RedeemerEnforcer }
  } = environment;
  if (!RedeemerEnforcer) {
    throw new Error("RedeemerEnforcer not found in environment");
  }
  return {
    enforcer: RedeemerEnforcer,
    terms,
    args: "0x00"
  };
};

// src/caveatBuilder/specificActionERC20TransferBatchBuilder.ts


var specificActionERC20TransferBatch = "specificActionERC20TransferBatch";
var specificActionERC20TransferBatchBuilder = (environment, config) => {
  const { tokenAddress, recipient, amount, target, calldata } = config;
  if (!_viem.isAddress.call(void 0, tokenAddress, { strict: false })) {
    throw new Error("Invalid tokenAddress: must be a valid address");
  }
  if (!_viem.isAddress.call(void 0, recipient, { strict: false })) {
    throw new Error("Invalid recipient: must be a valid address");
  }
  if (!_viem.isAddress.call(void 0, target, { strict: false })) {
    throw new Error("Invalid target: must be a valid address");
  }
  if (amount <= 0n) {
    throw new Error("Invalid amount: must be a positive number");
  }
  const terms = _delegationcore.createSpecificActionERC20TransferBatchTerms.call(void 0, {
    tokenAddress,
    recipient,
    amount,
    target,
    calldata
  });
  const {
    caveatEnforcers: { SpecificActionERC20TransferBatchEnforcer }
  } = environment;
  if (!SpecificActionERC20TransferBatchEnforcer) {
    throw new Error(
      "SpecificActionERC20TransferBatchEnforcer not found in environment"
    );
  }
  return {
    enforcer: SpecificActionERC20TransferBatchEnforcer,
    terms,
    args: "0x00"
  };
};

// src/caveatBuilder/timestampBuilder.ts

var timestamp = "timestamp";
var timestampBuilder = (environment, config) => {
  const { afterThreshold, beforeThreshold } = config;
  const terms = _delegationcore.createTimestampTerms.call(void 0, {
    afterThreshold,
    beforeThreshold
  });
  const {
    caveatEnforcers: { TimestampEnforcer }
  } = environment;
  if (!TimestampEnforcer) {
    throw new Error("TimestampEnforcer not found in environment");
  }
  return {
    enforcer: TimestampEnforcer,
    terms,
    args: "0x00"
  };
};

// src/caveatBuilder/valueLteBuilder.ts

var valueLte = "valueLte";
var valueLteBuilder = (environment, config) => {
  const { maxValue } = config;
  const terms = _delegationcore.createValueLteTerms.call(void 0, { maxValue });
  const {
    caveatEnforcers: { ValueLteEnforcer }
  } = environment;
  if (!ValueLteEnforcer) {
    throw new Error("ValueLteEnforcer not found in environment");
  }
  return {
    enforcer: ValueLteEnforcer,
    terms,
    args: "0x00"
  };
};

// src/caveatBuilder/coreCaveatBuilder.ts
var createCaveatBuilder = (environment, config) => {
  const caveatBuilder = new CaveatBuilder(environment, config).extend(allowedMethods, allowedMethodsBuilder).extend(allowedTargets, allowedTargetsBuilder).extend(approvalRevocation, approvalRevocationBuilder).extend(deployed, deployedBuilder).extend(allowedCalldata, allowedCalldataBuilder).extend(erc20BalanceChange, erc20BalanceChangeBuilder).extend(erc721BalanceChange, erc721BalanceChangeBuilder).extend(erc1155BalanceChange, erc1155BalanceChangeBuilder).extend(valueLte, valueLteBuilder).extend(limitedCalls, limitedCallsBuilder).extend(id, idBuilder).extend(nonce, nonceBuilder).extend(timestamp, timestampBuilder).extend(blockNumber, blockNumberBuilder).extend(erc20TransferAmount, erc20TransferAmountBuilder).extend(erc20Streaming, erc20StreamingBuilder).extend(nativeTokenStreaming, nativeTokenStreamingBuilder).extend(erc721Transfer, erc721TransferBuilder).extend(nativeTokenTransferAmount, nativeTokenTransferAmountBuilder).extend(nativeBalanceChange, nativeBalanceChangeBuilder).extend(redeemer, redeemerBuilder).extend(nativeTokenPayment, nativeTokenPaymentBuilder).extend(argsEqualityCheck, argsEqualityCheckBuilder).extend(
    specificActionERC20TransferBatch,
    specificActionERC20TransferBatchBuilder
  ).extend(erc20PeriodTransfer, erc20PeriodTransferBuilder).extend(nativeTokenPeriodTransfer, nativeTokenPeriodTransferBuilder).extend(exactCalldataBatch, exactCalldataBatchBuilder).extend(exactCalldata, exactCalldataBuilder).extend(exactExecution, exactExecutionBuilder).extend(exactExecutionBatch, exactExecutionBatchBuilder).extend(multiTokenPeriod, multiTokenPeriodBuilder).extend(ownershipTransfer, ownershipTransferBuilder);
  return caveatBuilder;
};

// src/caveatBuilder/scope/erc20PeriodicScope.ts
function createErc20PeriodicCaveatBuilder(environment, config) {
  return createCaveatBuilder(environment).addCaveat("valueLte", {
    maxValue: 0n
  }).addCaveat("erc20PeriodTransfer", {
    tokenAddress: config.tokenAddress,
    periodAmount: config.periodAmount,
    periodDuration: config.periodDuration,
    startDate: config.startDate
  });
}

// src/caveatBuilder/scope/erc20StreamingScope.ts
function createErc20StreamingCaveatBuilder(environment, config) {
  return createCaveatBuilder(environment).addCaveat("valueLte", {
    maxValue: 0n
  }).addCaveat("erc20Streaming", {
    tokenAddress: config.tokenAddress,
    initialAmount: config.initialAmount,
    maxAmount: config.maxAmount,
    amountPerSecond: config.amountPerSecond,
    startTime: config.startTime
  });
}

// src/caveatBuilder/scope/erc20TransferScope.ts
function createErc20TransferCaveatBuilder(environment, config) {
  return createCaveatBuilder(environment).addCaveat("valueLte", {
    maxValue: 0n
  }).addCaveat("erc20TransferAmount", {
    tokenAddress: config.tokenAddress,
    maxAmount: config.maxAmount
  });
}

// src/utils.ts

var hasProperties = (object, properties) => {
  return properties.every(
    (prop) => prop in object && object[prop] !== void 0
  );
};
function isDefined(value) {
  return value !== void 0 && value !== null;
}
function assertIsDefined(value, parameterName) {
  if (!isDefined(value)) {
    throw new Error(
      `Invalid parameters: ${_nullishCoalesce(parameterName, () => ( "value"))} is required`
    );
  }
}
function toHexOrThrow(value, parameterName) {
  assertIsDefined(value, parameterName);
  if (typeof value === "string") {
    if (!_viem.isHex.call(void 0, value)) {
      throw new Error(
        `Invalid parameters: ${_nullishCoalesce(parameterName, () => ( "value"))} is not a valid hex value`
      );
    }
    return value;
  }
  return _viem.toHex.call(void 0, value);
}

// src/caveatBuilder/scope/erc721Scope.ts
var isErc721TransferConfig = (config) => {
  return hasProperties(
    config,
    ["tokenAddress", "tokenId"]
  );
};
function createErc721CaveatBuilder(environment, config) {
  if (!isErc721TransferConfig(config)) {
    throw new Error("Invalid ERC721 configuration");
  }
  const caveatBuilder = createCaveatBuilder(environment).addCaveat(
    "erc721Transfer",
    config
  );
  return caveatBuilder;
}

// src/caveatBuilder/scope/functionCallScope.ts
var isFunctionCallConfig = (config) => {
  return hasProperties(config, ["targets", "selectors"]);
};
function createFunctionCallCaveatBuilder(environment, config) {
  const { targets, selectors, allowedCalldata: allowedCalldata2, exactCalldata: exactCalldata2 } = config;
  if (!isFunctionCallConfig(config)) {
    throw new Error("Invalid Function Call configuration");
  }
  if (allowedCalldata2 && allowedCalldata2.length > 0 && exactCalldata2) {
    throw new Error(
      "Cannot specify both allowedCalldata and exactCalldata. Please use only one calldata restriction type."
    );
  }
  const valueLteConfig = _nullishCoalesce(config.valueLte, () => ( { maxValue: 0n }));
  const caveatBuilder = createCaveatBuilder(environment).addCaveat("allowedTargets", { targets }).addCaveat("allowedMethods", { selectors }).addCaveat("valueLte", valueLteConfig);
  if (allowedCalldata2 && allowedCalldata2.length > 0) {
    allowedCalldata2.forEach((calldataConfig) => {
      caveatBuilder.addCaveat("allowedCalldata", calldataConfig);
    });
  } else if (exactCalldata2) {
    caveatBuilder.addCaveat("exactCalldata", exactCalldata2);
  }
  return caveatBuilder;
}

// src/caveatBuilder/scope/nativeTokenPeriodicScope.ts
function createNativeTokenPeriodicCaveatBuilder(environment, config) {
  const {
    periodAmount,
    periodDuration,
    startDate,
    allowedCalldata: allowedCalldata2,
    exactCalldata: exactCalldata2
  } = config;
  if (allowedCalldata2 && allowedCalldata2.length > 0 && exactCalldata2) {
    throw new Error(
      "Cannot specify both allowedCalldata and exactCalldata. Please use only one calldata restriction type."
    );
  }
  const caveatBuilder = createCaveatBuilder(environment);
  if (allowedCalldata2 && allowedCalldata2.length > 0) {
    allowedCalldata2.forEach((calldataConfig) => {
      caveatBuilder.addCaveat("allowedCalldata", calldataConfig);
    });
  } else if (exactCalldata2) {
    caveatBuilder.addCaveat("exactCalldata", exactCalldata2);
  } else {
    caveatBuilder.addCaveat("exactCalldata", {
      calldata: "0x"
    });
  }
  caveatBuilder.addCaveat("nativeTokenPeriodTransfer", {
    periodAmount,
    periodDuration,
    startDate
  });
  return caveatBuilder;
}

// src/caveatBuilder/scope/nativeTokenStreamingScope.ts
function createNativeTokenStreamingCaveatBuilder(environment, config) {
  const {
    initialAmount,
    maxAmount,
    amountPerSecond,
    startTime,
    allowedCalldata: allowedCalldata2,
    exactCalldata: exactCalldata2
  } = config;
  if (allowedCalldata2 && allowedCalldata2.length > 0 && exactCalldata2) {
    throw new Error(
      "Cannot specify both allowedCalldata and exactCalldata. Please use only one calldata restriction type."
    );
  }
  const caveatBuilder = createCaveatBuilder(environment);
  if (allowedCalldata2 && allowedCalldata2.length > 0) {
    allowedCalldata2.forEach((calldataConfig) => {
      caveatBuilder.addCaveat("allowedCalldata", calldataConfig);
    });
  } else if (exactCalldata2) {
    caveatBuilder.addCaveat("exactCalldata", exactCalldata2);
  } else {
    caveatBuilder.addCaveat("exactCalldata", {
      calldata: "0x"
    });
  }
  caveatBuilder.addCaveat("nativeTokenStreaming", {
    initialAmount,
    maxAmount,
    amountPerSecond,
    startTime
  });
  return caveatBuilder;
}

// src/caveatBuilder/scope/nativeTokenTransferScope.ts
function createNativeTokenTransferCaveatBuilder(environment, config) {
  const { maxAmount, allowedCalldata: allowedCalldata2, exactCalldata: exactCalldata2 } = config;
  if (allowedCalldata2 && allowedCalldata2.length > 0 && exactCalldata2) {
    throw new Error(
      "Cannot specify both allowedCalldata and exactCalldata. Please use only one calldata restriction type."
    );
  }
  const caveatBuilder = createCaveatBuilder(environment);
  if (allowedCalldata2 && allowedCalldata2.length > 0) {
    allowedCalldata2.forEach((calldataConfig) => {
      caveatBuilder.addCaveat("allowedCalldata", calldataConfig);
    });
  } else if (exactCalldata2) {
    caveatBuilder.addCaveat("exactCalldata", exactCalldata2);
  } else {
    caveatBuilder.addCaveat("exactCalldata", {
      calldata: "0x"
    });
  }
  caveatBuilder.addCaveat("nativeTokenTransferAmount", {
    maxAmount
  });
  return caveatBuilder;
}

// src/caveatBuilder/scope/ownershipScope.ts
var isOwnershipTransferConfig = (config) => {
  return hasProperties(
    config,
    ["contractAddress"]
  );
};
function createOwnershipCaveatBuilder(environment, config) {
  if (!isOwnershipTransferConfig(config)) {
    throw new Error("Invalid ownership transfer configuration");
  }
  const caveatBuilder = createCaveatBuilder(environment).addCaveat(
    "ownershipTransfer",
    config
  );
  return caveatBuilder;
}

// src/caveatBuilder/scope/index.ts
var normalizeScopeConfig = (config) => {
  return {
    ...config,
    type: config.type
  };
};
var createCaveatBuilderFromScope = (environment, scopeConfig) => {
  const normalizedScopeConfig = normalizeScopeConfig(scopeConfig);
  switch (normalizedScopeConfig.type) {
    case "erc20TransferAmount" /* Erc20TransferAmount */:
      return createErc20TransferCaveatBuilder(
        environment,
        normalizedScopeConfig
      );
    case "erc20Streaming" /* Erc20Streaming */:
      return createErc20StreamingCaveatBuilder(
        environment,
        normalizedScopeConfig
      );
    case "erc20PeriodTransfer" /* Erc20PeriodTransfer */:
      return createErc20PeriodicCaveatBuilder(
        environment,
        normalizedScopeConfig
      );
    case "nativeTokenTransferAmount" /* NativeTokenTransferAmount */:
      return createNativeTokenTransferCaveatBuilder(
        environment,
        normalizedScopeConfig
      );
    case "nativeTokenStreaming" /* NativeTokenStreaming */:
      return createNativeTokenStreamingCaveatBuilder(
        environment,
        normalizedScopeConfig
      );
    case "nativeTokenPeriodTransfer" /* NativeTokenPeriodTransfer */:
      return createNativeTokenPeriodicCaveatBuilder(
        environment,
        normalizedScopeConfig
      );
    case "erc721Transfer" /* Erc721Transfer */:
      return createErc721CaveatBuilder(environment, normalizedScopeConfig);
    case "ownershipTransfer" /* OwnershipTransfer */:
      return createOwnershipCaveatBuilder(environment, normalizedScopeConfig);
    case "functionCall" /* FunctionCall */:
      return createFunctionCallCaveatBuilder(
        environment,
        normalizedScopeConfig
      );
    default:
      const exhaustivenessCheck = normalizedScopeConfig;
      throw new Error(
        `Invalid scope type: ${exhaustivenessCheck.type}`
      );
  }
};

// src/caveatBuilder/resolveCaveats.ts
var resolveCaveats = ({
  environment,
  scope,
  caveats,
  isScopeOptional
}) => {
  if (!scope && !isScopeOptional) {
    throw new Error("Scope is required");
  }
  const scopeCaveatBuilder = scope ? createCaveatBuilderFromScope(environment, scope) : createCaveatBuilder(environment, {
    allowInsecureUnrestrictedDelegation: true
  });
  if (caveats) {
    if ("build" in caveats && typeof caveats.build === "function") {
      caveats.build().forEach((caveat) => {
        scopeCaveatBuilder.addCaveat(caveat);
      });
    } else if (Array.isArray(caveats)) {
      caveats.forEach((caveat) => {
        try {
          if ("type" in caveat) {
            const { type, ...config } = caveat;
            scopeCaveatBuilder.addCaveat(type, config);
          } else {
            scopeCaveatBuilder.addCaveat(caveat);
          }
        } catch (error) {
          throw new Error(`Invalid caveat: ${error.message}`);
        }
      });
    }
  }
  return scopeCaveatBuilder.build();
};

// src/caveats.ts








































var CAVEAT_ABI_TYPE_COMPONENTS = [
  { type: "address", name: "enforcer" },
  { type: "bytes", name: "terms" },
  { type: "bytes", name: "args" }
];
var CAVEAT_TYPEHASH = _viem.keccak256.call(void 0, 
  _viem.toHex.call(void 0, "Caveat(address enforcer,bytes terms)")
);
var createCaveat = (enforcer, terms, args = "0x00") => {
  return {
    enforcer,
    terms,
    args
  };
};
var decodeCaveat = ({
  caveat: { enforcer, terms },
  environment: { caveatEnforcers }
}) => {
  switch (enforcer.toLowerCase()) {
    case _optionalChain([caveatEnforcers, 'access', _2 => _2.AllowedCalldataEnforcer, 'optionalAccess', _3 => _3.toLowerCase, 'call', _4 => _4()]):
      return {
        type: "allowedCalldata" /* AllowedCalldata */,
        ..._delegationcore.decodeAllowedCalldataTerms.call(void 0, terms)
      };
    case _optionalChain([caveatEnforcers, 'access', _5 => _5.AllowedMethodsEnforcer, 'optionalAccess', _6 => _6.toLowerCase, 'call', _7 => _7()]):
      return {
        type: "allowedMethods" /* AllowedMethods */,
        ..._delegationcore.decodeAllowedMethodsTerms.call(void 0, terms)
      };
    case _optionalChain([caveatEnforcers, 'access', _8 => _8.ApprovalRevocationEnforcer, 'optionalAccess', _9 => _9.toLowerCase, 'call', _10 => _10()]):
      return {
        type: "approvalRevocation" /* ApprovalRevocation */,
        ..._delegationcore.decodeApprovalRevocationTerms.call(void 0, terms)
      };
    case _optionalChain([caveatEnforcers, 'access', _11 => _11.AllowedTargetsEnforcer, 'optionalAccess', _12 => _12.toLowerCase, 'call', _13 => _13()]):
      return {
        type: "allowedTargets" /* AllowedTargets */,
        ..._delegationcore.decodeAllowedTargetsTerms.call(void 0, terms)
      };
    case _optionalChain([caveatEnforcers, 'access', _14 => _14.ArgsEqualityCheckEnforcer, 'optionalAccess', _15 => _15.toLowerCase, 'call', _16 => _16()]):
      return {
        type: "argsEqualityCheck" /* ArgsEqualityCheck */,
        ..._delegationcore.decodeArgsEqualityCheckTerms.call(void 0, terms)
      };
    case _optionalChain([caveatEnforcers, 'access', _17 => _17.BlockNumberEnforcer, 'optionalAccess', _18 => _18.toLowerCase, 'call', _19 => _19()]):
      return { type: "blockNumber" /* BlockNumber */, ..._delegationcore.decodeBlockNumberTerms.call(void 0, terms) };
    case _optionalChain([caveatEnforcers, 'access', _20 => _20.DeployedEnforcer, 'optionalAccess', _21 => _21.toLowerCase, 'call', _22 => _22()]):
      return { type: "deployed" /* Deployed */, ..._delegationcore.decodeDeployedTerms.call(void 0, terms) };
    case _optionalChain([caveatEnforcers, 'access', _23 => _23.ERC20BalanceChangeEnforcer, 'optionalAccess', _24 => _24.toLowerCase, 'call', _25 => _25()]):
      return {
        type: "erc20BalanceChange" /* Erc20BalanceChange */,
        ..._delegationcore.decodeERC20BalanceChangeTerms.call(void 0, terms)
      };
    case _optionalChain([caveatEnforcers, 'access', _26 => _26.ERC20TransferAmountEnforcer, 'optionalAccess', _27 => _27.toLowerCase, 'call', _28 => _28()]):
      return {
        type: "erc20TransferAmount" /* Erc20TransferAmount */,
        ..._delegationcore.decodeERC20TransferAmountTerms.call(void 0, terms)
      };
    case _optionalChain([caveatEnforcers, 'access', _29 => _29.ERC20StreamingEnforcer, 'optionalAccess', _30 => _30.toLowerCase, 'call', _31 => _31()]):
      return {
        type: "erc20Streaming" /* Erc20Streaming */,
        ..._delegationcore.decodeERC20StreamingTerms.call(void 0, terms)
      };
    case _optionalChain([caveatEnforcers, 'access', _32 => _32.ERC721BalanceChangeEnforcer, 'optionalAccess', _33 => _33.toLowerCase, 'call', _34 => _34()]):
      return {
        type: "erc721BalanceChange" /* Erc721BalanceChange */,
        ..._delegationcore.decodeERC721BalanceChangeTerms.call(void 0, terms)
      };
    case _optionalChain([caveatEnforcers, 'access', _35 => _35.ERC721TransferEnforcer, 'optionalAccess', _36 => _36.toLowerCase, 'call', _37 => _37()]):
      return {
        type: "erc721Transfer" /* Erc721Transfer */,
        ..._delegationcore.decodeERC721TransferTerms.call(void 0, terms)
      };
    case _optionalChain([caveatEnforcers, 'access', _38 => _38.ERC1155BalanceChangeEnforcer, 'optionalAccess', _39 => _39.toLowerCase, 'call', _40 => _40()]):
      return {
        type: "erc1155BalanceChange" /* Erc1155BalanceChange */,
        ..._delegationcore.decodeERC1155BalanceChangeTerms.call(void 0, terms)
      };
    case _optionalChain([caveatEnforcers, 'access', _41 => _41.IdEnforcer, 'optionalAccess', _42 => _42.toLowerCase, 'call', _43 => _43()]):
      return { type: "id" /* Id */, ..._delegationcore.decodeIdTerms.call(void 0, terms) };
    case _optionalChain([caveatEnforcers, 'access', _44 => _44.LimitedCallsEnforcer, 'optionalAccess', _45 => _45.toLowerCase, 'call', _46 => _46()]):
      return {
        type: "limitedCalls" /* LimitedCalls */,
        ..._delegationcore.decodeLimitedCallsTerms.call(void 0, terms)
      };
    case _optionalChain([caveatEnforcers, 'access', _47 => _47.NonceEnforcer, 'optionalAccess', _48 => _48.toLowerCase, 'call', _49 => _49()]):
      return { type: "nonce" /* Nonce */, ..._delegationcore.decodeNonceTerms.call(void 0, terms) };
    case _optionalChain([caveatEnforcers, 'access', _50 => _50.TimestampEnforcer, 'optionalAccess', _51 => _51.toLowerCase, 'call', _52 => _52()]):
      return { type: "timestamp" /* Timestamp */, ..._delegationcore.decodeTimestampTerms.call(void 0, terms) };
    case _optionalChain([caveatEnforcers, 'access', _53 => _53.ValueLteEnforcer, 'optionalAccess', _54 => _54.toLowerCase, 'call', _55 => _55()]):
      return { type: "valueLte" /* ValueLte */, ..._delegationcore.decodeValueLteTerms.call(void 0, terms) };
    case _optionalChain([caveatEnforcers, 'access', _56 => _56.NativeTokenTransferAmountEnforcer, 'optionalAccess', _57 => _57.toLowerCase, 'call', _58 => _58()]):
      return {
        type: "nativeTokenTransferAmount" /* NativeTokenTransferAmount */,
        ..._delegationcore.decodeNativeTokenTransferAmountTerms.call(void 0, terms)
      };
    case _optionalChain([caveatEnforcers, 'access', _59 => _59.NativeBalanceChangeEnforcer, 'optionalAccess', _60 => _60.toLowerCase, 'call', _61 => _61()]):
      return {
        type: "nativeBalanceChange" /* NativeBalanceChange */,
        ..._delegationcore.decodeNativeBalanceChangeTerms.call(void 0, terms)
      };
    case _optionalChain([caveatEnforcers, 'access', _62 => _62.NativeTokenStreamingEnforcer, 'optionalAccess', _63 => _63.toLowerCase, 'call', _64 => _64()]):
      return {
        type: "nativeTokenStreaming" /* NativeTokenStreaming */,
        ..._delegationcore.decodeNativeTokenStreamingTerms.call(void 0, terms)
      };
    case _optionalChain([caveatEnforcers, 'access', _65 => _65.NativeTokenPaymentEnforcer, 'optionalAccess', _66 => _66.toLowerCase, 'call', _67 => _67()]):
      return {
        type: "nativeTokenPayment" /* NativeTokenPayment */,
        ..._delegationcore.decodeNativeTokenPaymentTerms.call(void 0, terms)
      };
    case _optionalChain([caveatEnforcers, 'access', _68 => _68.RedeemerEnforcer, 'optionalAccess', _69 => _69.toLowerCase, 'call', _70 => _70()]):
      return { type: "redeemer" /* Redeemer */, ..._delegationcore.decodeRedeemerTerms.call(void 0, terms) };
    case _optionalChain([caveatEnforcers, 'access', _71 => _71.SpecificActionERC20TransferBatchEnforcer, 'optionalAccess', _72 => _72.toLowerCase, 'call', _73 => _73()]):
      return {
        type: "specificActionERC20TransferBatch" /* SpecificActionERC20TransferBatch */,
        ..._delegationcore.decodeSpecificActionERC20TransferBatchTerms.call(void 0, terms)
      };
    case _optionalChain([caveatEnforcers, 'access', _74 => _74.ERC20PeriodTransferEnforcer, 'optionalAccess', _75 => _75.toLowerCase, 'call', _76 => _76()]):
      return {
        type: "erc20PeriodTransfer" /* Erc20PeriodTransfer */,
        ..._delegationcore.decodeERC20TokenPeriodTransferTerms.call(void 0, terms)
      };
    case _optionalChain([caveatEnforcers, 'access', _77 => _77.NativeTokenPeriodTransferEnforcer, 'optionalAccess', _78 => _78.toLowerCase, 'call', _79 => _79()]):
      return {
        type: "nativeTokenPeriodTransfer" /* NativeTokenPeriodTransfer */,
        ..._delegationcore.decodeNativeTokenPeriodTransferTerms.call(void 0, terms)
      };
    case _optionalChain([caveatEnforcers, 'access', _80 => _80.ExactCalldataBatchEnforcer, 'optionalAccess', _81 => _81.toLowerCase, 'call', _82 => _82()]):
      return {
        type: "exactCalldataBatch" /* ExactCalldataBatch */,
        ..._delegationcore.decodeExactCalldataBatchTerms.call(void 0, terms)
      };
    case _optionalChain([caveatEnforcers, 'access', _83 => _83.ExactCalldataEnforcer, 'optionalAccess', _84 => _84.toLowerCase, 'call', _85 => _85()]):
      return {
        type: "exactCalldata" /* ExactCalldata */,
        ..._delegationcore.decodeExactCalldataTerms.call(void 0, terms)
      };
    case _optionalChain([caveatEnforcers, 'access', _86 => _86.ExactExecutionEnforcer, 'optionalAccess', _87 => _87.toLowerCase, 'call', _88 => _88()]):
      return {
        type: "exactExecution" /* ExactExecution */,
        ..._delegationcore.decodeExactExecutionTerms.call(void 0, terms)
      };
    case _optionalChain([caveatEnforcers, 'access', _89 => _89.ExactExecutionBatchEnforcer, 'optionalAccess', _90 => _90.toLowerCase, 'call', _91 => _91()]):
      return {
        type: "exactExecutionBatch" /* ExactExecutionBatch */,
        ..._delegationcore.decodeExactExecutionBatchTerms.call(void 0, terms)
      };
    case _optionalChain([caveatEnforcers, 'access', _92 => _92.MultiTokenPeriodEnforcer, 'optionalAccess', _93 => _93.toLowerCase, 'call', _94 => _94()]):
      return {
        type: "multiTokenPeriod" /* MultiTokenPeriod */,
        ..._delegationcore.decodeMultiTokenPeriodTerms.call(void 0, terms)
      };
    case _optionalChain([caveatEnforcers, 'access', _95 => _95.OwnershipTransferEnforcer, 'optionalAccess', _96 => _96.toLowerCase, 'call', _97 => _97()]):
      return {
        type: "ownershipTransfer" /* OwnershipTransfer */,
        ..._delegationcore.decodeOwnershipTransferTerms.call(void 0, terms)
      };
    default:
      throw new Error(`Unknown enforcer address: ${enforcer}`);
  }
};

// src/delegation.ts












var _accounts = require('viem/accounts');

// src/analytics/index.ts
var _openapifetch = require('openapi-fetch'); var _openapifetch2 = _interopRequireDefault(_openapifetch);

// src/analytics/environment.ts
var session;
function createAnonId() {
  if (typeof _optionalChain([globalThis, 'access', _98 => _98.crypto, 'optionalAccess', _99 => _99.randomUUID]) === "function") {
    return globalThis.crypto.randomUUID();
  }
  return `00000000-0000-4000-8000-${Math.random().toString(16).slice(2, 14).padEnd(12, "0")}`;
}
function inferPlatform() {
  if (typeof globalThis === "undefined" || !("window" in globalThis)) {
    return "nodejs";
  }
  const nav = globalThis.navigator;
  const ua = typeof _optionalChain([nav, 'optionalAccess', _100 => _100.userAgent]) === "string" ? nav.userAgent : "";
  if (/Mobile|Android|iPhone|iPod|webOS|BlackBerry|IEMobile|Opera Mini/iu.test(ua)) {
    return "web-mobile";
  }
  return "web-desktop";
}
function inferDomain() {
  if (typeof globalThis !== "undefined" && "window" in globalThis) {
    const win = globalThis;
    const hostname = _optionalChain([win, 'access', _101 => _101.location, 'optionalAccess', _102 => _102.hostname]);
    if (typeof hostname === "string" && hostname.length > 0) {
      return { domain: hostname };
    }
  }
  return {};
}
function getInitializationContext(params) {
  if (!session) {
    session = {
      sdk_version: params.sdk_version,
      anon_id: _nullishCoalesce(params.anon_id, () => ( createAnonId())),
      platform: inferPlatform(),
      ...inferDomain()
    };
    return { ...session };
  }
  session = {
    ...session,
    sdk_version: params.sdk_version
  };
  return { ...session };
}
function getSessionBaseProperties() {
  if (!session) {
    throw new Error(
      "Smart Accounts Kit analytics: call getInitializationContext({ sdk_version }) at SDK startup before recording events."
    );
  }
  return { ...session };
}
function mergeSessionProperties(partial) {
  if (!session) {
    throw new Error(
      "Smart Accounts Kit analytics: call getInitializationContext before mergeSessionProperties."
    );
  }
  session = { ...session, ...partial };
}

// src/analytics/sender.ts
var Sender = class {
  #sendFn;
  #batchSize;
  #baseTimeoutMs;
  #maxFailureCount;
  #maxTimeoutMs;
  #isDisabled = false;
  #batch = [];
  #failureCount = 0;
  #timeoutId = null;
  #isSending = false;
  constructor(options) {
    this.#batchSize = options.batchSize;
    this.#baseTimeoutMs = options.baseTimeoutMs;
    this.#maxFailureCount = options.maxFailureCount;
    this.#sendFn = options.sendFn;
    this.#maxTimeoutMs = options.maxTimeoutMs;
  }
  enqueue(item) {
    if (this.#isDisabled) {
      return;
    }
    this.#batch.push(item);
    this.#schedule();
  }
  #schedule() {
    if (this.#isDisabled) {
      return;
    }
    if (this.#batch.length > 0 && !this.#timeoutId) {
      const { isRetry, timeoutMs } = this.#getTimeout();
      const timeout = setTimeout(() => {
        this.#timeoutId = null;
        this.#flush();
      }, timeoutMs);
      if (isRetry && timeout.unref) {
        timeout.unref();
      }
      this.#timeoutId = timeout;
    }
  }
  async #flush() {
    if (this.#isDisabled || this.#isSending || this.#batch.length === 0) {
      return;
    }
    this.#isSending = true;
    const current = this.#batch.slice(0, this.#batchSize);
    this.#batch = this.#batch.slice(this.#batchSize);
    try {
      await this.#sendFn(current);
      this.#failureCount = 0;
    } catch (e) {
      this.#failureCount += 1;
      if (this.#failureCount >= this.#maxFailureCount) {
        this.#isDisabled = true;
        this.#batch = [];
        if (this.#timeoutId !== null) {
          clearTimeout(this.#timeoutId);
          this.#timeoutId = null;
        }
      } else {
        this.#batch = [...current, ...this.#batch];
      }
    } finally {
      this.#isSending = false;
      this.#schedule();
    }
  }
  #getTimeout() {
    const isRetry = this.#failureCount > 0;
    const timeoutMs = Math.min(
      this.#baseTimeoutMs * 2 ** this.#failureCount,
      this.#maxTimeoutMs
    );
    return { isRetry, timeoutMs };
  }
};

// package.json
var version = "1.6.0";

// src/analytics/index.ts
function isCompleteBase(value) {
  return typeof value.sdk_version === "string" && value.sdk_version.length > 0 && typeof value.anon_id === "string" && value.anon_id.length > 0 && typeof value.platform === "string" && value.platform.length > 0;
}
function normalise(batch) {
  const walk = (value) => {
    if (typeof value === "bigint") {
      return `${value}n`;
    }
    if (value === null || typeof value !== "object") {
      return value;
    }
    if (Array.isArray(value)) {
      return value.map((item) => walk(item));
    }
    const result = {};
    for (const [key, child] of Object.entries(
      value
    )) {
      result[key] = walk(child);
    }
    return result;
  };
  return walk(batch);
}
var METAMASK_ANALYTICS_ENDPOINT = "https://mm-sdk-analytics.api.cx.metamask.io/";
var Analytics = (_class = class {
  __init() {this.enabled = false}
  
  constructor(baseUrl) {;_class.prototype.__init.call(this);
    const client = _openapifetch2.default.call(void 0, { baseUrl });
    const sendFn = async (batch) => {
      const normalisedBatch = normalise(batch);
      const res = await client.POST("/v2/events", { body: normalisedBatch });
      if (res.response.status !== 200) {
        throw new Error(String(res.error));
      }
    };
    this.sender = new Sender({
      batchSize: 100,
      baseTimeoutMs: 100,
      maxFailureCount: 6,
      maxTimeoutMs: 1e4,
      sendFn
    });
  }
  enable() {
    this.enabled = true;
  }
  /**
   * Merges a field into the session base (shared with {@link getSessionBaseProperties}).
   *
   * @param key - Base property name.
   * @param value - Value for that property.
   */
  setGlobalProperty(key, value) {
    mergeSessionProperties({
      [key]: value
    });
  }
  /**
   * Sends `smart_accounts_kit_initialized` using the session base from {@link getInitializationContext}
   * plus optional per-event overrides. Updates the stored session with the merged snapshot.
   *
   * @param properties - Optional overrides; omit to use the current session base only.
   */
  trackInitialized(properties = {}) {
    if (!this.enabled) {
      return;
    }
    const merged = {
      ...getSessionBaseProperties(),
      ...properties
    };
    if (!isCompleteBase(merged)) {
      throw new Error(
        "Analytics: trackInitialized produced incomplete base configuration (ensure getInitializationContext ran and sdk_version, anon_id, platform are set)"
      );
    }
    mergeSessionProperties(merged);
    const event = {
      namespace: "metamask/smart-accounts-kit",
      event_name: "smart_accounts_kit_initialized",
      properties: merged
    };
    this.sender.enqueue(event);
  }
  /**
   * Sends `smart_accounts_kit_function_called` with session base plus the function name and optional
   * non-sensitive parameters. Does not mutate the session store (unlike {@link trackInitialized}).
   *
   * @param functionName - Public SDK entry name (use a stable string, e.g. `createDelegation`).
   * @param parameters - Safe primitives only; omit secrets, keys, and raw addresses if sensitive.
   * @param baseOverrides - Optional overrides for base fields (same as {@link trackInitialized}).
   */
  trackSdkFunctionCall(functionName, parameters, baseOverrides = {}) {
    if (!this.enabled) {
      return;
    }
    const mergedBase = {
      ...getSessionBaseProperties(),
      ...baseOverrides
    };
    if (!isCompleteBase(mergedBase)) {
      throw new Error(
        "Analytics: trackSdkFunctionCall requires session (call getInitializationContext before tracking)"
      );
    }
    const props = {
      ...mergedBase,
      function_name: functionName,
      ...parameters !== void 0 && Object.keys(parameters).length > 0 ? { parameters } : {}
    };
    const event = {
      namespace: "metamask/smart-accounts-kit",
      event_name: "smart_accounts_kit_function_called",
      properties: props
    };
    this.sender.enqueue(event);
  }
}, _class);
var analytics = new Analytics(METAMASK_ANALYTICS_ENDPOINT);
function isAnalyticsDisabled() {
  const dntValues = [];
  if (typeof process !== "undefined") {
    dntValues.push(_optionalChain([process, 'access', _103 => _103.env, 'optionalAccess', _104 => _104.CI]));
    dntValues.push(_optionalChain([process, 'access', _105 => _105.env, 'optionalAccess', _106 => _106.DO_NOT_TRACK]));
  }
  if (typeof navigator !== "undefined") {
    dntValues.push(navigator.doNotTrack);
  }
  if (typeof window !== "undefined") {
    dntValues.push(window.doNotTrack);
  }
  return dntValues.some(
    (dntValue) => dntValue === "1" || _optionalChain([dntValue, 'optionalAccess', _107 => _107.toLowerCase, 'call', _108 => _108()]) === "yes" || _optionalChain([dntValue, 'optionalAccess', _109 => _109.toLowerCase, 'call', _110 => _110()]) === "true"
  );
}
var hasBootstrapped = false;
function ensureSmartAccountsKitAnalyticsBootstrapped() {
  if (hasBootstrapped) {
    return;
  }
  hasBootstrapped = true;
  if (isAnalyticsDisabled()) {
    return;
  }
  try {
    getInitializationContext({ sdk_version: version });
    analytics.enable();
    analytics.trackInitialized();
  } catch (e2) {
  }
}
function trackSmartAccountsKitFunctionCall(functionName, parameters) {
  ensureSmartAccountsKitAnalyticsBootstrapped();
  try {
    analytics.trackSdkFunctionCall(functionName, parameters);
  } catch (e3) {
  }
}

// src/delegation.ts
var DELEGATION_ABI_TYPE_COMPONENTS = [
  { type: "address", name: "delegate" },
  { type: "address", name: "delegator" },
  { type: "bytes32", name: "authority" },
  { type: "tuple[]", name: "caveats", components: CAVEAT_ABI_TYPE_COMPONENTS },
  { type: "uint256", name: "salt" },
  { type: "bytes", name: "signature" }
];
var toDelegationStruct = (delegation) => {
  const caveats = delegation.caveats.map((caveat) => ({
    enforcer: _viem.getAddress.call(void 0, caveat.enforcer),
    terms: caveat.terms,
    args: caveat.args
  }));
  const salt = delegation.salt === "0x" ? 0n : BigInt(delegation.salt);
  return {
    delegate: _viem.getAddress.call(void 0, delegation.delegate),
    delegator: _viem.getAddress.call(void 0, delegation.delegator),
    authority: _nullishCoalesce(delegation.authority, () => ( _delegationcore.ROOT_AUTHORITY)),
    caveats,
    salt,
    signature: delegation.signature
  };
};
var toDelegation = (delegationStruct) => {
  return {
    ...delegationStruct,
    salt: _viem.toHex.call(void 0, delegationStruct.salt)
  };
};
var encodeDelegations = (delegations) => {
  if (Array.isArray(delegations)) {
    const delegationStructs = delegations.map(toDelegationStruct);
    return _delegationcore.encodeDelegations.call(void 0, delegationStructs);
  }
  if (_viem.isHex.call(void 0, delegations)) {
    return delegations;
  }
  throw new Error(
    "Invalid delegations - must be an array of delegations or a hex string"
  );
};
var encodeDelegation = (delegation) => {
  const delegationStruct = toDelegationStruct(delegation);
  return _delegationcore.encodeDelegation.call(void 0, delegationStruct);
};
var decodeDelegations = (delegations) => {
  if (Array.isArray(delegations)) {
    return delegations;
  }
  if (_viem.isHex.call(void 0, delegations)) {
    return _delegationcore.decodeDelegations.call(void 0, delegations).map(toDelegation);
  }
  throw new Error(
    "Invalid delegations - must be an array of delegations or a hex string"
  );
};
var decodeDelegation = (encoded) => {
  return toDelegation(_delegationcore.decodeDelegation.call(void 0, encoded));
};
var SIGNABLE_DELEGATION_TYPED_DATA = {
  Caveat: [
    { name: "enforcer", type: "address" },
    { name: "terms", type: "bytes" }
  ],
  Delegation: [
    { name: "delegate", type: "address" },
    { name: "delegator", type: "address" },
    { name: "authority", type: "bytes32" },
    { name: "caveats", type: "Caveat[]" },
    { name: "salt", type: "uint256" }
  ]
};
var DELEGATION_ARRAY_ABI_TYPE = {
  type: "tuple[]",
  components: DELEGATION_ABI_TYPE_COMPONENTS
};
var DELEGATION_ABI_TYPE = {
  type: "tuple",
  components: DELEGATION_ABI_TYPE_COMPONENTS
};
var hashDelegation = (input) => {
  const delegationStruct = toDelegationStruct(input);
  return _delegationcore.hashDelegation.call(void 0, delegationStruct);
};
var extractLeafDelegation = (permissionContext) => {
  const delegations = decodeDelegations(permissionContext);
  if (delegations.length === 0) {
    throw new Error("Permission context must contain at least one delegation");
  }
  const leafDelegation = delegations[0];
  if (!leafDelegation) {
    throw new Error("Failed to extract leaf delegation");
  }
  return leafDelegation;
};
var getCaveatNames = ({
  caveats,
  environment: { caveatEnforcers }
}) => {
  if (Array.isArray(caveats)) {
    const knownEnforcers = Object.entries(caveatEnforcers).map(
      ([name, address]) => ({ name, address: address.toLowerCase() })
    );
    return caveats.map((caveat) => {
      const enforcerAddressLowercase = _optionalChain([caveat, 'access', _111 => _111.enforcer, 'optionalAccess', _112 => _112.toLowerCase, 'call', _113 => _113()]);
      const matchingCaveat = knownEnforcers.find(
        ({ address }) => address === enforcerAddressLowercase
      );
      return _nullishCoalesce(_optionalChain([matchingCaveat, 'optionalAccess', _114 => _114.name]), () => ( "Unknown"));
    });
  }
  return [];
};
var resolveDelegationArgs = (options) => {
  const optionsHasParentPermissionContext = "parentPermissionContext" in options && options.parentPermissionContext;
  const resolvedParentDelegation = optionsHasParentPermissionContext ? extractLeafDelegation(options.parentPermissionContext) : options.parentDelegation;
  const authority = (() => {
    if (!resolvedParentDelegation) {
      if (!options.scope) {
        throw new Error(
          "Invalid arguments - must specify either parentDelegation, parentPermissionContext, or scope"
        );
      }
      return _delegationcore.ROOT_AUTHORITY;
    }
    if (typeof resolvedParentDelegation === "string") {
      if (resolvedParentDelegation.toLowerCase() === _delegationcore.ROOT_AUTHORITY.toLowerCase()) {
        throw new Error(
          `Invalid parent delegation - cannot be ${_delegationcore.ROOT_AUTHORITY}`
        );
      }
      return resolvedParentDelegation;
    }
    return hashDelegation(resolvedParentDelegation);
  })();
  const hasInheritance = resolvedParentDelegation !== void 0;
  const caveats = resolveCaveats({
    environment: options.environment,
    scope: options.scope,
    caveats: options.caveats,
    isScopeOptional: hasInheritance
  });
  return {
    authority,
    caveats,
    hasInheritance,
    salt: _nullishCoalesce(options.salt, () => ( "0x00"))
  };
};
var createDelegation = (options) => {
  const { authority, caveats, hasInheritance, salt } = resolveDelegationArgs(options);
  trackSmartAccountsKitFunctionCall("createDelegation", {
    hasInheritance,
    scope: _nullishCoalesce(_optionalChain([options, 'access', _115 => _115.scope, 'optionalAccess', _116 => _116.type]), () => ( null)),
    caveatNames: getCaveatNames({
      caveats,
      environment: options.environment
    })
  });
  return {
    delegate: options.to,
    delegator: options.from,
    authority,
    caveats,
    salt,
    signature: "0x"
  };
};
var createOpenDelegation = (options) => {
  const { authority, caveats, hasInheritance, salt } = resolveDelegationArgs(options);
  trackSmartAccountsKitFunctionCall("createOpenDelegation", {
    hasInheritance,
    scope: _nullishCoalesce(_optionalChain([options, 'access', _117 => _117.scope, 'optionalAccess', _118 => _118.type]), () => ( null)),
    caveatNames: getCaveatNames({
      caveats,
      environment: options.environment
    })
  });
  return {
    delegate: _delegationcore.ANY_BENEFICIARY,
    delegator: options.from,
    authority,
    caveats,
    salt,
    signature: "0x"
  };
};
var prepareSignDelegationTypedData = ({
  delegation,
  delegationManager,
  chainId,
  name = "DelegationManager",
  version: version2 = "1",
  allowInsecureUnrestrictedDelegation = false
}) => {
  const delegationStruct = toDelegationStruct({
    ...delegation,
    signature: "0x"
  });
  if (delegationStruct.caveats.length === 0 && !allowInsecureUnrestrictedDelegation) {
    throw new Error(
      "No caveats found. If you definitely want to sign a delegation without caveats, set `allowInsecureUnrestrictedDelegation` to `true`."
    );
  }
  return {
    domain: {
      chainId,
      name,
      version: version2,
      verifyingContract: delegationManager
    },
    types: SIGNABLE_DELEGATION_TYPED_DATA,
    primaryType: "Delegation",
    message: delegationStruct
  };
};
var signDelegation = async ({
  privateKey,
  delegation,
  delegationManager,
  chainId,
  name = "DelegationManager",
  version: version2 = "1",
  allowInsecureUnrestrictedDelegation = false
}) => {
  const typedData = prepareSignDelegationTypedData({
    delegation,
    delegationManager,
    chainId,
    name,
    version: version2,
    allowInsecureUnrestrictedDelegation
  });
  return _accounts.signTypedData.call(void 0, {
    privateKey,
    ...typedData
  });
};

// src/executions.ts




var createExecution = ({
  target,
  value = 0n,
  callData = "0x"
}) => {
  return {
    target,
    value,
    callData
  };
};
var ExecutionMode = /* @__PURE__ */ ((ExecutionMode2) => {
  ExecutionMode2["SingleDefault"] = "0x0000000000000000000000000000000000000000000000000000000000000000";
  ExecutionMode2["SingleTry"] = "0x0001000000000000000000000000000000000000000000000000000000000000";
  ExecutionMode2["BatchDefault"] = "0x0100000000000000000000000000000000000000000000000000000000000000";
  ExecutionMode2["BatchTry"] = "0x0101000000000000000000000000000000000000000000000000000000000000";
  return ExecutionMode2;
})(ExecutionMode || {});
var EXECUTION_ABI_TYPE_COMPONENTS = [
  { type: "address", name: "target" },
  { type: "uint256", name: "value" },
  { type: "bytes", name: "callData" }
];
var encodeSingleExecution = (execution) => {
  return _viem.encodePacked.call(void 0, 
    ["address", "uint256", "bytes"],
    [execution.target, execution.value, execution.callData]
  );
};
var encodeBatchExecution = (executions) => {
  return _viem.encodeAbiParameters.call(void 0, 
    [
      {
        components: EXECUTION_ABI_TYPE_COMPONENTS,
        name: "executions",
        type: "tuple[]"
      }
    ],
    [executions]
  );
};
var encodeExecutionCalldata = (executions) => {
  if (executions.length === 0) {
    throw new Error(
      "Error while getting the execution calldatas, executions is empty"
    );
  }
  if (executions.length === 1) {
    const execution = executions[0];
    if (!execution) {
      throw new Error("Execution not found");
    }
    return encodeSingleExecution(execution);
  }
  return encodeBatchExecution(executions);
};
var encodeExecutionCalldatas = (executionsBatch) => {
  if (executionsBatch.length === 0) {
    throw new Error(
      "Error while getting the execution calldatas, executionsBatch is empty"
    );
  }
  return executionsBatch.map(encodeExecutionCalldata);
};







































exports.__export = __export; exports.trackSmartAccountsKitFunctionCall = trackSmartAccountsKitFunctionCall; exports.Implementation = Implementation; exports.TransferWindow = TransferWindow; exports.ScopeType = ScopeType; exports.CaveatType = CaveatType; exports.CaveatBuilder = CaveatBuilder; exports.createCaveatBuilder = createCaveatBuilder; exports.isDefined = isDefined; exports.toHexOrThrow = toHexOrThrow; exports.resolveCaveats = resolveCaveats; exports.createCaveat = createCaveat; exports.decodeCaveat = decodeCaveat; exports.ANY_BENEFICIARY = _delegationcore.ANY_BENEFICIARY; exports.DELEGATION_TYPEHASH = _delegationcore.DELEGATION_TYPEHASH; exports.ROOT_AUTHORITY = _delegationcore.ROOT_AUTHORITY; exports.DELEGATION_ABI_TYPE_COMPONENTS = DELEGATION_ABI_TYPE_COMPONENTS; exports.toDelegationStruct = toDelegationStruct; exports.toDelegation = toDelegation; exports.encodeDelegations = encodeDelegations; exports.encodeDelegation = encodeDelegation; exports.decodeDelegations = decodeDelegations; exports.decodeDelegation = decodeDelegation; exports.SIGNABLE_DELEGATION_TYPED_DATA = SIGNABLE_DELEGATION_TYPED_DATA; exports.DELEGATION_ARRAY_ABI_TYPE = DELEGATION_ARRAY_ABI_TYPE; exports.DELEGATION_ABI_TYPE = DELEGATION_ABI_TYPE; exports.hashDelegation = hashDelegation; exports.createDelegation = createDelegation; exports.createOpenDelegation = createOpenDelegation; exports.prepareSignDelegationTypedData = prepareSignDelegationTypedData; exports.signDelegation = signDelegation; exports.createExecution = createExecution; exports.ExecutionMode = ExecutionMode; exports.encodeSingleExecution = encodeSingleExecution; exports.encodeBatchExecution = encodeBatchExecution; exports.encodeExecutionCalldata = encodeExecutionCalldata; exports.encodeExecutionCalldatas = encodeExecutionCalldatas;
