// @metamask/smart-accounts-kit/dist/chunk-JKVSTPEC.cjs
"use strict";Object.defineProperty(exports, "__esModule", {value: true}); function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { newObj[key] = obj[key]; } } } newObj.default = obj; return newObj; } } function _nullishCoalesce(lhs, rhsFn) { if (lhs != null) { return lhs; } else { return rhsFn(); } }



var _chunkW6HPAK4Mcjs = require('./chunk-W6HPAK4M.cjs');

// src/DelegationFramework/utils.ts
var _delegationabis = require('@metamask/delegation-abis'); var delegationAbis = _interopRequireWildcard(_delegationabis);
var _bytecode = require('@metamask/delegation-abis/bytecode');
var _viem = require('viem');
var _actions = require('viem/actions');

// src/DelegationFramework/DeleGatorCore/read.ts
var read_exports = {};
_chunkW6HPAK4Mcjs.__export.call(void 0, read_exports, {
  getDelegationManager: () => read,
  getDeposit: () => read2,
  getEntryPoint: () => read3,
  getNonce: () => read4,
  getProxyImplementation: () => read5,
  getProxyVersion: () => read6,
  isValidSignature: () => read7
});

// src/DelegationFramework/DeleGatorCore/methods/getDelegationManager.ts


var read = async ({
  client,
  contractAddress
}) => await _actions.readContract.call(void 0, client, {
  address: contractAddress,
  abi: _delegationabis.DeleGatorCore,
  functionName: "delegationManager"
});

// src/DelegationFramework/DeleGatorCore/methods/getDeposit.ts


var read2 = async ({
  client,
  contractAddress
}) => await _actions.readContract.call(void 0, client, {
  address: contractAddress,
  abi: _delegationabis.DeleGatorCore,
  functionName: "getDeposit"
});

// src/DelegationFramework/DeleGatorCore/methods/getEntryPoint.ts


var read3 = async ({
  client,
  contractAddress
}) => await _actions.readContract.call(void 0, client, {
  address: contractAddress,
  abi: _delegationabis.DeleGatorCore,
  functionName: "entryPoint"
});

// src/DelegationFramework/DeleGatorCore/methods/getNonce.ts


var read4 = async ({
  client,
  contractAddress,
  key
}) => await _actions.readContract.call(void 0, client, {
  address: contractAddress,
  abi: _delegationabis.DeleGatorCore,
  functionName: "getNonce",
  args: key ? [key] : void 0
});

// src/DelegationFramework/DeleGatorCore/methods/getProxyImplementation.ts


var read5 = async ({
  client,
  contractAddress
}) => await _actions.readContract.call(void 0, client, {
  address: contractAddress,
  abi: _delegationabis.DeleGatorCore,
  functionName: "getImplementation"
});

// src/DelegationFramework/DeleGatorCore/methods/getProxyVersion.ts


var read6 = async ({
  client,
  contractAddress
}) => await _actions.readContract.call(void 0, client, {
  address: contractAddress,
  abi: _delegationabis.DeleGatorCore,
  functionName: "getInitializedVersion"
});

// src/DelegationFramework/DeleGatorCore/methods/isValidSignature.ts



var read7 = async ({
  client,
  contractAddress,
  hash,
  signature
}) => await _actions.readContract.call(void 0, client, {
  address: contractAddress,
  abi: _delegationabis.DeleGatorCore,
  functionName: "isValidSignature",
  args: [hash, signature]
});
var encode = ({
  hash,
  signature
}) => {
  return _viem.encodeFunctionData.call(void 0, {
    abi: _delegationabis.DeleGatorCore,
    functionName: "isValidSignature",
    args: [hash, signature]
  });
};

// src/DelegationFramework/utils.ts
async function isContractDeployed({
  client,
  contractAddress
}) {
  const code = await _actions.getCode.call(void 0, client, {
    address: contractAddress
  });
  return Boolean(code) && code !== "0x";
}
async function isImplementationExpected({
  client,
  contractAddress,
  expectedImplementationAddress
}) {
  if (!await isContractDeployed({ client, contractAddress })) {
    return false;
  }
  const implementationAddress = await read5({
    client,
    contractAddress
  });
  return implementationAddress === expectedImplementationAddress;
}
var encodeProxyCreationCode = ({
  implementationAddress,
  initcode
}) => _viem.encodeDeployData.call(void 0, {
  abi: _delegationabis.ERC1967Proxy,
  args: [implementationAddress, initcode],
  bytecode: _bytecode.ERC1967Proxy
});

// src/decodeRevertReason.ts


var _utils = require('viem/utils');
var knownRevertAbis = Object.values(delegationAbis).filter(
  (abi) => abi.length > 0
);
var standardSolidityErrorAbi = [];
var revertReasonAbis = [
  standardSolidityErrorAbi,
  ...knownRevertAbis
];
var MAX_ERROR_TRAVERSAL_DEPTH = 8;
var panicReasons = {
  "1": "An `assert` condition failed.",
  "17": "Arithmetic operation resulted in underflow or overflow.",
  "18": "Division or modulo by zero.",
  "33": "Attempted to convert to an invalid type.",
  "34": "Attempted to access a storage byte array that is incorrectly encoded.",
  "49": "Performed `.pop()` on an empty array.",
  "50": "Array index is out of bounds.",
  "65": "Allocated too much memory or created an array which is too large.",
  "81": "Attempted to call a zero-initialized variable of internal function type."
};
function decodeRevertReason(error) {
  const decodedViemError = decodeViemContractRevert(error);
  if (decodedViemError) {
    return decodedViemError;
  }
  for (const rawData of getRevertDataCandidates(error)) {
    const decoded = decodeRevertData(rawData);
    if (decoded) {
      return decoded;
    }
  }
  return void 0;
}
function decodeViemContractRevert(error) {
  if (!(error instanceof _viem.BaseError)) {
    return void 0;
  }
  const revertError = error.walk(
    (cause) => cause instanceof _viem.ContractFunctionRevertedError
  );
  if (!(revertError instanceof _viem.ContractFunctionRevertedError)) {
    return void 0;
  }
  if (!revertError.raw) {
    return void 0;
  }
  if (revertError.data) {
    const { abiItem, args, errorName } = revertError.data;
    return {
      errorName,
      message: formatDecodedError(errorName, args, abiItem),
      rawData: revertError.raw
    };
  }
  return decodeRevertData(revertError.raw);
}
function decodeRevertData(rawData) {
  for (const abi of revertReasonAbis) {
    try {
      const { abiItem, args, errorName } = _utils.decodeErrorResult.call(void 0, {
        abi,
        data: rawData
      });
      return {
        errorName,
        message: formatDecodedError(errorName, args, abiItem),
        rawData
      };
    } catch (e) {
    }
  }
  const decodedString = decodeRawString(rawData);
  if (decodedString) {
    return {
      errorName: "Error",
      message: decodedString,
      rawData
    };
  }
  return void 0;
}
function formatDecodedError(errorName, args, abiItem) {
  const [firstArg] = _nullishCoalesce(args, () => ( []));
  if (errorName === "Error") {
    return typeof firstArg === "string" ? firstArg : errorName;
  }
  if (errorName === "Panic") {
    const panicCode = String(firstArg);
    return _nullishCoalesce(panicReasons[panicCode], () => ( `Panic(${panicCode})`));
  }
  const formattedArgs = _utils.formatAbiItemWithArgs.call(void 0, {
    abiItem,
    args: _nullishCoalesce(args, () => ( [])),
    includeFunctionName: false,
    includeName: false
  });
  return `${errorName}${_nullishCoalesce(formattedArgs, () => ( ""))}`;
}
function decodeRawString(rawData) {
  const bytes = rawData.slice(2);
  if (bytes.length === 0 || bytes.length % 2 !== 0) {
    return void 0;
  }
  let value = "";
  for (let index = 0; index < bytes.length; index += 2) {
    const charCode = Number.parseInt(bytes.slice(index, index + 2), 16);
    if (Number.isNaN(charCode) || charCode < 32 || charCode > 126) {
      return void 0;
    }
    value += String.fromCharCode(charCode);
  }
  return value;
}
function getRevertDataCandidates(error) {
  const candidates = [];
  const seen = /* @__PURE__ */ new Set();
  const seenCandidates = /* @__PURE__ */ new Set();
  const addHexCandidate = (candidate) => {
    if (candidate.length < 10 || candidate.length % 2 !== 0 || !_viem.isHex.call(void 0, candidate)) {
      return;
    }
    if (!seenCandidates.has(candidate)) {
      seenCandidates.add(candidate);
      candidates.push(candidate);
    }
  };
  const addLabeledHexCandidates = (value) => {
    for (const match of value.matchAll(
      /\b(?:reason|revertData|raw|data):\s*(0x[0-9a-fA-F]+)/giu
    )) {
      const [, candidate] = match;
      if (candidate) {
        addHexCandidate(candidate);
      }
    }
  };
  const addHexCandidates = (value) => {
    if (typeof value !== "string") {
      return;
    }
    addLabeledHexCandidates(value);
    for (const [candidate] of value.matchAll(/0x[0-9a-fA-F]+/gu)) {
      addHexCandidate(candidate);
    }
  };
  const visit = (value, depth = 0) => {
    if (value === null || value === void 0 || seen.has(value) || depth > MAX_ERROR_TRAVERSAL_DEPTH) {
      return;
    }
    addHexCandidates(value);
    if (Array.isArray(value)) {
      seen.add(value);
      value.forEach((item) => visit(item, depth + 1));
      return;
    }
    if (typeof value !== "object") {
      return;
    }
    seen.add(value);
    const record = value;
    addHexCandidates(record.revertData);
    addHexCandidates(record.raw);
    addHexCandidates(record.data);
    addHexCandidates(record.details);
    addHexCandidates(record.reason);
    addHexCandidates(record.shortMessage);
    addHexCandidates(record.message);
    visit(record.data, depth + 1);
    visit(record.details, depth + 1);
    visit(record.error, depth + 1);
    visit(record.metaMessages, depth + 1);
    visit(record.originalError, depth + 1);
    visit(record.cause, depth + 1);
  };
  visit(error);
  return candidates;
}

// src/DelegationFramework/HybridDeleGator/methods/initialize.ts



var simulate = async ({
  client,
  hybridDeleGatorAddress,
  eoaOwner,
  p256Owners
}) => {
  return _actions.simulateContract.call(void 0, client, {
    address: hybridDeleGatorAddress,
    abi: _delegationabis.HybridDeleGator,
    functionName: "initialize",
    args: [
      eoaOwner,
      p256Owners.map((p256Owner) => p256Owner.keyId),
      p256Owners.map((p256Owner) => p256Owner.x),
      p256Owners.map((p256Owner) => p256Owner.y)
    ]
  });
};
var encode2 = ({
  eoaOwner,
  p256Owners
}) => {
  return _viem.encodeFunctionData.call(void 0, {
    abi: _delegationabis.HybridDeleGator,
    functionName: "initialize",
    args: [
      eoaOwner,
      p256Owners.map((p256Owner) => p256Owner.keyId),
      p256Owners.map((p256Owner) => p256Owner.x),
      p256Owners.map((p256Owner) => p256Owner.y)
    ]
  });
};

// src/DelegationFramework/HybridDeleGator/encode.ts
var encode_exports = {};
_chunkW6HPAK4Mcjs.__export.call(void 0, encode_exports, {
  addKey: () => encode3,
  initializeHybridDeleGator: () => encode2,
  reinitializeHybridDeleGator: () => encode4,
  removeKey: () => encode5,
  updateSigners: () => encode6
});

// src/DelegationFramework/HybridDeleGator/methods/addKey.ts



var simulate2 = async ({
  client,
  hybridDeleGatorAddress,
  p256Owner
}) => {
  return _actions.simulateContract.call(void 0, client, {
    address: hybridDeleGatorAddress,
    abi: _delegationabis.HybridDeleGator,
    functionName: "addKey",
    args: [p256Owner.keyId, p256Owner.x, p256Owner.y]
  });
};
var encode3 = ({ p256Owner }) => {
  return _viem.encodeFunctionData.call(void 0, {
    abi: _delegationabis.HybridDeleGator,
    functionName: "addKey",
    args: [p256Owner.keyId, p256Owner.x, p256Owner.y]
  });
};

// src/DelegationFramework/HybridDeleGator/methods/reinitialize.ts



var simulate3 = async ({
  client,
  hybridDeleGatorAddress,
  version,
  eoaOwner,
  p256Owners,
  removeExistingP256Owners
}) => {
  return _actions.simulateContract.call(void 0, client, {
    address: hybridDeleGatorAddress,
    abi: _delegationabis.HybridDeleGator,
    functionName: "reinitialize",
    args: [
      version,
      eoaOwner,
      p256Owners.map((p256Owner) => p256Owner.keyId),
      p256Owners.map((p256Owner) => p256Owner.x),
      p256Owners.map((p256Owner) => p256Owner.y),
      removeExistingP256Owners
    ]
  });
};
var encode4 = ({
  version,
  eoaOwner,
  p256Owners,
  removeExistingP256Owners
}) => {
  return _viem.encodeFunctionData.call(void 0, {
    abi: _delegationabis.HybridDeleGator,
    functionName: "reinitialize",
    args: [
      version,
      eoaOwner,
      p256Owners.map((p256Owner) => p256Owner.keyId),
      p256Owners.map((p256Owner) => p256Owner.x),
      p256Owners.map((p256Owner) => p256Owner.y),
      removeExistingP256Owners
    ]
  });
};

// src/DelegationFramework/HybridDeleGator/methods/removeKey.ts



var simulate4 = async ({
  client,
  hybridDeleGatorAddress,
  keyId
}) => {
  return _actions.simulateContract.call(void 0, client, {
    address: hybridDeleGatorAddress,
    abi: _delegationabis.HybridDeleGator,
    functionName: "removeKey",
    args: [keyId]
  });
};
var encode5 = ({ keyId }) => {
  return _viem.encodeFunctionData.call(void 0, {
    abi: _delegationabis.HybridDeleGator,
    functionName: "removeKey",
    args: [keyId]
  });
};

// src/DelegationFramework/HybridDeleGator/methods/updateSigners.ts



var simulate5 = async ({
  client,
  hybridDeleGatorAddress,
  eoaOwner,
  p256Owners
}) => {
  return _actions.simulateContract.call(void 0, client, {
    address: hybridDeleGatorAddress,
    abi: _delegationabis.HybridDeleGator,
    functionName: "updateSigners",
    args: [
      eoaOwner,
      p256Owners.map((p256Owner) => p256Owner.keyId),
      p256Owners.map((p256Owner) => p256Owner.x),
      p256Owners.map((p256Owner) => p256Owner.y)
    ]
  });
};
var encode6 = ({
  eoaOwner,
  p256Owners
}) => {
  return _viem.encodeFunctionData.call(void 0, {
    abi: _delegationabis.HybridDeleGator,
    functionName: "updateSigners",
    args: [
      eoaOwner,
      p256Owners.map((p256Owner) => p256Owner.keyId),
      p256Owners.map((p256Owner) => p256Owner.x),
      p256Owners.map((p256Owner) => p256Owner.y)
    ]
  });
};

// src/DelegationFramework/MultiSigDeleGator/methods/initialize.ts



var simulate6 = async ({
  client,
  multiSigDeleGatorAddress,
  owners,
  threshold
}) => {
  return _actions.simulateContract.call(void 0, client, {
    address: multiSigDeleGatorAddress,
    abi: _delegationabis.MultiSigDeleGator,
    functionName: "initialize",
    args: [owners, threshold]
  });
};
var encode7 = ({ owners, threshold }) => {
  return _viem.encodeFunctionData.call(void 0, {
    abi: _delegationabis.MultiSigDeleGator,
    functionName: "initialize",
    args: [owners, threshold]
  });
};

// src/DelegationFramework/MultiSigDeleGator/encode.ts
var encode_exports2 = {};
_chunkW6HPAK4Mcjs.__export.call(void 0, encode_exports2, {
  addSigner: () => encode8,
  initializeMultiSigDeleGator: () => encode7,
  reinitializeMultiSigDeleGator: () => encode9,
  removeSigner: () => encode10,
  replaceSigner: () => encode11,
  updateMultiSigParameters: () => encode12,
  updateThreshold: () => encode13
});

// src/DelegationFramework/MultiSigDeleGator/methods/addSigner.ts



var simulate7 = async ({
  client,
  multiSigDeleGatorAddress,
  signer
}) => {
  return _actions.simulateContract.call(void 0, client, {
    address: multiSigDeleGatorAddress,
    abi: _delegationabis.MultiSigDeleGator,
    functionName: "addSigner",
    args: [signer]
  });
};
var encode8 = ({ signer }) => {
  return _viem.encodeFunctionData.call(void 0, {
    abi: _delegationabis.MultiSigDeleGator,
    functionName: "addSigner",
    args: [signer]
  });
};

// src/DelegationFramework/MultiSigDeleGator/methods/reinitialize.ts



var simulate8 = async ({
  client,
  multiSigDeleGatorAddress,
  version,
  owners,
  threshold,
  removeExistingOwners
}) => {
  return _actions.simulateContract.call(void 0, client, {
    address: multiSigDeleGatorAddress,
    abi: _delegationabis.MultiSigDeleGator,
    functionName: "reinitialize",
    args: [version, owners, threshold, removeExistingOwners]
  });
};
var encode9 = ({
  version,
  owners,
  threshold,
  removeExistingOwners
}) => {
  return _viem.encodeFunctionData.call(void 0, {
    abi: _delegationabis.MultiSigDeleGator,
    functionName: "reinitialize",
    args: [version, owners, threshold, removeExistingOwners]
  });
};

// src/DelegationFramework/MultiSigDeleGator/methods/removeSigner.ts



var simulate9 = async ({
  client,
  multiSigDeleGatorAddress,
  signer
}) => {
  return _actions.simulateContract.call(void 0, client, {
    address: multiSigDeleGatorAddress,
    abi: _delegationabis.MultiSigDeleGator,
    functionName: "removeSigner",
    args: [signer]
  });
};
var encode10 = ({ signer }) => {
  return _viem.encodeFunctionData.call(void 0, {
    abi: _delegationabis.MultiSigDeleGator,
    functionName: "removeSigner",
    args: [signer]
  });
};

// src/DelegationFramework/MultiSigDeleGator/methods/replaceSigner.ts



var simulate10 = async ({
  client,
  multiSigDeleGatorAddress,
  oldSigner,
  newSigner
}) => {
  return _actions.simulateContract.call(void 0, client, {
    address: multiSigDeleGatorAddress,
    abi: _delegationabis.MultiSigDeleGator,
    functionName: "replaceSigner",
    args: [oldSigner, newSigner]
  });
};
var encode11 = ({
  oldSigner,
  newSigner
}) => {
  return _viem.encodeFunctionData.call(void 0, {
    abi: _delegationabis.MultiSigDeleGator,
    functionName: "replaceSigner",
    args: [oldSigner, newSigner]
  });
};

// src/DelegationFramework/MultiSigDeleGator/methods/updateMultiSigParameters.ts



var simulate11 = async ({
  client,
  multiSigDeleGatorAddress,
  owners,
  threshold,
  removeExistingOwners
}) => {
  return _actions.simulateContract.call(void 0, client, {
    address: multiSigDeleGatorAddress,
    abi: _delegationabis.MultiSigDeleGator,
    functionName: "updateMultiSigParameters",
    args: [owners, threshold, removeExistingOwners]
  });
};
var encode12 = ({
  owners,
  threshold,
  removeExistingOwners
}) => {
  return _viem.encodeFunctionData.call(void 0, {
    abi: _delegationabis.MultiSigDeleGator,
    functionName: "updateMultiSigParameters",
    args: [owners, threshold, removeExistingOwners]
  });
};

// src/DelegationFramework/MultiSigDeleGator/methods/updateThreshold.ts



var simulate12 = async ({
  client,
  multiSigDeleGatorAddress,
  threshold
}) => {
  return _actions.simulateContract.call(void 0, client, {
    address: multiSigDeleGatorAddress,
    abi: _delegationabis.MultiSigDeleGator,
    functionName: "updateThreshold",
    args: [threshold]
  });
};
var encode13 = ({ threshold }) => {
  return _viem.encodeFunctionData.call(void 0, {
    abi: _delegationabis.MultiSigDeleGator,
    functionName: "updateThreshold",
    args: [threshold]
  });
};

// src/DelegationFramework/SimpleFactory/methods/create2Deploy.ts



var simulate13 = async ({
  client,
  factoryAddress,
  creationCode,
  salt
}) => {
  return _actions.simulateContract.call(void 0, client, {
    address: factoryAddress,
    abi: _delegationabis.SimpleFactory,
    functionName: "deploy",
    args: [creationCode, salt]
  });
};
var encode14 = (creationCode, salt) => {
  return _viem.encodeFunctionData.call(void 0, {
    abi: _delegationabis.SimpleFactory,
    functionName: "deploy",
    args: [creationCode, salt]
  });
};
var execute = async ({
  client,
  factoryAddress,
  creationCode,
  salt
}) => {
  const { request } = await simulate13({
    client,
    factoryAddress,
    creationCode,
    salt
  });
  return _actions.writeContract.call(void 0, client, request);
};

// src/DelegationFramework/SimpleFactory/encode.ts
var encode_exports3 = {};
_chunkW6HPAK4Mcjs.__export.call(void 0, encode_exports3, {
  create2Deploy: () => encode14
});

// src/DelegationFramework/DeleGatorCore/methods/execute.ts



var simulate14 = async ({
  client,
  contractAddress,
  execution
}) => {
  return _actions.simulateContract.call(void 0, client, {
    address: contractAddress,
    abi: _delegationabis.DeleGatorCore,
    functionName: "execute",
    args: [execution]
  });
};
var execute2 = async ({
  client,
  contractAddress,
  execution
}) => {
  const { request } = await simulate14({
    client,
    contractAddress,
    execution
  });
  return _actions.writeContract.call(void 0, client, request);
};
var encode15 = ({ execution }) => {
  return _viem.encodeFunctionData.call(void 0, {
    abi: _delegationabis.DeleGatorCore,
    functionName: "execute",
    args: [execution]
  });
};

// src/DelegationFramework/DeleGatorCore/methods/executeWithMode.ts



var simulate15 = async ({
  client,
  contractAddress,
  mode,
  executions
}) => {
  return _actions.simulateContract.call(void 0, client, {
    address: contractAddress,
    abi: _delegationabis.DeleGatorCore,
    functionName: "execute",
    args: [mode, _chunkW6HPAK4Mcjs.encodeExecutionCalldata.call(void 0, executions)]
  });
};
var execute3 = async ({
  client,
  contractAddress,
  mode,
  executions
}) => {
  const { request } = await simulate15({
    client,
    contractAddress,
    mode,
    executions
  });
  return _actions.writeContract.call(void 0, client, request);
};
var encode16 = ({
  mode,
  executions
}) => {
  return _viem.encodeFunctionData.call(void 0, {
    abi: _delegationabis.DeleGatorCore,
    functionName: "execute",
    args: [mode, _chunkW6HPAK4Mcjs.encodeExecutionCalldata.call(void 0, executions)]
  });
};

// src/DelegationFramework/DeleGatorCore/encode.ts
var encode_exports4 = {};
_chunkW6HPAK4Mcjs.__export.call(void 0, encode_exports4, {
  disableDelegation: () => encode17,
  enableDelegation: () => encode18,
  execute: () => encode15,
  executeWithMode: () => encode16,
  isValidSignature: () => encode,
  upgradeToAndCall: () => encode19
});

// src/DelegationFramework/DeleGatorCore/methods/disableDelegation.ts



var simulate16 = async ({
  client,
  delegationManagerAddress,
  delegation
}) => {
  const delegationStruct = _chunkW6HPAK4Mcjs.toDelegationStruct.call(void 0, delegation);
  return _actions.simulateContract.call(void 0, client, {
    address: delegationManagerAddress,
    abi: _delegationabis.DeleGatorCore,
    functionName: "disableDelegation",
    args: [delegationStruct]
  });
};
var execute4 = async ({
  client,
  delegationManagerAddress,
  delegation
}) => {
  const { request } = await simulate16({
    client,
    delegationManagerAddress,
    delegation
  });
  return _actions.writeContract.call(void 0, client, request);
};
var encode17 = ({ delegation }) => {
  const delegationStruct = _chunkW6HPAK4Mcjs.toDelegationStruct.call(void 0, delegation);
  return _viem.encodeFunctionData.call(void 0, {
    abi: _delegationabis.DeleGatorCore,
    functionName: "disableDelegation",
    args: [delegationStruct]
  });
};

// src/DelegationFramework/DeleGatorCore/methods/enableDelegation.ts



var simulate17 = async ({
  client,
  delegationManagerAddress,
  delegation
}) => {
  const delegationStruct = _chunkW6HPAK4Mcjs.toDelegationStruct.call(void 0, delegation);
  return _actions.simulateContract.call(void 0, client, {
    address: delegationManagerAddress,
    abi: _delegationabis.DeleGatorCore,
    functionName: "enableDelegation",
    args: [delegationStruct]
  });
};
var execute5 = async ({
  client,
  delegationManagerAddress,
  delegation
}) => {
  const { request } = await simulate17({
    client,
    delegationManagerAddress,
    delegation
  });
  return _actions.writeContract.call(void 0, client, request);
};
var encode18 = ({ delegation }) => {
  const delegationStruct = _chunkW6HPAK4Mcjs.toDelegationStruct.call(void 0, delegation);
  return _viem.encodeFunctionData.call(void 0, {
    abi: _delegationabis.DeleGatorCore,
    functionName: "enableDelegation",
    args: [delegationStruct]
  });
};

// src/DelegationFramework/DeleGatorCore/methods/upgradeToAndCall.ts



var simulate18 = async ({
  client,
  contractAddress,
  implementation,
  data
}) => {
  return _actions.simulateContract.call(void 0, client, {
    address: contractAddress,
    abi: _delegationabis.DeleGatorCore,
    functionName: "upgradeToAndCall",
    args: [implementation, data]
  });
};
var execute6 = async ({
  client,
  contractAddress,
  implementation,
  data
}) => {
  const { request } = await simulate18({
    client,
    contractAddress,
    implementation,
    data
  });
  return _actions.writeContract.call(void 0, client, request);
};
var encode19 = ({
  implementation,
  data
}) => {
  return _viem.encodeFunctionData.call(void 0, {
    abi: _delegationabis.DeleGatorCore,
    functionName: "upgradeToAndCall",
    args: [implementation, data]
  });
};









































exports.simulate = simulate2; exports.simulate2 = simulate; exports.encode = encode2; exports.simulate3 = simulate3; exports.simulate4 = simulate4; exports.simulate5 = simulate5; exports.encode_exports = encode_exports; exports.simulate6 = simulate7; exports.simulate7 = simulate6; exports.encode2 = encode7; exports.simulate8 = simulate8; exports.simulate9 = simulate9; exports.simulate10 = simulate10; exports.simulate11 = simulate11; exports.simulate12 = simulate12; exports.encode_exports2 = encode_exports2; exports.simulate13 = simulate13; exports.encode3 = encode14; exports.execute = execute; exports.encode_exports3 = encode_exports3; exports.read_exports = read_exports; exports.isContractDeployed = isContractDeployed; exports.isImplementationExpected = isImplementationExpected; exports.encodeProxyCreationCode = encodeProxyCreationCode; exports.simulate14 = simulate16; exports.execute2 = execute4; exports.simulate15 = simulate17; exports.execute3 = execute5; exports.simulate16 = simulate14; exports.execute4 = execute2; exports.encode4 = encode15; exports.simulate17 = simulate15; exports.execute5 = execute3; exports.encode5 = encode16; exports.simulate18 = simulate18; exports.execute6 = execute6; exports.encode_exports4 = encode_exports4; exports.decodeRevertReason = decodeRevertReason; exports.decodeRevertData = decodeRevertData;
