// @metamask/smart-accounts-kit/dist/index.cjs
"use strict";Object.defineProperty(exports, "__esModule", {value: true}); function _nullishCoalesce(lhs, rhsFn) { if (lhs != null) { return lhs; } else { return rhsFn(); } } function _optionalChain(ops) { let lastAccessLHS = undefined; let value = ops[0]; let i = 1; while (i < ops.length) { const op = ops[i]; const fn = ops[i + 1]; i += 2; if ((op === 'optionalAccess' || op === 'optionalCall') && value == null) { return undefined; } if (op === 'access' || op === 'optionalAccess') { lastAccessLHS = value; value = fn(value); } else if (op === 'call' || op === 'optionalCall') { value = fn((...args) => value.call(lastAccessLHS, ...args)); lastAccessLHS = undefined; } } return value; }



var _chunk2ZOWIAP3cjs = require('./chunk-2ZOWIAP3.cjs');


var _chunkEPU4O7FYcjs = require('./chunk-EPU4O7FY.cjs');
require('./chunk-PQPAEHJZ.cjs');





var _chunkDCI2AN4Ncjs = require('./chunk-DCI2AN4N.cjs');





var _chunk2OS36DVVcjs = require('./chunk-2OS36DVV.cjs');
require('./chunk-JKVSTPEC.cjs');
















var _chunkW6HPAK4Mcjs = require('./chunk-W6HPAK4M.cjs');

// src/toMetaMaskSmartAccount.ts




var _delegationabis = require('@metamask/delegation-abis');




var _accountabstraction = require('viem/account-abstraction');

// src/signer.ts
var _viem = require('viem');

// src/signatures.ts

var signatureTypes = ["ECDSA"];
var aggregateSignature = ({
  signatures
}) => {
  _chunkW6HPAK4Mcjs.trackSmartAccountsKitFunctionCall.call(void 0, "aggregateSignature", {
    signatureCount: signatures.length
  });
  if (signatures.length === 0) {
    return "0x";
  }
  for (const { type } of signatures) {
    if (!signatureTypes.includes(type)) {
      throw new Error(`Invalid signature type: ${type}`);
    }
  }
  const sortedSignatures = [...signatures].sort(
    (a, b) => a.signer.localeCompare(b.signer)
  );
  return _viem.concat.call(void 0, sortedSignatures.map(({ signature }) => signature));
};

// src/signer.ts
var EOA_STUB_SIGNATURE = "0x000000000000000000000000000000000000000000000000000000000000000100000000000000000000000000000000000000000000000000000000000000011b";
var resolveSignerFromWalletConfig = (config) => {
  return {
    signMessage: config.walletClient.signMessage,
    signTypedData: async (typedData) => {
      return config.walletClient.signTypedData(typedData);
    },
    getStubSignature: async () => EOA_STUB_SIGNATURE
  };
};
var resolveSignerFromAccountConfig = (config) => {
  return {
    signMessage: config.account.signMessage,
    signTypedData: config.account.signTypedData,
    getStubSignature: async () => EOA_STUB_SIGNATURE
  };
};
var resolveHybridSigner = (config) => {
  if ("walletClient" in config) {
    return resolveSignerFromWalletConfig(config);
  } else if ("account" in config) {
    const { signMessage: signMessage2, signTypedData: signTypedData2, getStubSignature: getStubSignature2 } = resolveSignerFromAccountConfig(config);
    if (!signMessage2) {
      throw new Error("Account does not support signMessage");
    }
    if (!signTypedData2) {
      throw new Error("Account does not support signTypedData");
    }
    return {
      signMessage: signMessage2,
      signTypedData: signTypedData2,
      getStubSignature: getStubSignature2
    };
  }
  const { keyId, webAuthnAccount } = config;
  if (webAuthnAccount.type !== "webAuthn") {
    throw new Error("Account is not a webAuthn account");
  }
  const encodeSignature = (result) => _chunkDCI2AN4Ncjs.encodeDeleGatorSignature.call(void 0, 
    keyId,
    result.signature,
    result.webauthn.clientDataJSON,
    result.webauthn.authenticatorData
  );
  const signMessage = async (args) => webAuthnAccount.signMessage(args).then(encodeSignature);
  const signTypedData = async (typedDataDefinition) => webAuthnAccount.signTypedData(typedDataDefinition).then(encodeSignature);
  const getStubSignature = async () => _chunkDCI2AN4Ncjs.createDummyWebAuthnSignature.call(void 0, keyId);
  return {
    signMessage,
    signTypedData,
    getStubSignature
  };
};
var resolveMultiSigSigner = (config) => {
  const resolvedSigners = config.map((signer) => {
    let individualSignMessage;
    let individualSignTypedData;
    let address;
    if ("walletClient" in signer) {
      const { signMessage: signMessage2, signTypedData: signTypedData2 } = resolveSignerFromWalletConfig(signer);
      individualSignMessage = signMessage2;
      individualSignTypedData = signTypedData2;
      address = signer.walletClient.account.address;
    } else {
      const { signMessage: signMessage2, signTypedData: signTypedData2 } = resolveSignerFromAccountConfig(signer);
      if (!signMessage2) {
        throw new Error("Account does not support signMessage");
      }
      if (!signTypedData2) {
        throw new Error("Account does not support signTypedData");
      }
      individualSignMessage = signMessage2;
      individualSignTypedData = signTypedData2;
      address = signer.account.address;
    }
    return {
      address,
      individualSignMessage,
      individualSignTypedData
    };
  });
  const signMessage = async (args) => {
    const addressAndSignatures = resolvedSigners.map(
      async ({ individualSignMessage, address }) => ({
        signature: await individualSignMessage(args),
        signer: address,
        type: "ECDSA"
      })
    );
    const signatures = await Promise.all(addressAndSignatures);
    return aggregateSignature({
      signatures
    });
  };
  const signTypedData = async (typedDataDefinition) => {
    const addressAndSignatures = resolvedSigners.map(
      async ({ individualSignTypedData, address }) => ({
        signature: await individualSignTypedData(typedDataDefinition),
        signer: address,
        type: "ECDSA"
      })
    );
    const signatures = await Promise.all(addressAndSignatures);
    return aggregateSignature({
      signatures
    });
  };
  const getStubSignature = async () => _viem.concat.call(void 0, resolvedSigners.map(() => EOA_STUB_SIGNATURE));
  return {
    signMessage,
    signTypedData,
    getStubSignature
  };
};
var resolveStateless7702Signer = (config) => {
  if ("walletClient" in config) {
    return resolveSignerFromWalletConfig(config);
  } else if ("account" in config) {
    const { signMessage, signTypedData, getStubSignature } = resolveSignerFromAccountConfig(config);
    if (!signMessage) {
      throw new Error("Account does not support signMessage");
    }
    if (!signTypedData) {
      throw new Error("Account does not support signTypedData");
    }
    return {
      signMessage,
      signTypedData,
      getStubSignature
    };
  }
  throw new Error("Invalid signer config");
};
function resolveSigner(config) {
  const { implementation, signer } = config;
  if (!signer) {
    return null;
  }
  if (implementation === "Hybrid" /* Hybrid */) {
    return resolveHybridSigner(signer);
  } else if (implementation === "MultiSig" /* MultiSig */) {
    return resolveMultiSigSigner(signer);
  } else if (implementation === "Stateless7702" /* Stateless7702 */) {
    return resolveStateless7702Signer(signer);
  }
  throw new Error(`Implementation type '${implementation}' not supported`);
}

// src/toMetaMaskSmartAccount.ts
var ENTRYPOINT_VERSION = "0.7";
async function toMetaMaskSmartAccount(params) {
  const {
    client,
    client: { chain },
    implementation,
    nonceKeyManager
  } = params;
  _chunkW6HPAK4Mcjs.trackSmartAccountsKitFunctionCall.call(void 0, "toMetaMaskSmartAccount", {
    implementation,
    hasAddress: params.address !== void 0,
    hasEnvironment: params.environment !== void 0,
    chainId: _nullishCoalesce(_optionalChain([chain, 'optionalAccess', _ => _.id]), () => ( null))
  });
  if (!chain) {
    throw new Error("Chain not specified");
  }
  const signer = resolveSigner({
    implementation,
    signer: params.signer
  });
  const environment = _nullishCoalesce(params.environment, () => ( _chunk2OS36DVVcjs.getSmartAccountsEnvironment.call(void 0, chain.id)));
  let address, factoryData;
  if (params.address) {
    factoryData = void 0;
    address = params.address;
  } else {
    if (implementation === "Stateless7702" /* Stateless7702 */) {
      throw new Error("Stateless7702 does not support counterfactual accounts");
    }
    const accountData = await _chunkDCI2AN4Ncjs.getCounterfactualAccountData.call(void 0, {
      factory: environment.SimpleFactory,
      implementations: environment.implementations,
      implementation,
      deployParams: params.deployParams,
      deploySalt: params.deploySalt
    });
    address = accountData.address;
    factoryData = accountData.factoryData;
  }
  const entryPoint = {
    abi: _accountabstraction.entryPoint07Abi,
    address: environment.EntryPoint,
    version: ENTRYPOINT_VERSION
  };
  const { abi, contractName } = {
    ["Hybrid" /* Hybrid */]: {
      contractName: "HybridDeleGator",
      abi: _delegationabis.HybridDeleGator
    },
    ["MultiSig" /* MultiSig */]: {
      contractName: "MultiSigDeleGator",
      abi: _delegationabis.MultiSigDeleGator
    },
    ["Stateless7702" /* Stateless7702 */]: {
      contractName: "EIP7702StatelessDeleGator",
      abi: _delegationabis.EIP7702StatelessDeleGator
    }
  }[implementation];
  const getFactoryArgs = async () => {
    if (factoryData === void 0) {
      throw new Error(
        "Deploy params were not provided, so factory args cannot be inferred"
      );
    }
    return {
      factoryData,
      factory: environment.SimpleFactory
    };
  };
  const signDelegation2 = async (delegationParams) => {
    if (!signer) {
      throw new Error(
        "Cannot sign delegation: signer not provided. Specify a signer in toMetaMaskSmartAccount() to perform signing operations."
      );
    }
    const { delegation, chainId } = delegationParams;
    const delegationStruct = _chunkW6HPAK4Mcjs.toDelegationStruct.call(void 0, {
      ...delegation,
      signature: "0x"
    });
    const signature = await signer.signTypedData({
      domain: {
        chainId: _nullishCoalesce(chainId, () => ( chain.id)),
        name: "DelegationManager",
        version: "1",
        verifyingContract: environment.DelegationManager
      },
      types: _chunkW6HPAK4Mcjs.SIGNABLE_DELEGATION_TYPED_DATA,
      primaryType: "Delegation",
      message: delegationStruct
    });
    return signature;
  };
  const signUserOperation2 = async (userOpParams) => {
    if (!signer) {
      throw new Error(
        "Cannot sign user operation: signer not provided. Specify a signer in toMetaMaskSmartAccount() to perform signing operations."
      );
    }
    const { chainId } = userOpParams;
    const packedUserOp = _accountabstraction.toPackedUserOperation.call(void 0, {
      sender: address,
      ...userOpParams,
      signature: "0x"
    });
    const signature = await signer.signTypedData({
      domain: {
        chainId: _nullishCoalesce(chainId, () => ( chain.id)),
        name: contractName,
        version: "1",
        verifyingContract: address
      },
      types: _chunk2OS36DVVcjs.SIGNABLE_USER_OP_TYPED_DATA,
      primaryType: "PackedUserOperation",
      message: { ...packedUserOp, entryPoint: entryPoint.address }
    });
    return signature;
  };
  const getAddress = async () => address;
  const encodeCalls = async (calls) => _chunkDCI2AN4Ncjs.encodeCallsForCaller.call(void 0, address, calls);
  const signerMethods = _nullishCoalesce(signer, () => ( {
    signMessage: async () => {
      throw new Error(
        "Cannot sign message: signer not provided. Specify a signer in toMetaMaskSmartAccount() to perform signing operations."
      );
    },
    signTypedData: async () => {
      throw new Error(
        "Cannot sign typed data: signer not provided. Specify a signer in toMetaMaskSmartAccount() to perform signing operations."
      );
    },
    getStubSignature: async () => {
      throw new Error(
        "Cannot get stub signature: signer not provided. Specify a signer in toMetaMaskSmartAccount() to perform signing operations."
      );
    }
  }));
  const smartAccount = await _accountabstraction.toSmartAccount.call(void 0, {
    abi,
    client,
    entryPoint,
    environment,
    getAddress,
    getFactoryArgs,
    encodeCalls,
    signUserOperation: signUserOperation2,
    signDelegation: signDelegation2,
    nonceKeyManager,
    ...signerMethods
  });
  if (implementation === "Stateless7702" /* Stateless7702 */) {
    return {
      ...smartAccount,
      isDeployed: async () => _chunk2ZOWIAP3cjs.isValid7702Implementation.call(void 0, {
        client,
        accountAddress: address,
        environment
      })
    };
  }
  return smartAccount;
}

// src/index.ts
var _delegationcore = require('@metamask/delegation-core');

// src/actions/caveatEnforcerClient.ts
function createCaveatEnforcerClient({
  client,
  environment
}) {
  _chunkW6HPAK4Mcjs.trackSmartAccountsKitFunctionCall.call(void 0, "createCaveatEnforcerClient", {
    chainId: _nullishCoalesce(_optionalChain([client, 'access', _2 => _2.chain, 'optionalAccess', _3 => _3.id]), () => ( null))
  });
  return client.extend(_chunk2ZOWIAP3cjs.caveatEnforcerActions.call(void 0, { environment }));
}

// src/actions/infuraBundlerClient.ts



var infuraBundlerActions = () => (client) => ({
  /**
   * Get user operation gas prices from Infura bundler.
   * Calls the pimlico_getUserOperationGasPrice RPC method.
   *
   * @returns Promise resolving to gas price tiers (slow, standard, fast).
   * @example
   * ```typescript
   * const gasPrices = await bundlerClient.getUserOperationGasPrice();
   * console.log(gasPrices.standard.maxFeePerGas);
   * ```
   */
  async getUserOperationGasPrice() {
    const pimlicoClient = client;
    return await pimlicoClient.request({
      method: "pimlico_getUserOperationGasPrice",
      params: []
    });
  }
});
function createInfuraBundlerClient(config) {
  _chunkW6HPAK4Mcjs.trackSmartAccountsKitFunctionCall.call(void 0, "createInfuraBundlerClient", {
    chainId: _nullishCoalesce(_optionalChain([config, 'access', _4 => _4.chain, 'optionalAccess', _5 => _5.id]), () => ( null))
  });
  const baseBundlerClient = _accountabstraction.createBundlerClient.call(void 0, config);
  return baseBundlerClient.extend(
    infuraBundlerActions()
  );
}























exports.ANY_BENEFICIARY = _chunkW6HPAK4Mcjs.ANY_BENEFICIARY; exports.BalanceChangeType = _delegationcore.BalanceChangeType; exports.CaveatType = _chunkW6HPAK4Mcjs.CaveatType; exports.ExecutionMode = _chunkW6HPAK4Mcjs.ExecutionMode; exports.Implementation = _chunkW6HPAK4Mcjs.Implementation; exports.PREFERRED_VERSION = _chunk2OS36DVVcjs.PREFERRED_VERSION; exports.ROOT_AUTHORITY = _chunkW6HPAK4Mcjs.ROOT_AUTHORITY; exports.ScopeType = _chunkW6HPAK4Mcjs.ScopeType; exports.TransferWindow = _chunkW6HPAK4Mcjs.TransferWindow; exports.actions = _chunk2ZOWIAP3cjs.actions_exports; exports.aggregateSignature = aggregateSignature; exports.contracts = _chunkEPU4O7FYcjs.contracts_exports; exports.createCaveat = _chunkW6HPAK4Mcjs.createCaveat; exports.createCaveatEnforcerClient = createCaveatEnforcerClient; exports.createDelegation = _chunkW6HPAK4Mcjs.createDelegation; exports.createExecution = _chunkW6HPAK4Mcjs.createExecution; exports.createInfuraBundlerClient = createInfuraBundlerClient; exports.createOpenDelegation = _chunkW6HPAK4Mcjs.createOpenDelegation; exports.getSmartAccountsEnvironment = _chunk2OS36DVVcjs.getSmartAccountsEnvironment; exports.signDelegation = _chunkW6HPAK4Mcjs.signDelegation; exports.signUserOperation = _chunk2OS36DVVcjs.signUserOperation; exports.toMetaMaskSmartAccount = toMetaMaskSmartAccount;
