// @metamask/delegation-core/dist/index.mjs
// src/caveats/types.ts
var BalanceChangeType = /* @__PURE__ */ ((BalanceChangeType2) => {
  BalanceChangeType2[BalanceChangeType2["Increase"] = 0] = "Increase";
  BalanceChangeType2[BalanceChangeType2["Decrease"] = 1] = "Decrease";
  return BalanceChangeType2;
})(BalanceChangeType || {});

// src/internalUtils.ts
import {
  bytesToHex,
  hexToBytes,
  isHexString,
  remove0x
} from "@metamask/utils";
var toHexString = ({
  value,
  size
}) => {
  return value.toString(16).padStart(size * 2, "0");
};
var normalizeHex = (value, errorMessage) => {
  if (typeof value === "string") {
    if (!isHexString(value)) {
      throw new Error(errorMessage);
    }
    return value;
  }
  return bytesToHex(value);
};
var normalizeAddress = (value, errorMessage) => {
  if (typeof value === "string") {
    if (!isHexString(value) || value.length !== 42) {
      throw new Error(errorMessage);
    }
    return value;
  }
  if (value.length !== 20) {
    throw new Error(errorMessage);
  }
  return bytesToHex(value);
};
var normalizeAddressLowercase = (value, errorMessage) => {
  if (typeof value === "string") {
    if (!isHexString(value) || value.length !== 42) {
      throw new Error(errorMessage);
    }
    return bytesToHex(hexToBytes(value));
  }
  if (value.length !== 20) {
    throw new Error(errorMessage);
  }
  return bytesToHex(value);
};
var concatHex = (parts) => {
  return `0x${parts.map(remove0x).join("")}`;
};
var extractBigInt = (value, offset, size) => {
  const start = 2 + offset * 2;
  const end = start + size * 2;
  const slice = value.slice(start, end);
  return BigInt(`0x${slice}`);
};
var extractNumber = (value, offset, size) => {
  const bigIntValue = extractBigInt(value, offset, size);
  if (bigIntValue > Number.MAX_SAFE_INTEGER) {
    throw new Error("Number is too large");
  }
  return Number(bigIntValue);
};
var extractAddress = (value, offset) => {
  const start = 2 + offset * 2;
  const end = start + 40;
  return `0x${value.slice(start, end)}`;
};
var extractHex = (value, offset, size) => {
  const start = 2 + offset * 2;
  const end = start + size * 2;
  return `0x${value.slice(start, end)}`;
};
var extractRemainingHex = (value, offset) => {
  const start = 2 + offset * 2;
  return `0x${value.slice(start)}`;
};
function getByteLength(value) {
  return (value.length - 2) / 2;
}
function assertHexByteExactLength(hexTerms, expectedBytes, errorMessage) {
  if (getByteLength(hexTerms) !== expectedBytes) {
    throw new Error(errorMessage);
  }
}
function assertHexByteLengthAtLeastOneMultipleOf(hexTerms, unitBytes, errorMessage) {
  const byteLength = getByteLength(hexTerms);
  if (byteLength === 0 || byteLength % unitBytes !== 0) {
    throw new Error(errorMessage);
  }
}
function assertHexBytesMinLength(hexTerms, minBytes, errorMessage) {
  if (getByteLength(hexTerms) < minBytes) {
    throw new Error(errorMessage);
  }
}

// src/returns.ts
import { bytesToHex as bytesToHex2, hexToBytes as hexToBytes2 } from "@metamask/utils";
var defaultOptions = { out: "hex" };
function prepareResult(result, options) {
  if (options.out === "hex") {
    const hexValue = typeof result === "string" ? result : bytesToHex2(result);
    return hexValue.startsWith("0x") ? hexValue : `0x${hexValue}`;
  }
  const bytesValue = result instanceof Uint8Array ? result : hexToBytes2(result);
  return bytesValue;
}
var bytesLikeToHex = (bytesLike) => {
  if (typeof bytesLike === "string") {
    return bytesLike;
  }
  return bytesToHex2(bytesLike);
};
var bytesLikeToBytes = (bytesLike) => {
  if (typeof bytesLike === "string") {
    return hexToBytes2(bytesLike);
  }
  return bytesLike;
};

// src/caveats/valueLte.ts
function createValueLteTerms(terms, options = defaultOptions) {
  const { maxValue } = terms;
  if (maxValue < 0n) {
    throw new Error("Invalid maxValue: must be greater than or equal to zero");
  }
  const hexValue = toHexString({ value: maxValue, size: 32 });
  return prepareResult(hexValue, options);
}
function decodeValueLteTerms(terms) {
  const hexTerms = bytesLikeToHex(terms);
  assertHexByteExactLength(
    hexTerms,
    32,
    "Invalid ValueLte terms: must be exactly 32 bytes"
  );
  const maxValue = extractBigInt(hexTerms, 0, 32);
  return { maxValue };
}

// src/caveats/timestamp.ts
var TIMESTAMP_UPPER_BOUND_SECONDS = 253402300799;
function createTimestampTerms(terms, encodingOptions = defaultOptions) {
  const { afterThreshold, beforeThreshold } = terms;
  if (afterThreshold < 0) {
    throw new Error("Invalid afterThreshold: must be zero or positive");
  }
  if (beforeThreshold < 0) {
    throw new Error("Invalid beforeThreshold: must be zero or positive");
  }
  if (beforeThreshold > TIMESTAMP_UPPER_BOUND_SECONDS) {
    throw new Error(
      `Invalid beforeThreshold: must be less than or equal to ${TIMESTAMP_UPPER_BOUND_SECONDS}`
    );
  }
  if (afterThreshold > TIMESTAMP_UPPER_BOUND_SECONDS) {
    throw new Error(
      `Invalid afterThreshold: must be less than or equal to ${TIMESTAMP_UPPER_BOUND_SECONDS}`
    );
  }
  if (beforeThreshold !== 0 && afterThreshold >= beforeThreshold) {
    throw new Error(
      "Invalid thresholds: beforeThreshold must be greater than afterThreshold when both are specified"
    );
  }
  const afterThresholdHex = toHexString({
    value: afterThreshold,
    size: 16
  });
  const beforeThresholdHex = toHexString({
    value: beforeThreshold,
    size: 16
  });
  const hexValue = `0x${afterThresholdHex}${beforeThresholdHex}`;
  return prepareResult(hexValue, encodingOptions);
}
function decodeTimestampTerms(terms) {
  const hexTerms = bytesLikeToHex(terms);
  assertHexByteExactLength(
    hexTerms,
    32,
    "Invalid Timestamp terms: must be exactly 32 bytes"
  );
  const afterThreshold = extractNumber(hexTerms, 0, 16);
  const beforeThreshold = extractNumber(hexTerms, 16, 16);
  return { afterThreshold, beforeThreshold };
}

// src/caveats/nativeTokenPeriodTransfer.ts
function createNativeTokenPeriodTransferTerms(terms, encodingOptions = defaultOptions) {
  const { periodAmount, periodDuration, startDate } = terms;
  if (periodAmount <= 0n) {
    throw new Error("Invalid periodAmount: must be a positive number");
  }
  if (periodDuration <= 0) {
    throw new Error("Invalid periodDuration: must be a positive number");
  }
  if (startDate <= 0) {
    throw new Error("Invalid startDate: must be a positive number");
  }
  const periodAmountHex = toHexString({ value: periodAmount, size: 32 });
  const periodDurationHex = toHexString({ value: periodDuration, size: 32 });
  const startDateHex = toHexString({ value: startDate, size: 32 });
  const hexValue = `0x${periodAmountHex}${periodDurationHex}${startDateHex}`;
  return prepareResult(hexValue, encodingOptions);
}
function decodeNativeTokenPeriodTransferTerms(terms) {
  const hexTerms = bytesLikeToHex(terms);
  assertHexByteExactLength(
    hexTerms,
    96,
    "Invalid NativeTokenPeriodTransfer terms: must be exactly 96 bytes"
  );
  const periodAmount = extractBigInt(hexTerms, 0, 32);
  const periodDuration = extractNumber(hexTerms, 32, 32);
  const startDate = extractNumber(hexTerms, 64, 32);
  return { periodAmount, periodDuration, startDate };
}

// src/caveats/exactCalldata.ts
function createExactCalldataTerms(terms, encodingOptions = defaultOptions) {
  const { calldata } = terms;
  if (calldata === void 0 || calldata === null) {
    throw new Error("Invalid calldata: calldata is required");
  }
  if (typeof calldata === "string" && !calldata.startsWith("0x")) {
    throw new Error("Invalid calldata: must be a hex string starting with 0x");
  }
  return prepareResult(calldata, encodingOptions);
}
function decodeExactCalldataTerms(terms, encodingOptions = defaultOptions) {
  const calldataHex = bytesLikeToHex(terms);
  const calldata = prepareResult(calldataHex, encodingOptions);
  return { calldata };
}

// src/caveats/exactCalldataBatch.ts
import { decodeSingle, encodeSingle } from "@metamask/abi-utils";
import { bytesToHex as bytesToHex3 } from "@metamask/utils";
var EXECUTION_ARRAY_ABI = "(address,uint256,bytes)[]";
function createExactCalldataBatchTerms(terms, encodingOptions = defaultOptions) {
  const { executions } = terms;
  if (executions.length === 0) {
    throw new Error("Invalid executions: array cannot be empty");
  }
  const encodableExecutions = executions.map((execution) => {
    const targetHex = normalizeAddress(
      execution.target,
      "Invalid target: must be a valid address"
    );
    if (execution.value < 0n) {
      throw new Error("Invalid value: must be a non-negative number");
    }
    let callDataHex;
    if (typeof execution.callData === "string") {
      if (!execution.callData.startsWith("0x")) {
        throw new Error(
          "Invalid calldata: must be a hex string starting with 0x"
        );
      }
      callDataHex = execution.callData;
    } else {
      callDataHex = bytesToHex3(execution.callData);
    }
    return [targetHex, execution.value, callDataHex];
  });
  const hexValue = encodeSingle(EXECUTION_ARRAY_ABI, encodableExecutions);
  return prepareResult(hexValue, encodingOptions);
}
function decodeExactCalldataBatchTerms(terms, encodingOptions = defaultOptions) {
  const hexTerms = bytesLikeToHex(terms);
  const decoded = decodeSingle(EXECUTION_ARRAY_ABI, hexTerms);
  const executions = decoded.map(
    ([target, value, callData]) => ({
      target: prepareResult(target, encodingOptions),
      value,
      callData: prepareResult(bytesToHex3(callData), encodingOptions)
    })
  );
  return { executions };
}

// src/caveats/exactExecution.ts
import { bytesToHex as bytesToHex4 } from "@metamask/utils";
function createExactExecutionTerms(terms, encodingOptions = defaultOptions) {
  const { execution } = terms;
  const targetHex = normalizeAddress(
    execution.target,
    "Invalid target: must be a valid address"
  );
  if (execution.value < 0n) {
    throw new Error("Invalid value: must be a non-negative number");
  }
  let callDataHex;
  if (typeof execution.callData === "string") {
    if (!execution.callData.startsWith("0x")) {
      throw new Error(
        "Invalid calldata: must be a hex string starting with 0x"
      );
    }
    callDataHex = execution.callData;
  } else {
    callDataHex = bytesToHex4(execution.callData);
  }
  const valueHex = `0x${toHexString({ value: execution.value, size: 32 })}`;
  const hexValue = concatHex([targetHex, valueHex, callDataHex]);
  return prepareResult(hexValue, encodingOptions);
}
function decodeExactExecutionTerms(terms, encodingOptions = defaultOptions) {
  const hexTerms = bytesLikeToHex(terms);
  assertHexBytesMinLength(
    hexTerms,
    52,
    "Invalid ExactExecution terms: must be at least 52 bytes"
  );
  const targetHex = extractAddress(hexTerms, 0);
  const value = extractBigInt(hexTerms, 20, 32);
  const callDataHex = extractRemainingHex(hexTerms, 52);
  return {
    execution: {
      target: prepareResult(targetHex, encodingOptions),
      value,
      callData: prepareResult(callDataHex, encodingOptions)
    }
  };
}

// src/caveats/exactExecutionBatch.ts
import { decodeSingle as decodeSingle2, encodeSingle as encodeSingle2 } from "@metamask/abi-utils";
import { bytesToHex as bytesToHex5 } from "@metamask/utils";
var EXECUTION_ARRAY_ABI2 = "(address,uint256,bytes)[]";
function createExactExecutionBatchTerms(terms, encodingOptions = defaultOptions) {
  const { executions } = terms;
  if (executions.length === 0) {
    throw new Error("Invalid executions: array cannot be empty");
  }
  const encodableExecutions = executions.map((execution) => {
    const targetHex = normalizeAddress(
      execution.target,
      "Invalid target: must be a valid address"
    );
    if (execution.value < 0n) {
      throw new Error("Invalid value: must be a non-negative number");
    }
    let callDataHex;
    if (typeof execution.callData === "string") {
      if (!execution.callData.startsWith("0x")) {
        throw new Error(
          "Invalid calldata: must be a hex string starting with 0x"
        );
      }
      callDataHex = execution.callData;
    } else {
      callDataHex = bytesToHex5(execution.callData);
    }
    return [targetHex, execution.value, callDataHex];
  });
  const hexValue = encodeSingle2(EXECUTION_ARRAY_ABI2, encodableExecutions);
  return prepareResult(hexValue, encodingOptions);
}
function decodeExactExecutionBatchTerms(terms, encodingOptions = defaultOptions) {
  const hexTerms = bytesLikeToHex(terms);
  const decoded = decodeSingle2(EXECUTION_ARRAY_ABI2, hexTerms);
  const executions = decoded.map(
    ([target, value, callData]) => ({
      target: prepareResult(target, encodingOptions),
      value,
      callData: prepareResult(bytesToHex5(callData), encodingOptions)
    })
  );
  return { executions };
}

// src/caveats/nativeTokenStreaming.ts
var TIMESTAMP_UPPER_BOUND_SECONDS2 = 253402300799;
function createNativeTokenStreamingTerms(terms, encodingOptions = defaultOptions) {
  const { initialAmount, maxAmount, amountPerSecond, startTime } = terms;
  if (initialAmount < 0n) {
    throw new Error("Invalid initialAmount: must be greater than zero");
  }
  if (maxAmount <= 0n) {
    throw new Error("Invalid maxAmount: must be a positive number");
  }
  if (maxAmount < initialAmount) {
    throw new Error("Invalid maxAmount: must be greater than initialAmount");
  }
  if (amountPerSecond <= 0n) {
    throw new Error("Invalid amountPerSecond: must be a positive number");
  }
  if (startTime <= 0) {
    throw new Error("Invalid startTime: must be a positive number");
  }
  if (startTime > TIMESTAMP_UPPER_BOUND_SECONDS2) {
    throw new Error(
      "Invalid startTime: must be less than or equal to 253402300799"
    );
  }
  const initialAmountHex = toHexString({ value: initialAmount, size: 32 });
  const maxAmountHex = toHexString({ value: maxAmount, size: 32 });
  const amountPerSecondHex = toHexString({ value: amountPerSecond, size: 32 });
  const startTimeHex = toHexString({ value: startTime, size: 32 });
  const hexValue = `0x${initialAmountHex}${maxAmountHex}${amountPerSecondHex}${startTimeHex}`;
  return prepareResult(hexValue, encodingOptions);
}
function decodeNativeTokenStreamingTerms(terms) {
  const hexTerms = bytesLikeToHex(terms);
  assertHexByteExactLength(
    hexTerms,
    128,
    "Invalid NativeTokenStreaming terms: must be exactly 128 bytes"
  );
  const initialAmount = extractBigInt(hexTerms, 0, 32);
  const maxAmount = extractBigInt(hexTerms, 32, 32);
  const amountPerSecond = extractBigInt(hexTerms, 64, 32);
  const startTime = extractNumber(hexTerms, 96, 32);
  return { initialAmount, maxAmount, amountPerSecond, startTime };
}

// src/caveats/nativeTokenTransferAmount.ts
function createNativeTokenTransferAmountTerms(terms, encodingOptions = defaultOptions) {
  const { maxAmount } = terms;
  if (maxAmount < 0n) {
    throw new Error("Invalid maxAmount: must be zero or positive");
  }
  const hexValue = `0x${toHexString({ value: maxAmount, size: 32 })}`;
  return prepareResult(hexValue, encodingOptions);
}
function decodeNativeTokenTransferAmountTerms(terms) {
  const hexTerms = bytesLikeToHex(terms);
  assertHexByteExactLength(
    hexTerms,
    32,
    "Invalid NativeTokenTransferAmount terms: must be exactly 32 bytes"
  );
  const maxAmount = extractBigInt(hexTerms, 0, 32);
  return { maxAmount };
}

// src/caveats/nativeTokenPayment.ts
function createNativeTokenPaymentTerms(terms, encodingOptions = defaultOptions) {
  const { recipient, amount } = terms;
  const recipientHex = normalizeAddressLowercase(
    recipient,
    "Invalid recipient: must be a valid address"
  );
  if (amount <= 0n) {
    throw new Error("Invalid amount: must be positive");
  }
  const amountHex = `0x${toHexString({ value: amount, size: 32 })}`;
  const hexValue = concatHex([recipientHex, amountHex]);
  return prepareResult(hexValue, encodingOptions);
}
function decodeNativeTokenPaymentTerms(terms, encodingOptions = defaultOptions) {
  const hexTerms = bytesLikeToHex(terms);
  assertHexByteExactLength(
    hexTerms,
    52,
    "Invalid NativeTokenPayment terms: must be exactly 52 bytes"
  );
  const recipientHex = extractAddress(hexTerms, 0);
  const amount = extractBigInt(hexTerms, 20, 32);
  return {
    recipient: prepareResult(recipientHex, encodingOptions),
    amount
  };
}

// src/caveats/nativeBalanceChange.ts
function createNativeBalanceChangeTerms(terms, encodingOptions = defaultOptions) {
  const { recipient, balance, changeType } = terms;
  const recipientHex = normalizeAddressLowercase(
    recipient,
    "Invalid recipient: must be a valid Address"
  );
  if (balance <= 0n) {
    throw new Error("Invalid balance: must be a positive number");
  }
  if (changeType !== 0 /* Increase */ && changeType !== 1 /* Decrease */) {
    throw new Error("Invalid changeType: must be either Increase or Decrease");
  }
  const changeTypeHex = `0x${toHexString({ value: changeType, size: 1 })}`;
  const balanceHex = `0x${toHexString({ value: balance, size: 32 })}`;
  const hexValue = concatHex([changeTypeHex, recipientHex, balanceHex]);
  return prepareResult(hexValue, encodingOptions);
}
function decodeNativeBalanceChangeTerms(terms, encodingOptions = defaultOptions) {
  const hexTerms = bytesLikeToHex(terms);
  assertHexByteExactLength(
    hexTerms,
    53,
    "Invalid NativeBalanceChange terms: must be exactly 53 bytes"
  );
  const changeType = extractNumber(hexTerms, 0, 1);
  const recipientHex = extractAddress(hexTerms, 1);
  const balance = extractBigInt(hexTerms, 21, 32);
  return {
    changeType,
    recipient: prepareResult(recipientHex, encodingOptions),
    balance
  };
}

// src/caveats/erc20Streaming.ts
import { bytesToHex as bytesToHex6, isHexString as isHexString2 } from "@metamask/utils";
var TIMESTAMP_UPPER_BOUND_SECONDS3 = 253402300799;
function createERC20StreamingTerms(terms, encodingOptions = defaultOptions) {
  const { tokenAddress, initialAmount, maxAmount, amountPerSecond, startTime } = terms;
  if (!tokenAddress) {
    throw new Error("Invalid tokenAddress: must be a valid address");
  }
  let prefixedTokenAddressHex;
  if (typeof tokenAddress === "string") {
    if (!isHexString2(tokenAddress) || tokenAddress.length !== 42) {
      throw new Error("Invalid tokenAddress: must be a valid address");
    }
    prefixedTokenAddressHex = tokenAddress;
  } else {
    if (tokenAddress.length !== 20) {
      throw new Error("Invalid tokenAddress: must be a valid address");
    }
    prefixedTokenAddressHex = bytesToHex6(tokenAddress);
  }
  if (initialAmount < 0n) {
    throw new Error("Invalid initialAmount: must be greater than zero");
  }
  if (maxAmount <= 0n) {
    throw new Error("Invalid maxAmount: must be a positive number");
  }
  if (maxAmount < initialAmount) {
    throw new Error("Invalid maxAmount: must be greater than initialAmount");
  }
  if (amountPerSecond <= 0n) {
    throw new Error("Invalid amountPerSecond: must be a positive number");
  }
  if (startTime <= 0) {
    throw new Error("Invalid startTime: must be a positive number");
  }
  if (startTime > TIMESTAMP_UPPER_BOUND_SECONDS3) {
    throw new Error(
      "Invalid startTime: must be less than or equal to 253402300799"
    );
  }
  const initialAmountHex = toHexString({ value: initialAmount, size: 32 });
  const maxAmountHex = toHexString({ value: maxAmount, size: 32 });
  const amountPerSecondHex = toHexString({ value: amountPerSecond, size: 32 });
  const startTimeHex = toHexString({ value: startTime, size: 32 });
  const hexValue = `${prefixedTokenAddressHex}${initialAmountHex}${maxAmountHex}${amountPerSecondHex}${startTimeHex}`;
  return prepareResult(hexValue, encodingOptions);
}
function decodeERC20StreamingTerms(terms, encodingOptions = defaultOptions) {
  const hexTerms = bytesLikeToHex(terms);
  assertHexByteExactLength(
    hexTerms,
    148,
    "Invalid ERC20Streaming terms: must be exactly 148 bytes"
  );
  const tokenAddressHex = extractAddress(hexTerms, 0);
  const initialAmount = extractBigInt(hexTerms, 20, 32);
  const maxAmount = extractBigInt(hexTerms, 52, 32);
  const amountPerSecond = extractBigInt(hexTerms, 84, 32);
  const startTime = extractNumber(hexTerms, 116, 32);
  return {
    tokenAddress: prepareResult(tokenAddressHex, encodingOptions),
    initialAmount,
    maxAmount,
    amountPerSecond,
    startTime
  };
}

// src/caveats/erc20TokenPeriodTransfer.ts
import { isHexString as isHexString3, bytesToHex as bytesToHex7 } from "@metamask/utils";
function createERC20TokenPeriodTransferTerms(terms, encodingOptions = defaultOptions) {
  const { tokenAddress, periodAmount, periodDuration, startDate } = terms;
  if (!tokenAddress) {
    throw new Error("Invalid tokenAddress: must be a valid address");
  }
  let prefixedTokenAddressHex;
  if (typeof tokenAddress === "string") {
    if (!isHexString3(tokenAddress) || tokenAddress.length !== 42) {
      throw new Error("Invalid tokenAddress: must be a valid address");
    }
    prefixedTokenAddressHex = tokenAddress;
  } else {
    if (tokenAddress.length !== 20) {
      throw new Error("Invalid tokenAddress: must be a valid address");
    }
    prefixedTokenAddressHex = bytesToHex7(tokenAddress);
  }
  if (periodAmount <= 0n) {
    throw new Error("Invalid periodAmount: must be a positive number");
  }
  if (periodDuration <= 0) {
    throw new Error("Invalid periodDuration: must be a positive number");
  }
  if (startDate <= 0) {
    throw new Error("Invalid startDate: must be a positive number");
  }
  const periodAmountHex = toHexString({ value: periodAmount, size: 32 });
  const periodDurationHex = toHexString({ value: periodDuration, size: 32 });
  const startDateHex = toHexString({ value: startDate, size: 32 });
  const hexValue = `${prefixedTokenAddressHex}${periodAmountHex}${periodDurationHex}${startDateHex}`;
  return prepareResult(hexValue, encodingOptions);
}
function decodeERC20TokenPeriodTransferTerms(terms, encodingOptions = defaultOptions) {
  const hexTerms = bytesLikeToHex(terms);
  assertHexByteExactLength(
    hexTerms,
    116,
    "Invalid ERC20TokenPeriodTransfer terms: must be exactly 116 bytes"
  );
  const tokenAddressHex = extractAddress(hexTerms, 0);
  const periodAmount = extractBigInt(hexTerms, 20, 32);
  const periodDuration = extractNumber(hexTerms, 52, 32);
  const startDate = extractNumber(hexTerms, 84, 32);
  return {
    tokenAddress: prepareResult(tokenAddressHex, encodingOptions),
    periodAmount,
    periodDuration,
    startDate
  };
}

// src/caveats/erc20TransferAmount.ts
function createERC20TransferAmountTerms(terms, encodingOptions = defaultOptions) {
  const { tokenAddress, maxAmount } = terms;
  const tokenAddressHex = normalizeAddress(
    tokenAddress,
    "Invalid tokenAddress: must be a valid address"
  );
  if (maxAmount <= 0n) {
    throw new Error("Invalid maxAmount: must be a positive number");
  }
  const maxAmountHex = `0x${toHexString({ value: maxAmount, size: 32 })}`;
  const hexValue = concatHex([tokenAddressHex, maxAmountHex]);
  return prepareResult(hexValue, encodingOptions);
}
function decodeERC20TransferAmountTerms(terms, encodingOptions = defaultOptions) {
  const hexTerms = bytesLikeToHex(terms);
  assertHexByteExactLength(
    hexTerms,
    52,
    "Invalid ERC20TransferAmount terms: must be exactly 52 bytes"
  );
  const tokenAddressHex = extractAddress(hexTerms, 0);
  const maxAmount = extractBigInt(hexTerms, 20, 32);
  return {
    tokenAddress: prepareResult(tokenAddressHex, encodingOptions),
    maxAmount
  };
}

// src/caveats/erc20BalanceChange.ts
function createERC20BalanceChangeTerms(terms, encodingOptions = defaultOptions) {
  const { tokenAddress, recipient, balance, changeType } = terms;
  const tokenAddressHex = normalizeAddressLowercase(
    tokenAddress,
    "Invalid tokenAddress: must be a valid address"
  );
  const recipientHex = normalizeAddressLowercase(
    recipient,
    "Invalid recipient: must be a valid address"
  );
  if (balance <= 0n) {
    throw new Error("Invalid balance: must be a positive number");
  }
  if (changeType !== 0 /* Increase */ && changeType !== 1 /* Decrease */) {
    throw new Error("Invalid changeType: must be either Increase or Decrease");
  }
  const changeTypeHex = `0x${toHexString({ value: changeType, size: 1 })}`;
  const balanceHex = `0x${toHexString({ value: balance, size: 32 })}`;
  const hexValue = concatHex([
    changeTypeHex,
    tokenAddressHex,
    recipientHex,
    balanceHex
  ]);
  return prepareResult(hexValue, encodingOptions);
}
function decodeERC20BalanceChangeTerms(terms, encodingOptions = defaultOptions) {
  const hexTerms = bytesLikeToHex(terms);
  assertHexByteExactLength(
    hexTerms,
    73,
    "Invalid ERC20BalanceChange terms: must be exactly 73 bytes"
  );
  const changeType = extractNumber(hexTerms, 0, 1);
  const tokenAddressHex = extractAddress(hexTerms, 1);
  const recipientHex = extractAddress(hexTerms, 21);
  const balance = extractBigInt(hexTerms, 41, 32);
  return {
    changeType,
    tokenAddress: prepareResult(tokenAddressHex, encodingOptions),
    recipient: prepareResult(recipientHex, encodingOptions),
    balance
  };
}

// src/caveats/erc721BalanceChange.ts
function createERC721BalanceChangeTerms(terms, encodingOptions = defaultOptions) {
  const { tokenAddress, recipient, amount, changeType } = terms;
  const tokenAddressHex = normalizeAddressLowercase(
    tokenAddress,
    "Invalid tokenAddress: must be a valid address"
  );
  const recipientHex = normalizeAddressLowercase(
    recipient,
    "Invalid recipient: must be a valid address"
  );
  if (amount <= 0n) {
    throw new Error("Invalid balance: must be a positive number");
  }
  if (changeType !== 0 /* Increase */ && changeType !== 1 /* Decrease */) {
    throw new Error("Invalid changeType: must be either Increase or Decrease");
  }
  const changeTypeHex = `0x${toHexString({ value: changeType, size: 1 })}`;
  const amountHex = `0x${toHexString({ value: amount, size: 32 })}`;
  const hexValue = concatHex([
    changeTypeHex,
    tokenAddressHex,
    recipientHex,
    amountHex
  ]);
  return prepareResult(hexValue, encodingOptions);
}
function decodeERC721BalanceChangeTerms(terms, encodingOptions = defaultOptions) {
  const hexTerms = bytesLikeToHex(terms);
  assertHexByteExactLength(
    hexTerms,
    73,
    "Invalid ERC721BalanceChange terms: must be exactly 73 bytes"
  );
  const changeType = extractNumber(hexTerms, 0, 1);
  const tokenAddressHex = extractAddress(hexTerms, 1);
  const recipientHex = extractAddress(hexTerms, 21);
  const amount = extractBigInt(hexTerms, 41, 32);
  return {
    changeType,
    tokenAddress: prepareResult(tokenAddressHex, encodingOptions),
    recipient: prepareResult(recipientHex, encodingOptions),
    amount
  };
}

// src/caveats/erc721Transfer.ts
function createERC721TransferTerms(terms, encodingOptions = defaultOptions) {
  const { tokenAddress, tokenId } = terms;
  const tokenAddressHex = normalizeAddress(
    tokenAddress,
    "Invalid tokenAddress: must be a valid address"
  );
  if (tokenId < 0n) {
    throw new Error("Invalid tokenId: must be a non-negative number");
  }
  const tokenIdHex = `0x${toHexString({ value: tokenId, size: 32 })}`;
  const hexValue = concatHex([tokenAddressHex, tokenIdHex]);
  return prepareResult(hexValue, encodingOptions);
}
function decodeERC721TransferTerms(terms, encodingOptions = defaultOptions) {
  const hexTerms = bytesLikeToHex(terms);
  assertHexByteExactLength(
    hexTerms,
    52,
    "Invalid ERC721Transfer terms: must be exactly 52 bytes"
  );
  const tokenAddressHex = extractAddress(hexTerms, 0);
  const tokenId = extractBigInt(hexTerms, 20, 32);
  return {
    tokenAddress: prepareResult(tokenAddressHex, encodingOptions),
    tokenId
  };
}

// src/caveats/erc1155BalanceChange.ts
function createERC1155BalanceChangeTerms(terms, encodingOptions = defaultOptions) {
  const { tokenAddress, recipient, tokenId, balance, changeType } = terms;
  const tokenAddressHex = normalizeAddressLowercase(
    tokenAddress,
    "Invalid tokenAddress: must be a valid address"
  );
  const recipientHex = normalizeAddressLowercase(
    recipient,
    "Invalid recipient: must be a valid address"
  );
  if (balance <= 0n) {
    throw new Error("Invalid balance: must be a positive number");
  }
  if (tokenId < 0n) {
    throw new Error("Invalid tokenId: must be a non-negative number");
  }
  if (changeType !== 0 /* Increase */ && changeType !== 1 /* Decrease */) {
    throw new Error("Invalid changeType: must be either Increase or Decrease");
  }
  const changeTypeHex = `0x${toHexString({ value: changeType, size: 1 })}`;
  const tokenIdHex = `0x${toHexString({ value: tokenId, size: 32 })}`;
  const balanceHex = `0x${toHexString({ value: balance, size: 32 })}`;
  const hexValue = concatHex([
    changeTypeHex,
    tokenAddressHex,
    recipientHex,
    tokenIdHex,
    balanceHex
  ]);
  return prepareResult(hexValue, encodingOptions);
}
function decodeERC1155BalanceChangeTerms(terms, encodingOptions = defaultOptions) {
  const hexTerms = bytesLikeToHex(terms);
  assertHexByteExactLength(
    hexTerms,
    105,
    "Invalid ERC1155BalanceChange terms: must be exactly 105 bytes"
  );
  const changeType = extractNumber(hexTerms, 0, 1);
  const tokenAddressHex = extractAddress(hexTerms, 1);
  const recipientHex = extractAddress(hexTerms, 21);
  const tokenId = extractBigInt(hexTerms, 41, 32);
  const balance = extractBigInt(hexTerms, 73, 32);
  return {
    changeType,
    tokenAddress: prepareResult(tokenAddressHex, encodingOptions),
    recipient: prepareResult(recipientHex, encodingOptions),
    tokenId,
    balance
  };
}

// src/caveats/nonce.ts
import { isHexString as isHexString4 } from "@metamask/utils";
var MAX_NONCE_STRING_LENGTH = 66;
function createNonceTerms(terms, encodingOptions = defaultOptions) {
  const { nonce } = terms;
  if (nonce instanceof Uint8Array && nonce.length === 0) {
    throw new Error("Invalid nonce: Uint8Array must not be empty");
  }
  if (typeof nonce === "string" && !nonce.startsWith("0x")) {
    throw new Error("Invalid nonce: string must have 0x prefix");
  }
  const hexNonce = bytesLikeToHex(nonce);
  if (hexNonce === "0x") {
    throw new Error("Invalid nonce: must not be empty");
  }
  if (!isHexString4(hexNonce)) {
    throw new Error("Invalid nonce: must be a valid BytesLike value");
  }
  if (hexNonce.length > MAX_NONCE_STRING_LENGTH) {
    throw new Error("Invalid nonce: must be 32 bytes or less in length");
  }
  const nonceWithoutPrefix = hexNonce.slice(2);
  const paddedNonce = nonceWithoutPrefix.padStart(64, "0");
  const hexValue = `0x${paddedNonce}`;
  return prepareResult(hexValue, encodingOptions);
}
function decodeNonceTerms(terms, encodingOptions = defaultOptions) {
  const hexTerms = bytesLikeToHex(terms);
  assertHexByteExactLength(
    hexTerms,
    32,
    "Invalid Nonce terms: must be exactly 32 bytes"
  );
  const nonce = prepareResult(hexTerms, encodingOptions);
  return { nonce };
}

// src/caveats/allowedCalldata.ts
import { bytesToHex as bytesToHex8, remove0x as remove0x2 } from "@metamask/utils";
function createAllowedCalldataTerms(terms, encodingOptions = defaultOptions) {
  const { startIndex, value } = terms;
  if (startIndex < 0) {
    throw new Error("Invalid startIndex: must be zero or positive");
  }
  if (!Number.isInteger(startIndex)) {
    throw new Error("Invalid startIndex: must be a whole number");
  }
  let unprefixedValue;
  if (typeof value === "string") {
    if (!value.startsWith("0x")) {
      throw new Error("Invalid value: must be a hex string starting with 0x");
    }
    unprefixedValue = remove0x2(value);
  } else {
    unprefixedValue = remove0x2(bytesToHex8(value));
  }
  const indexHex = toHexString({ value: startIndex, size: 32 });
  return prepareResult(`0x${indexHex}${unprefixedValue}`, encodingOptions);
}
function decodeAllowedCalldataTerms(terms, encodingOptions = defaultOptions) {
  const hexTerms = bytesLikeToHex(terms);
  assertHexBytesMinLength(
    hexTerms,
    32,
    "Invalid AllowedCalldata terms: must be at least 32 bytes"
  );
  const startIndex = extractNumber(hexTerms, 0, 32);
  const valueHex = extractRemainingHex(hexTerms, 32);
  const value = prepareResult(valueHex, encodingOptions);
  return { startIndex, value };
}

// src/caveats/allowedMethods.ts
import { bytesToHex as bytesToHex9, isHexString as isHexString5 } from "@metamask/utils";
var FUNCTION_SELECTOR_STRING_LENGTH = 10;
var INVALID_SELECTOR_ERROR = "Invalid selector: must be a 4 byte hex string, abi function signature, or AbiFunction";
function createAllowedMethodsTerms(terms, encodingOptions = defaultOptions) {
  const { selectors } = terms;
  if (!selectors || selectors.length === 0) {
    throw new Error("Invalid selectors: must provide at least one selector");
  }
  const normalizedSelectors = selectors.map((selector) => {
    if (typeof selector === "string") {
      if (isHexString5(selector) && selector.length === FUNCTION_SELECTOR_STRING_LENGTH) {
        return selector;
      }
      throw new Error(INVALID_SELECTOR_ERROR);
    }
    if (selector.length !== 4) {
      throw new Error(INVALID_SELECTOR_ERROR);
    }
    return bytesToHex9(selector);
  });
  const hexValue = concatHex(normalizedSelectors);
  return prepareResult(hexValue, encodingOptions);
}
function decodeAllowedMethodsTerms(terms, encodingOptions = defaultOptions) {
  const hexTerms = bytesLikeToHex(terms);
  const selectorSize = 4;
  assertHexByteLengthAtLeastOneMultipleOf(
    hexTerms,
    selectorSize,
    "Invalid selectors: must be a multiple of 4"
  );
  const selectorCount = getByteLength(hexTerms) / selectorSize;
  const selectors = [];
  for (let i = 0; i < selectorCount; i++) {
    const selector = extractHex(hexTerms, i * selectorSize, selectorSize);
    selectors.push(prepareResult(selector, encodingOptions));
  }
  return { selectors };
}

// src/caveats/allowedTargets.ts
function createAllowedTargetsTerms(terms, encodingOptions = defaultOptions) {
  const { targets } = terms;
  if (!targets || targets.length === 0) {
    throw new Error(
      "Invalid targets: must provide at least one target address"
    );
  }
  const normalizedTargets = targets.map(
    (target) => normalizeAddress(target, "Invalid targets: must be valid addresses")
  );
  const hexValue = concatHex(normalizedTargets);
  return prepareResult(hexValue, encodingOptions);
}
function decodeAllowedTargetsTerms(terms, encodingOptions = defaultOptions) {
  const hexTerms = bytesLikeToHex(terms);
  const addressSize = 20;
  assertHexByteLengthAtLeastOneMultipleOf(
    hexTerms,
    addressSize,
    "Invalid targets: must be a multiple of 20"
  );
  const addressCount = getByteLength(hexTerms) / addressSize;
  const targets = [];
  for (let i = 0; i < addressCount; i++) {
    const target = extractAddress(hexTerms, i * addressSize);
    targets.push(prepareResult(target, encodingOptions));
  }
  return { targets };
}

// src/caveats/approvalRevocationEnforcer.ts
var BIT_ERC20_APPROVE_ZERO = 1;
var BIT_ERC721_PER_TOKEN_CLEAR = 2;
var BIT_SET_APPROVAL_FOR_ALL_REVOKE = 4;
var BIT_PERMIT2_APPROVE_ZERO = 8;
var BIT_PERMIT2_LOCKDOWN = 16;
var BIT_PERMIT2_INVALIDATE_NONCES = 32;
var ALLOWED_APPROVAL_REVOCATION_MAX_MASK = BIT_ERC20_APPROVE_ZERO | BIT_ERC721_PER_TOKEN_CLEAR | BIT_SET_APPROVAL_FOR_ALL_REVOKE | BIT_PERMIT2_APPROVE_ZERO | BIT_PERMIT2_LOCKDOWN | BIT_PERMIT2_INVALIDATE_NONCES;
var NO_FLAGS_SET_ERROR = "Invalid ApprovalRevocation terms: at least one revocation primitive must be enabled";
function termsToMask(terms) {
  let mask = 0;
  if (terms.erc20Approve) {
    mask |= BIT_ERC20_APPROVE_ZERO;
  }
  if (terms.erc721Approve) {
    mask |= BIT_ERC721_PER_TOKEN_CLEAR;
  }
  if (terms.erc721SetApprovalForAll) {
    mask |= BIT_SET_APPROVAL_FOR_ALL_REVOKE;
  }
  if (terms.permit2Approve) {
    mask |= BIT_PERMIT2_APPROVE_ZERO;
  }
  if (terms.permit2Lockdown) {
    mask |= BIT_PERMIT2_LOCKDOWN;
  }
  if (terms.permit2InvalidateNonces) {
    mask |= BIT_PERMIT2_INVALIDATE_NONCES;
  }
  if (mask === 0) {
    throw new Error(NO_FLAGS_SET_ERROR);
  }
  return mask;
}
function maskToTerms(mask) {
  if (mask > ALLOWED_APPROVAL_REVOCATION_MAX_MASK) {
    throw new Error(
      "Invalid ApprovalRevocation terms: reserved bits must be zero (only bits 0-5 are defined)"
    );
  }
  if (mask === 0) {
    throw new Error(NO_FLAGS_SET_ERROR);
  }
  return {
    erc20Approve: (mask & BIT_ERC20_APPROVE_ZERO) !== 0,
    erc721Approve: (mask & BIT_ERC721_PER_TOKEN_CLEAR) !== 0,
    erc721SetApprovalForAll: (mask & BIT_SET_APPROVAL_FOR_ALL_REVOKE) !== 0,
    permit2Approve: (mask & BIT_PERMIT2_APPROVE_ZERO) !== 0,
    permit2Lockdown: (mask & BIT_PERMIT2_LOCKDOWN) !== 0,
    permit2InvalidateNonces: (mask & BIT_PERMIT2_INVALIDATE_NONCES) !== 0
  };
}
function createApprovalRevocationTerms(terms, encodingOptions = defaultOptions) {
  const mask = termsToMask(terms);
  const hexValue = `0x${toHexString({ value: mask, size: 1 })}`;
  return prepareResult(hexValue, encodingOptions);
}
function decodeApprovalRevocationTerms(terms) {
  const hexTerms = bytesLikeToHex(terms);
  assertHexByteExactLength(
    hexTerms,
    1,
    "Invalid ApprovalRevocation terms: must be exactly 1 byte"
  );
  const mask = extractNumber(hexTerms, 0, 1);
  return maskToTerms(mask);
}

// src/caveats/argsEqualityCheck.ts
function createArgsEqualityCheckTerms(terms, encodingOptions = defaultOptions) {
  const { args } = terms;
  if (typeof args === "string" && args === "0x") {
    return prepareResult(args, encodingOptions);
  }
  const hexValue = normalizeHex(
    args,
    "Invalid config: args must be a valid hex string"
  );
  return prepareResult(hexValue, encodingOptions);
}
function decodeArgsEqualityCheckTerms(terms, encodingOptions = defaultOptions) {
  const argsHex = bytesLikeToHex(terms);
  const args = prepareResult(argsHex, encodingOptions);
  return { args };
}

// src/caveats/blockNumber.ts
function createBlockNumberTerms(terms, encodingOptions = defaultOptions) {
  const { afterThreshold, beforeThreshold } = terms;
  if (afterThreshold < 0n || beforeThreshold < 0n) {
    throw new Error("Invalid thresholds: block numbers must be non-negative");
  }
  if (afterThreshold === 0n && beforeThreshold === 0n) {
    throw new Error(
      "Invalid thresholds: At least one of afterThreshold or beforeThreshold must be specified"
    );
  }
  if (beforeThreshold !== 0n && afterThreshold >= beforeThreshold) {
    throw new Error(
      "Invalid thresholds: afterThreshold must be less than beforeThreshold if both are specified"
    );
  }
  const afterThresholdHex = toHexString({ value: afterThreshold, size: 16 });
  const beforeThresholdHex = toHexString({ value: beforeThreshold, size: 16 });
  const hexValue = `0x${afterThresholdHex}${beforeThresholdHex}`;
  return prepareResult(hexValue, encodingOptions);
}
function decodeBlockNumberTerms(terms) {
  const hexTerms = bytesLikeToHex(terms);
  assertHexByteExactLength(
    hexTerms,
    32,
    "Invalid BlockNumber terms: must be exactly 32 bytes"
  );
  const afterThreshold = extractBigInt(hexTerms, 0, 16);
  const beforeThreshold = extractBigInt(hexTerms, 16, 16);
  return { afterThreshold, beforeThreshold };
}

// src/caveats/deployed.ts
import { remove0x as remove0x3 } from "@metamask/utils";
function createDeployedTerms(terms, encodingOptions = defaultOptions) {
  const { contractAddress, salt, bytecode } = terms;
  const contractAddressHex = normalizeAddress(
    contractAddress,
    "Invalid contractAddress: must be a valid Ethereum address"
  );
  const saltHex = normalizeHex(
    salt,
    "Invalid salt: must be a valid hexadecimal string"
  );
  const bytecodeHex = normalizeHex(
    bytecode,
    "Invalid bytecode: must be a valid hexadecimal string"
  );
  const unprefixedSalt = remove0x3(saltHex);
  if (unprefixedSalt.length > 64) {
    throw new Error("Invalid salt: must be a valid hexadecimal string");
  }
  const paddedSalt = `0x${unprefixedSalt.padStart(64, "0")}`;
  const hexValue = concatHex([contractAddressHex, paddedSalt, bytecodeHex]);
  return prepareResult(hexValue, encodingOptions);
}
function decodeDeployedTerms(terms, encodingOptions = defaultOptions) {
  const hexTerms = bytesLikeToHex(terms);
  assertHexBytesMinLength(
    hexTerms,
    52,
    "Invalid Deployed terms: must be at least 52 bytes"
  );
  const contractAddressHex = extractAddress(hexTerms, 0);
  const saltHex = extractHex(hexTerms, 20, 32);
  const bytecodeHex = extractRemainingHex(hexTerms, 52);
  return {
    contractAddress: prepareResult(contractAddressHex, encodingOptions),
    salt: prepareResult(saltHex, encodingOptions),
    bytecode: prepareResult(bytecodeHex, encodingOptions)
  };
}

// src/caveats/id.ts
var MAX_UINT256 = BigInt(`0x${"f".repeat(64)}`);
function createIdTerms(terms, encodingOptions = defaultOptions) {
  const { id } = terms;
  let idBigInt;
  if (typeof id === "number") {
    if (!Number.isInteger(id)) {
      throw new Error("Invalid id: must be an integer");
    }
    idBigInt = BigInt(id);
  } else if (typeof id === "bigint") {
    idBigInt = id;
  } else {
    throw new Error("Invalid id: must be a bigint or number");
  }
  if (idBigInt < 0n) {
    throw new Error("Invalid id: must be a non-negative number");
  }
  if (idBigInt > MAX_UINT256) {
    throw new Error("Invalid id: must be less than 2^256");
  }
  const hexValue = `0x${toHexString({ value: idBigInt, size: 32 })}`;
  return prepareResult(hexValue, encodingOptions);
}
function decodeIdTerms(terms) {
  const hexTerms = bytesLikeToHex(terms);
  assertHexByteExactLength(
    hexTerms,
    32,
    "Invalid Id terms: must be exactly 32 bytes"
  );
  const id = extractBigInt(hexTerms, 0, 32);
  return { id };
}

// src/caveats/limitedCalls.ts
function createLimitedCallsTerms(terms, encodingOptions = defaultOptions) {
  const { limit } = terms;
  if (!Number.isInteger(limit)) {
    throw new Error("Invalid limit: must be an integer");
  }
  if (limit <= 0) {
    throw new Error("Invalid limit: must be a positive integer");
  }
  const hexValue = `0x${toHexString({ value: limit, size: 32 })}`;
  return prepareResult(hexValue, encodingOptions);
}
function decodeLimitedCallsTerms(terms) {
  const hexTerms = bytesLikeToHex(terms);
  assertHexByteExactLength(
    hexTerms,
    32,
    "Invalid LimitedCalls terms: must be exactly 32 bytes"
  );
  const limit = extractNumber(hexTerms, 0, 32);
  return { limit };
}

// src/caveats/multiTokenPeriod.ts
function createMultiTokenPeriodTerms(terms, encodingOptions = defaultOptions) {
  const { tokenConfigs } = terms;
  if (!tokenConfigs || tokenConfigs.length === 0) {
    throw new Error(
      "MultiTokenPeriodBuilder: tokenConfigs array cannot be empty"
    );
  }
  const hexParts = [];
  for (const tokenConfig of tokenConfigs) {
    const tokenHex = normalizeAddress(
      tokenConfig.token,
      `Invalid token address: ${String(tokenConfig.token)}`
    );
    if (tokenConfig.periodAmount <= 0n) {
      throw new Error("Invalid period amount: must be greater than 0");
    }
    if (tokenConfig.periodDuration <= 0) {
      throw new Error("Invalid period duration: must be greater than 0");
    }
    if (tokenConfig.startDate <= 0) {
      throw new Error("Invalid start date: must be greater than 0");
    }
    hexParts.push(
      tokenHex,
      `0x${toHexString({ value: tokenConfig.periodAmount, size: 32 })}`,
      `0x${toHexString({ value: tokenConfig.periodDuration, size: 32 })}`,
      `0x${toHexString({ value: tokenConfig.startDate, size: 32 })}`
    );
  }
  const hexValue = concatHex(hexParts);
  return prepareResult(hexValue, encodingOptions);
}
function decodeMultiTokenPeriodTerms(terms, encodingOptions = defaultOptions) {
  const hexTerms = bytesLikeToHex(terms);
  const configSize = 116;
  assertHexByteLengthAtLeastOneMultipleOf(
    hexTerms,
    configSize,
    "Invalid MultiTokenPeriod terms: must be a multiple of 116 bytes"
  );
  const configCount = getByteLength(hexTerms) / configSize;
  const tokenConfigs = [];
  for (let i = 0; i < configCount; i++) {
    const offset = i * configSize;
    const tokenHex = extractAddress(hexTerms, offset);
    const periodAmount = extractBigInt(hexTerms, offset + 20, 32);
    const periodDuration = extractNumber(hexTerms, offset + 52, 32);
    const startDate = extractNumber(hexTerms, offset + 84, 32);
    tokenConfigs.push({
      token: prepareResult(tokenHex, encodingOptions),
      periodAmount,
      periodDuration,
      startDate
    });
  }
  return { tokenConfigs };
}

// src/caveats/ownershipTransfer.ts
function createOwnershipTransferTerms(terms, encodingOptions = defaultOptions) {
  const { contractAddress } = terms;
  const contractAddressHex = normalizeAddress(
    contractAddress,
    "Invalid contractAddress: must be a valid address"
  );
  return prepareResult(contractAddressHex, encodingOptions);
}
function decodeOwnershipTransferTerms(terms, encodingOptions = defaultOptions) {
  const hexTerms = bytesLikeToHex(terms);
  assertHexByteExactLength(
    hexTerms,
    20,
    "Invalid OwnershipTransfer terms: must be exactly 20 bytes"
  );
  const contractAddressHex = extractAddress(hexTerms, 0);
  return {
    contractAddress: prepareResult(contractAddressHex, encodingOptions)
  };
}

// src/caveats/redeemer.ts
function createRedeemerTerms(terms, encodingOptions = defaultOptions) {
  const { redeemers } = terms;
  if (!redeemers || redeemers.length === 0) {
    throw new Error(
      "Invalid redeemers: must specify at least one redeemer address"
    );
  }
  const normalizedRedeemers = redeemers.map(
    (redeemer) => normalizeAddress(redeemer, "Invalid redeemers: must be a valid address")
  );
  const hexValue = concatHex(normalizedRedeemers);
  return prepareResult(hexValue, encodingOptions);
}
function decodeRedeemerTerms(terms, encodingOptions = defaultOptions) {
  const hexTerms = bytesLikeToHex(terms);
  const addressSize = 20;
  assertHexByteLengthAtLeastOneMultipleOf(
    hexTerms,
    addressSize,
    "Invalid redeemers: must be a multiple of 20"
  );
  const addressCount = getByteLength(hexTerms) / addressSize;
  const redeemers = [];
  for (let i = 0; i < addressCount; i++) {
    const redeemer = extractAddress(hexTerms, i * addressSize);
    redeemers.push(prepareResult(redeemer, encodingOptions));
  }
  return { redeemers };
}

// src/caveats/specificActionERC20TransferBatch.ts
import { bytesToHex as bytesToHex10 } from "@metamask/utils";
function createSpecificActionERC20TransferBatchTerms(terms, encodingOptions = defaultOptions) {
  const { tokenAddress, recipient, amount, target, calldata } = terms;
  const tokenAddressHex = normalizeAddress(
    tokenAddress,
    "Invalid tokenAddress: must be a valid address"
  );
  const recipientHex = normalizeAddress(
    recipient,
    "Invalid recipient: must be a valid address"
  );
  const targetHex = normalizeAddress(
    target,
    "Invalid target: must be a valid address"
  );
  let calldataHex;
  if (typeof calldata === "string") {
    if (!calldata.startsWith("0x")) {
      throw new Error(
        "Invalid calldata: must be a hex string starting with 0x"
      );
    }
    calldataHex = calldata;
  } else {
    calldataHex = bytesToHex10(calldata);
  }
  if (amount <= 0n) {
    throw new Error("Invalid amount: must be a positive number");
  }
  const amountHex = `0x${toHexString({ value: amount, size: 32 })}`;
  const hexValue = concatHex([
    tokenAddressHex,
    recipientHex,
    amountHex,
    targetHex,
    calldataHex
  ]);
  return prepareResult(hexValue, encodingOptions);
}
function decodeSpecificActionERC20TransferBatchTerms(terms, encodingOptions = defaultOptions) {
  const hexTerms = bytesLikeToHex(terms);
  assertHexBytesMinLength(
    hexTerms,
    92,
    "Invalid SpecificActionERC20TransferBatch terms: must be at least 92 bytes"
  );
  const tokenAddressHex = extractAddress(hexTerms, 0);
  const recipientHex = extractAddress(hexTerms, 20);
  const amount = extractBigInt(hexTerms, 40, 32);
  const targetHex = extractAddress(hexTerms, 72);
  const calldataHex = extractRemainingHex(hexTerms, 92);
  return {
    tokenAddress: prepareResult(tokenAddressHex, encodingOptions),
    recipient: prepareResult(recipientHex, encodingOptions),
    amount,
    target: prepareResult(targetHex, encodingOptions),
    calldata: prepareResult(calldataHex, encodingOptions)
  };
}

// src/caveats/logicalOrWrapper.ts
import { decodeSingle as decodeSingle3, encodeSingle as encodeSingle3 } from "@metamask/abi-utils";
import { bytesToHex as bytesToHex11 } from "@metamask/utils";
var CAVEAT_GROUPS_ABI = "((address,bytes,bytes)[])[]";
var SELECTED_GROUP_ABI = "(uint256,bytes[])";
function assertValidCaveatGroups(groups) {
  if (!groups?.length) {
    throw new Error(
      "Invalid caveatGroups: must provide at least one caveat group"
    );
  }
  for (let i = 0; i < groups.length; i++) {
    const group = groups[i];
    if (!group?.length) {
      throw new Error(
        `Invalid caveatGroups: group at index ${i} must contain at least one caveat`
      );
    }
  }
}
function normalizeCaveatTuple(caveat) {
  return [
    normalizeAddress(
      caveat.enforcer,
      "Invalid enforcer: must be a valid address"
    ),
    normalizeHex(caveat.terms, "Invalid terms: must be a valid hex string"),
    normalizeHex(caveat.args, "Invalid args: must be a valid hex string")
  ];
}
function encodeCaveatGroupTuple(group) {
  return [group.map(normalizeCaveatTuple)];
}
function createLogicalOrWrapperTerms(terms, encodingOptions = defaultOptions) {
  assertValidCaveatGroups(terms.caveatGroups);
  const encodableGroups = terms.caveatGroups.map(encodeCaveatGroupTuple);
  const hexValue = encodeSingle3(CAVEAT_GROUPS_ABI, encodableGroups);
  return prepareResult(hexValue, encodingOptions);
}
function decodeLogicalOrWrapperTerms(terms, encodingOptions = defaultOptions) {
  const hexTerms = bytesLikeToHex(terms);
  const decoded = decodeSingle3(
    CAVEAT_GROUPS_ABI,
    hexTerms
  );
  const caveatGroups = decoded.map(
    ([caveats]) => caveats.map(([enforcer, caveatTerms, args]) => ({
      enforcer: prepareResult(enforcer, encodingOptions),
      terms: prepareResult(bytesToHex11(caveatTerms), encodingOptions),
      args: prepareResult(bytesToHex11(args), encodingOptions)
    }))
  );
  return { caveatGroups };
}
function createLogicalOrWrapperArgs(args, encodingOptions = defaultOptions) {
  if (args.groupIndex < 0n) {
    throw new Error("Invalid groupIndex: must be a non-negative number");
  }
  const caveatArgsHex = args.caveatArgs.map(
    (arg) => normalizeHex(arg, "Invalid caveatArgs: must be valid hex strings")
  );
  const hexValue = encodeSingle3(SELECTED_GROUP_ABI, [
    args.groupIndex,
    caveatArgsHex
  ]);
  return prepareResult(hexValue, encodingOptions);
}
function decodeLogicalOrWrapperArgs(args, encodingOptions = defaultOptions) {
  const hexArgs = bytesLikeToHex(args);
  const [groupIndex, caveatArgsRaw] = decodeSingle3(
    SELECTED_GROUP_ABI,
    hexArgs
  );
  const caveatArgs = caveatArgsRaw.map(
    (arg) => prepareResult(bytesToHex11(arg), encodingOptions)
  );
  return { groupIndex, caveatArgs };
}

// src/delegation.ts
import { encode, encodeSingle as encodeSingle4, decodeSingle as decodeSingle4 } from "@metamask/abi-utils";
import { hexToBytes as hexToBytes3 } from "@metamask/utils";
import { keccak_256 as keccak256 } from "@noble/hashes/sha3";
var ANY_BENEFICIARY = "0x0000000000000000000000000000000000000a11";
var ROOT_AUTHORITY = "0xffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff";
var DELEGATION_TYPEHASH = "0x88c1d2ecf185adf710588203a5f263f0ff61be0d33da39792cde19ba9aa4331e";
var CAVEAT_TYPEHASH = "0x80ad7e1b04ee6d994a125f4714ca0720908bd80ed16063ec8aee4b88e9253e2d";
var DELEGATION_ARRAY_ABI_TYPES = "(address,address,bytes32,(address,bytes,bytes)[],uint256,bytes)[]";
var DELEGATION_ABI_TYPE = "(address,address,bytes32,(address,bytes,bytes)[],uint256,bytes)";
function encodeDelegations(delegations, options = defaultOptions) {
  let result;
  if (delegations.length === 0) {
    result = new Uint8Array(64);
    result[31] = 32;
  } else {
    const encodableStructs = delegations.map((struct) => [
      struct.delegate,
      struct.delegator,
      struct.authority,
      struct.caveats.map((caveat) => [
        caveat.enforcer,
        caveat.terms,
        caveat.args
      ]),
      struct.salt,
      struct.signature
    ]);
    result = encodeSingle4(DELEGATION_ARRAY_ABI_TYPES, encodableStructs);
  }
  return prepareResult(result, options);
}
function encodeDelegation(delegation, options = defaultOptions) {
  const encodableStruct = [
    delegation.delegate,
    delegation.delegator,
    delegation.authority,
    delegation.caveats.map((caveat) => [
      caveat.enforcer,
      caveat.terms,
      caveat.args
    ]),
    delegation.salt,
    delegation.signature
  ];
  const result = encodeSingle4(DELEGATION_ABI_TYPE, encodableStruct);
  return prepareResult(result, options);
}
var delegationFromDecodedDelegation = (decodedDelegation, convertFn) => {
  const [delegate, delegator, authority, caveats, salt, signature] = decodedDelegation;
  return {
    delegate: convertFn(delegate),
    delegator: convertFn(delegator),
    authority: convertFn(authority),
    caveats: caveats.map(([enforcer, terms, args]) => ({
      enforcer: convertFn(enforcer),
      terms: convertFn(terms),
      args: convertFn(args)
    })),
    salt,
    signature: convertFn(signature)
  };
};
function decodeDelegations(encoded, options = defaultOptions) {
  const decodedStructs = decodeSingle4(
    DELEGATION_ARRAY_ABI_TYPES,
    encoded
    // return types cannot be inferred from complex ABI types, so we must assert the type
  );
  if (options.out === "bytes") {
    return decodedStructs.map(
      (struct) => delegationFromDecodedDelegation(struct, bytesLikeToBytes)
    );
  }
  return decodedStructs.map(
    (struct) => delegationFromDecodedDelegation(struct, bytesLikeToHex)
  );
}
function decodeDelegation(encoded, options = defaultOptions) {
  const decodedStruct = decodeSingle4(
    DELEGATION_ABI_TYPE,
    encoded
    // return types cannot be inferred from complex ABI types, so we must assert the type
  );
  if (options.out === "bytes") {
    return delegationFromDecodedDelegation(decodedStruct, bytesLikeToBytes);
  }
  return delegationFromDecodedDelegation(decodedStruct, bytesLikeToHex);
}
function hashDelegation(delegation, options = defaultOptions) {
  const encoded = encode(
    ["bytes32", "address", "address", "bytes32", "bytes32", "uint256"],
    [
      DELEGATION_TYPEHASH,
      delegation.delegate,
      delegation.delegator,
      delegation.authority,
      getCaveatsArrayHash(delegation.caveats),
      delegation.salt
    ]
  );
  const hash = keccak256(encoded);
  return prepareResult(hash, options);
}
function getCaveatsArrayHash(caveats) {
  const byteLength = 32 * caveats.length;
  const encoded = new Uint8Array(byteLength);
  for (let i = 0; i < caveats.length; i++) {
    const caveat = caveats[i];
    if (!caveat) {
      throw new Error(`Caveat was undefined at index ${i}`);
    }
    const caveatHash = getCaveatHash(caveat);
    encoded.set(caveatHash, i * 32);
  }
  return keccak256(encoded);
}
function getCaveatHash(caveat) {
  const termsBytes = typeof caveat.terms === "string" ? hexToBytes3(caveat.terms) : caveat.terms;
  const termsHash = keccak256(termsBytes);
  const encoded = encode(
    ["bytes32", "address", "bytes32"],
    [CAVEAT_TYPEHASH, caveat.enforcer, termsHash]
  );
  const hash = keccak256(encoded);
  return hash;
}
export {
  ANY_BENEFICIARY,
  BalanceChangeType,
  CAVEAT_TYPEHASH,
  DELEGATION_TYPEHASH,
  ROOT_AUTHORITY,
  createAllowedCalldataTerms,
  createAllowedMethodsTerms,
  createAllowedTargetsTerms,
  createApprovalRevocationTerms,
  createArgsEqualityCheckTerms,
  createBlockNumberTerms,
  createDeployedTerms,
  createERC1155BalanceChangeTerms,
  createERC20BalanceChangeTerms,
  createERC20StreamingTerms,
  createERC20TokenPeriodTransferTerms,
  createERC20TransferAmountTerms,
  createERC721BalanceChangeTerms,
  createERC721TransferTerms,
  createExactCalldataBatchTerms,
  createExactCalldataTerms,
  createExactExecutionBatchTerms,
  createExactExecutionTerms,
  createIdTerms,
  createLimitedCallsTerms,
  createLogicalOrWrapperArgs,
  createLogicalOrWrapperTerms,
  createMultiTokenPeriodTerms,
  createNativeBalanceChangeTerms,
  createNativeTokenPaymentTerms,
  createNativeTokenPeriodTransferTerms,
  createNativeTokenStreamingTerms,
  createNativeTokenTransferAmountTerms,
  createNonceTerms,
  createOwnershipTransferTerms,
  createRedeemerTerms,
  createSpecificActionERC20TransferBatchTerms,
  createTimestampTerms,
  createValueLteTerms,
  decodeAllowedCalldataTerms,
  decodeAllowedMethodsTerms,
  decodeAllowedTargetsTerms,
  decodeApprovalRevocationTerms,
  decodeArgsEqualityCheckTerms,
  decodeBlockNumberTerms,
  decodeDelegation,
  decodeDelegations,
  decodeDeployedTerms,
  decodeERC1155BalanceChangeTerms,
  decodeERC20BalanceChangeTerms,
  decodeERC20StreamingTerms,
  decodeERC20TokenPeriodTransferTerms,
  decodeERC20TransferAmountTerms,
  decodeERC721BalanceChangeTerms,
  decodeERC721TransferTerms,
  decodeExactCalldataBatchTerms,
  decodeExactCalldataTerms,
  decodeExactExecutionBatchTerms,
  decodeExactExecutionTerms,
  decodeIdTerms,
  decodeLimitedCallsTerms,
  decodeLogicalOrWrapperArgs,
  decodeLogicalOrWrapperTerms,
  decodeMultiTokenPeriodTerms,
  decodeNativeBalanceChangeTerms,
  decodeNativeTokenPaymentTerms,
  decodeNativeTokenPeriodTransferTerms,
  decodeNativeTokenStreamingTerms,
  decodeNativeTokenTransferAmountTerms,
  decodeNonceTerms,
  decodeOwnershipTransferTerms,
  decodeRedeemerTerms,
  decodeSpecificActionERC20TransferBatchTerms,
  decodeTimestampTerms,
  decodeValueLteTerms,
  encodeDelegation,
  encodeDelegations,
  hashDelegation
};
