// @metamask/smart-accounts-kit/dist/experimental/index.cjs
"use strict";Object.defineProperty(exports, "__esModule", {value: true}); function _nullishCoalesce(lhs, rhsFn) { if (lhs != null) { return lhs; } else { return rhsFn(); } } async function _asyncNullishCoalesce(lhs, rhsFn) { if (lhs != null) { return lhs; } else { return await rhsFn(); } } function _optionalChain(ops) { let lastAccessLHS = undefined; let value = ops[0]; let i = 1; while (i < ops.length) { const op = ops[i]; const fn = ops[i + 1]; i += 2; if ((op === 'optionalAccess' || op === 'optionalCall') && value == null) { return undefined; } if (op === 'access' || op === 'optionalAccess') { lastAccessLHS = value; value = fn(value); } else if (op === 'call' || op === 'optionalCall') { value = fn((...args) => value.call(lastAccessLHS, ...args)); lastAccessLHS = undefined; } } return value; }

var _chunkIQT6XFDTcjs = require('../chunk-IQT6XFDT.cjs');
require('../chunk-DCI2AN4N.cjs');


var _chunk2OS36DVVcjs = require('../chunk-2OS36DVV.cjs');
require('../chunk-JKVSTPEC.cjs');








var _chunkW6HPAK4Mcjs = require('../chunk-W6HPAK4M.cjs');

// src/experimental/delegationStorage.ts
var _viem = require('viem');
var DelegationStorageEnvironment = {
  dev: { apiUrl: "https://passkeys.dev-api.cx.metamask.io" },
  prod: { apiUrl: "https://passkeys.api.cx.metamask.io" }
};
function getDelegationStorageEnvironment(apiUrl) {
  const matchingEnvironment = Object.entries(DelegationStorageEnvironment).find(
    ([_, env]) => env.apiUrl === apiUrl
  );
  return matchingEnvironment ? matchingEnvironment[0] : "custom";
}
var DelegationStorageClient = class {
  #apiVersionPrefix = "api/v0";
  #config;
  #fetcher;
  #apiUrl;
  constructor(config) {
    const { apiUrl } = config.environment;
    _chunkW6HPAK4Mcjs.trackSmartAccountsKitFunctionCall.call(void 0, 
      "experimental.DelegationStorageClient.constructor",
      {
        environment: getDelegationStorageEnvironment(apiUrl)
      }
    );
    if (apiUrl.endsWith(this.#apiVersionPrefix)) {
      this.#apiUrl = apiUrl;
    } else {
      const separator = apiUrl.endsWith("/") ? "" : "/";
      this.#apiUrl = `${apiUrl}${separator}${this.#apiVersionPrefix}`;
    }
    this.#fetcher = this.#initializeFetcher(config);
    this.#config = config;
  }
  /**
   * Initializes the fetch function for HTTP requests.
   *
   * - Uses `config.fetcher` if provided.
   * - Falls back to global `fetch` if available.
   * - Throws an error if no fetch function is available.
   *
   * @param config - Configuration object that may include a custom fetch function.
   * @returns The fetch function to be used for HTTP requests.
   * @throws Error if no fetch function is available in the environment.
   */
  #initializeFetcher(config) {
    if (config.fetcher) {
      return config.fetcher;
    } else if (typeof _optionalChain([globalThis, 'optionalAccess', _2 => _2.fetch]) === "function") {
      return globalThis.fetch.bind(globalThis);
    }
    throw new Error(
      "Fetch API is not available in this environment. Please provide a fetch function in the config."
    );
  }
  /**
   * Fetches the delegation chain from the Delegation Storage Service, ending with
   * the specified leaf delegation.
   *
   * @param leafDelegationOrDelegationHash - The leaf delegation, or the hash
   * of the leaf delegation.
   * @returns A promise that resolves to the delegation chain - empty array if the delegation
   * is not found.
   */
  async getDelegationChain(leafDelegationOrDelegationHash) {
    _chunkW6HPAK4Mcjs.trackSmartAccountsKitFunctionCall.call(void 0, 
      "experimental.DelegationStorageClient.getDelegationChain",
      {
        inputKind: typeof leafDelegationOrDelegationHash === "string" ? "hash" : "delegation"
      }
    );
    const leafDelegationHash = typeof leafDelegationOrDelegationHash === "string" ? leafDelegationOrDelegationHash : _chunkW6HPAK4Mcjs.hashDelegation.call(void 0, leafDelegationOrDelegationHash);
    const response = await this.#fetcher(
      `${this.#apiUrl}/delegation/chain/${leafDelegationHash}`,
      {
        method: "GET",
        headers: {
          Authorization: `Bearer ${this.#config.apiKey}`,
          "x-api-key-id": this.#config.apiKeyId
        }
      }
    );
    const responseData = await response.json();
    if ("error" in responseData) {
      throw new Error(
        `Failed to fetch delegation chain: ${responseData.error}`
      );
    }
    return responseData;
  }
  /**
   * Fetches the delegations from the Delegation Storage Service, either `Received`
   * by, or `Given` by, (or both: `All`) the specified deleGatorAddress. Defaults
   * to `Received`.
   *
   * @param deleGatorAddress - The deleGatorAddress to retrieve the delegations for.
   * @param filterMode - The DelegationStoreFilter mode - defaults to Received.
   * @returns A promise that resolves to the list of delegations received by the deleGatorAddress,
   * empty array if the delegations are not found.
   */
  async fetchDelegations(deleGatorAddress, filterMode = "RECEIVED" /* Received */) {
    _chunkW6HPAK4Mcjs.trackSmartAccountsKitFunctionCall.call(void 0, 
      "experimental.DelegationStorageClient.fetchDelegations",
      { filterMode }
    );
    const response = await this.#fetcher(
      `${this.#apiUrl}/delegation/accounts/${deleGatorAddress}?filter=${filterMode}`,
      {
        method: "GET",
        headers: {
          Authorization: `Bearer ${this.#config.apiKey}`,
          "x-api-key-id": this.#config.apiKeyId
        }
      }
    );
    const responseData = await response.json();
    if ("error" in responseData) {
      throw new Error(`Failed to fetch delegations: ${responseData.error}`);
    }
    return responseData;
  }
  /**
   * Stores the specified delegation in the Delegation Storage Service.
   *
   * @param delegation - The delegation to store.
   * @returns A promise that resolves to the delegation hash indicating successful storage.
   */
  async storeDelegation(delegation) {
    _chunkW6HPAK4Mcjs.trackSmartAccountsKitFunctionCall.call(void 0, 
      "experimental.DelegationStorageClient.storeDelegation",
      { caveatCount: delegation.caveats.length }
    );
    if (!delegation.signature || delegation.signature === "0x") {
      throw new Error("Delegation must be signed to be stored");
    }
    const delegationHash = _chunkW6HPAK4Mcjs.hashDelegation.call(void 0, delegation);
    const body = JSON.stringify(
      {
        ...delegation,
        metadata: []
      },
      (_, value) => typeof value === "bigint" || typeof value === "number" ? _viem.toHex.call(void 0, value) : value,
      2
    );
    const response = await this.#fetcher(`${this.#apiUrl}/delegation/store`, {
      method: "POST",
      headers: {
        Authorization: `Bearer ${this.#config.apiKey}`,
        "x-api-key-id": this.#config.apiKeyId,
        "Content-Type": "application/json"
      },
      body
    });
    const responseData = await response.json();
    if ("error" in responseData) {
      throw new Error(responseData.error);
    }
    if (responseData.delegationHash !== delegationHash) {
      throw new Error(
        "Failed to store the Delegation, the hash returned from the MM delegation storage API does not match the hash of the delegation"
      );
    }
    return responseData.delegationHash;
  }
};

// src/experimental/x402DelegationProviderUtils.ts






var _delegationcore = require('@metamask/delegation-core');
var _utils = require('@metamask/utils');

async function resolveMaybeDeferred(maybeDeferred, requirements) {
  if (typeof maybeDeferred === "function") {
    const deferred = maybeDeferred;
    return await deferred(requirements);
  }
  return maybeDeferred;
}
function parseEip155ChainId(network) {
  const { namespace, reference } = _utils.parseCaipChainId.call(void 0, 
    network
  );
  if (namespace !== "eip155") {
    throw new Error("Unsupported chain namespace");
  }
  const parsedChainId = Number(reference);
  if (isNaN(parsedChainId)) {
    throw new Error("Invalid chain id");
  }
  return parsedChainId;
}
var TRANSFER_PAYEE_INDEX = 4;
var normalizeAddress = (address) => address.toLowerCase();
var hasMatchingCaveats = (caveats, delegations, match) => {
  for (const caveat of caveats) {
    if (match(caveat)) {
      return true;
    }
  }
  for (const delegation of delegations) {
    for (const caveat of delegation.caveats) {
      if (match(caveat)) {
        return true;
      }
    }
  }
  return false;
};
var ensureExpirySufficientlyConstrained = ({
  timestampEnforcer,
  caveats,
  existingDelegations,
  expirySeconds
}) => {
  if (expirySeconds < 0) {
    throw new Error("Expiry seconds must be a positive number");
  }
  const beforeThreshold = Math.floor(Date.now() / 1e3) + expirySeconds;
  const timestampEnforcerNormalized = normalizeAddress(timestampEnforcer);
  const hasSupersedingTimestampConstraint = hasMatchingCaveats(
    caveats,
    existingDelegations,
    (caveat) => {
      if (normalizeAddress(caveat.enforcer) !== timestampEnforcerNormalized) {
        return false;
      }
      const { beforeThreshold: existingBeforeThreshold } = _delegationcore.decodeTimestampTerms.call(void 0, 
        caveat.terms
      );
      return existingBeforeThreshold !== 0 && existingBeforeThreshold <= beforeThreshold;
    }
  );
  if (hasSupersedingTimestampConstraint) {
    return caveats;
  }
  const timestampCaveat = {
    enforcer: timestampEnforcer,
    terms: _delegationcore.createTimestampTerms.call(void 0, {
      afterThreshold: 0,
      beforeThreshold
    }),
    args: "0x"
  };
  return [...caveats, timestampCaveat];
};
var ensureRedeemerSufficientlyConstrained = ({
  redeemerEnforcer,
  caveats,
  existingDelegations,
  redeemerAddresses,
  requireRedeemers
}) => {
  const redeemerAddressNormalized = normalizeAddress(redeemerEnforcer);
  if (redeemerAddresses.length === 0) {
    if (!requireRedeemers) {
      return caveats;
    }
    const hasExistingRedeemerCaveat = hasMatchingCaveats(
      caveats,
      existingDelegations,
      ({ enforcer }) => normalizeAddress(enforcer) === redeemerAddressNormalized
    );
    if (!hasExistingRedeemerCaveat) {
      throw new Error(
        "Redeemer must be constrained, either in the specified `caveats`, `parentPermissionContext`, or the `PaymentRequirements` as `extra.facilitatorAddresses`."
      );
    }
    return caveats;
  }
  const redeemerAddressesNormalized = redeemerAddresses.map(normalizeAddress);
  const redeemerAddressesSet = new Set(redeemerAddressesNormalized);
  const hasSupersedingRedeemerCaveat = hasMatchingCaveats(
    caveats,
    existingDelegations,
    (caveat) => {
      if (normalizeAddress(caveat.enforcer) !== redeemerAddressNormalized) {
        return false;
      }
      const allowedRedeemerAddresses = _delegationcore.decodeRedeemerTerms.call(void 0, 
        caveat.terms
      ).redeemers.map(normalizeAddress);
      return allowedRedeemerAddresses.every(
        (item) => redeemerAddressesSet.has(item)
      );
    }
  );
  if (hasSupersedingRedeemerCaveat) {
    return caveats;
  }
  const redeemerCaveat = {
    enforcer: redeemerEnforcer,
    terms: _delegationcore.createRedeemerTerms.call(void 0, { redeemers: redeemerAddresses }),
    args: "0x"
  };
  return [...caveats, redeemerCaveat];
};
var ensurePayeeSufficientlyConstrained = ({
  allowedCalldataEnforcer,
  caveats,
  existingDelegations,
  payee
}) => {
  const allowedCalldataTerms = _delegationcore.createAllowedCalldataTerms.call(void 0, {
    startIndex: TRANSFER_PAYEE_INDEX,
    value: _viem.pad.call(void 0, payee, { size: 32 })
  });
  const allowedCalldataEnforcerNormalized = normalizeAddress(
    allowedCalldataEnforcer
  );
  const lowercaseCalldataTerms = allowedCalldataTerms.toLowerCase();
  const hasSupersedingAllowedCalldataConstraint = hasMatchingCaveats(
    caveats,
    existingDelegations,
    ({ enforcer, terms }) => normalizeAddress(enforcer) === allowedCalldataEnforcerNormalized && terms.toLowerCase() === lowercaseCalldataTerms
  );
  if (hasSupersedingAllowedCalldataConstraint) {
    return caveats;
  }
  const payeeCaveat = {
    enforcer: allowedCalldataEnforcer,
    terms: allowedCalldataTerms,
    args: "0x"
  };
  return [...caveats, payeeCaveat];
};
var resolvex402DelegationCaveats = ({
  environment,
  caveatsConfig,
  existingDelegations,
  facilitatorAddresses,
  payee,
  expirySeconds,
  requireRedeemers,
  redeemerAddresses
}) => {
  const {
    caveatEnforcers: {
      RedeemerEnforcer: redeemerEnforcer,
      AllowedCalldataEnforcer: allowedCalldataEnforcer,
      TimestampEnforcer: timestampEnforcer
    }
  } = environment;
  if (!redeemerEnforcer) {
    throw new Error("RedeemerEnforcer not found in environment");
  }
  if (!allowedCalldataEnforcer) {
    throw new Error("AllowedCalldataEnforcer not found in environment");
  }
  if (!timestampEnforcer) {
    throw new Error("TimestampEnforcer not found in environment");
  }
  const initialCaveats = _chunkW6HPAK4Mcjs.resolveCaveats.call(void 0, {
    environment,
    caveats: caveatsConfig,
    // Resolve caveats first so downstream constraint checks can append as needed.
    // Scope is still attached later during delegation creation.
    isScopeOptional: true
  });
  const allRedeemerAddresses = new Set(
    [..._nullishCoalesce(facilitatorAddresses, () => ( [])), ..._nullishCoalesce(redeemerAddresses, () => ( []))].map(
      normalizeAddress
    )
  );
  const caveatsWithRedeemer = ensureRedeemerSufficientlyConstrained({
    redeemerEnforcer,
    caveats: initialCaveats,
    existingDelegations,
    redeemerAddresses: Array.from(allRedeemerAddresses),
    requireRedeemers
  });
  const caveatsWithPayee = ensurePayeeSufficientlyConstrained({
    allowedCalldataEnforcer,
    caveats: caveatsWithRedeemer,
    existingDelegations,
    payee
  });
  if (expirySeconds === void 0) {
    return caveatsWithPayee;
  }
  return ensureExpirySufficientlyConstrained({
    timestampEnforcer,
    caveats: caveatsWithPayee,
    existingDelegations,
    expirySeconds
  });
};
var resolveDelegationCreationContext = async (config, requirements) => {
  const account = await resolveMaybeDeferred(config.account, requirements);
  const redeemersConfig = await resolveMaybeDeferred(
    config.redeemers,
    requirements
  );
  const requireRedeemers = _nullishCoalesce(_optionalChain([redeemersConfig, 'optionalAccess', _3 => _3.requireRedeemers]), () => ( false));
  const redeemerAddresses = await resolveMaybeDeferred(
    _optionalChain([redeemersConfig, 'optionalAccess', _4 => _4.addresses]),
    requirements
  );
  const specifiedEnvironment = await resolveMaybeDeferred(
    config.environment,
    requirements
  );
  const environment = _nullishCoalesce(specifiedEnvironment, () => ( _chunk2OS36DVVcjs.getSmartAccountsEnvironment.call(void 0, parseEip155ChainId(requirements.network))));
  const caveatsConfig = await resolveMaybeDeferred(
    config.caveats,
    requirements
  );
  const parentPermissionContext = await resolveMaybeDeferred(config.parentPermissionContext, requirements);
  const expirySeconds = await resolveMaybeDeferred(
    config.expirySeconds,
    requirements
  );
  const from = await _asyncNullishCoalesce(await resolveMaybeDeferred(config.from, requirements), async () => ( account.address));
  const salt = await _asyncNullishCoalesce(await resolveMaybeDeferred(config.salt, requirements), async () => ( _chunkIQT6XFDTcjs.generateSalt.call(void 0, )));
  const scope = {
    type: "erc20TransferAmount" /* Erc20TransferAmount */,
    tokenAddress: requirements.asset,
    maxAmount: BigInt(requirements.amount)
  };
  const facilitatorAddresses = _optionalChain([requirements, 'access', _5 => _5.extra, 'optionalAccess', _6 => _6.facilitatorAddresses]);
  const existingDelegations = parentPermissionContext ? _chunkW6HPAK4Mcjs.decodeDelegations.call(void 0, parentPermissionContext) : [];
  const { DelegationManager: delegationManager } = environment;
  const caveats = resolvex402DelegationCaveats({
    environment,
    caveatsConfig,
    existingDelegations,
    facilitatorAddresses,
    payee: requirements.payTo,
    expirySeconds,
    requireRedeemers,
    redeemerAddresses
  });
  let createDelegationConfig;
  if (parentPermissionContext) {
    const parentDelegation = existingDelegations[0];
    if (!parentDelegation) {
      throw new Error("Parent permission context is not a valid delegation");
    }
    createDelegationConfig = {
      environment,
      from,
      caveats,
      salt,
      scope,
      parentDelegation
    };
  } else {
    createDelegationConfig = {
      environment,
      from,
      caveats,
      salt,
      scope
    };
  }
  const rootDelegator = _nullishCoalesce(_optionalChain([existingDelegations, 'access', _7 => _7[existingDelegations.length - 1], 'optionalAccess', _8 => _8.delegator]), () => ( from));
  return {
    account,
    createDelegationConfig,
    delegationManager,
    existingDelegations,
    rootDelegator
  };
};

// src/experimental/x402DelegationProvider.ts
function createx402DelegationProvider(config) {
  return async (requirements) => {
    const {
      account,
      createDelegationConfig,
      delegationManager,
      existingDelegations,
      rootDelegator
    } = await resolveDelegationCreationContext(config, requirements);
    const delegation = _chunkW6HPAK4Mcjs.createOpenDelegation.call(void 0, createDelegationConfig);
    const chainId = parseEip155ChainId(requirements.network);
    const typedData = _chunkW6HPAK4Mcjs.prepareSignDelegationTypedData.call(void 0, {
      delegationManager,
      chainId,
      delegation
    });
    if (!account.signTypedData) {
      throw new Error("Account does not support signTypedData");
    }
    const signature = await account.signTypedData(typedData);
    const signedDelegation = {
      ...delegation,
      signature
    };
    const permissionContext = _chunkW6HPAK4Mcjs.encodeDelegations.call(void 0, [
      signedDelegation,
      ...existingDelegations
    ]);
    return {
      delegationManager,
      permissionContext,
      delegator: rootDelegator
    };
  };
}





exports.DelegationStorageClient = DelegationStorageClient; exports.DelegationStorageEnvironment = DelegationStorageEnvironment; exports.createx402DelegationProvider = createx402DelegationProvider; exports.parseEip155ChainId = parseEip155ChainId;
